
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Walkthrough for Whole Organisms Mutliplexed Data Analysis">
      
      
      
        <meta name="author" content="Tara Chari">
      
      
        <link rel="canonical" href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/notebooks/filterStarvCells_ClickTags/">
      
      <link rel="shortcut icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.2.7">
    
    
      
        <title>Initial cDNA Filtering with Cell Ranger Data (Starvation Data) - WHAMSeqAnalysis</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.cb6bc1d0.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.39b8e14a.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#download-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/" title="WHAMSeqAnalysis" class="md-header-nav__button md-logo" aria-label="WHAMSeqAnalysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      <div class="md-header-nav__ellipsis">
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            WHAMSeqAnalysis
          </span>
        </div>
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            
              Initial cDNA Filtering with Cell Ranger Data (Starvation Data)
            
          </span>
        </div>
      </div>
    </div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    




<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/" title="WHAMSeqAnalysis" class="md-nav__button md-logo" aria-label="WHAMSeqAnalysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    WHAMSeqAnalysis
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" >
      
      <label class="md-nav__link" for="nav-2">
        ClickTag Pre-processing
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="ClickTag Pre-processing" data-md-level="1">
        <label class="md-nav__title" for="nav-2">
          <span class="md-nav__icon md-icon"></span>
          ClickTag Pre-processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellRangerClickTagCounts/" class="md-nav__link">
        Initial Quantification of ClickTag Counts with Cell Ranger (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistobusStarvClickTags/" class="md-nav__link">
        Quantification of ClickTag Counts with kallisto-bustools (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistobusStimClickTags/" class="md-nav__link">
        Quantification of ClickTag Counts with kallisto-bustools (Stimulation Data)
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3" checked>
      
      <label class="md-nav__link" for="nav-3">
        cDNA Pre-processing
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="cDNA Pre-processing" data-md-level="1">
        <label class="md-nav__title" for="nav-3">
          <span class="md-nav__icon md-icon"></span>
          cDNA Pre-processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Initial cDNA Filtering with Cell Ranger Data (Starvation Data)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Initial cDNA Filtering with Cell Ranger Data (Starvation Data)
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import Packages
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#select-cells-for-downstream-analysis-by-associated-clicktag-counts" class="md-nav__link">
    Select Cells for Downstream Analysis by Associated ClickTag Counts
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistoBusRuns_StarvAndStim/" class="md-nav__link">
        Quantification of cDNA Counts with kallisto-bustools (All Data)
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4" >
      
      <label class="md-nav__link" for="nav-4">
        Cell Atlas and Perturbation Response
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Cell Atlas and Perturbation Response" data-md-level="1">
        <label class="md-nav__title" for="nav-4">
          <span class="md-nav__icon md-icon"></span>
          Cell Atlas and Perturbation Response
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellRangerClustering_Starvation/" class="md-nav__link">
        Initial Cell Atlas Clustering from Cell Ranger
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../starvation_Analysis/" class="md-nav__link">
        Cell Atlas Clustering and Analysis from kallisto-bustools
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../neuronSubpop_Analysis/" class="md-nav__link">
        Analysis of Neuron Subpopulations
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../pseudotime_Analysis/" class="md-nav__link">
        Pseudotime Analysis
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../deSeq2Analysis_StarvationResponse/" class="md-nav__link">
        Extraction of Differentially Expressed Genes Under Starvation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5" >
      
      <label class="md-nav__link" for="nav-5">
        Stimulation Data Analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Stimulation Data Analysis" data-md-level="1">
        <label class="md-nav__title" for="nav-5">
          <span class="md-nav__icon md-icon"></span>
          Stimulation Data Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../stimulation_Analysis/" class="md-nav__link">
        Cell Atlas Concordance and Perturbation Response
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../deSeq2Analysis_StimulationResponse/" class="md-nav__link">
        Extraction of Differentially Expressed Genes Under Starvation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6" >
      
      <label class="md-nav__link" for="nav-6">
        Inter-Experimental Analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Inter-Experimental Analysis" data-md-level="1">
        <label class="md-nav__title" for="nav-6">
          <span class="md-nav__icon md-icon"></span>
          Inter-Experimental Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../allDistanceCalculations/" class="md-nav__link">
        Comparative Analysis of Cell Atlas Concordance and Perturbation Responses
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-7" type="checkbox" id="nav-7" >
      
      <label class="md-nav__link" for="nav-7">
        Explore Mutliplexed Datasets
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Explore Mutliplexed Datasets" data-md-level="1">
        <label class="md-nav__title" for="nav-7">
          <span class="md-nav__icon md-icon"></span>
          Explore Mutliplexed Datasets
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../MARIMBAAnnosAnalysis/" class="md-nav__link">
        Generation Cell Atlas with MARIMBA Transcriptome
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../MARIMBAAtlasSearchAndPlot/" class="md-nav__link">
        Explore MARIMBA-Annotated Cell Atlas
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellAtlasSearchAndPlot/" class="md-nav__link">
        Explore Original Cell Atlas
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import Packages
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#select-cells-for-downstream-analysis-by-associated-clicktag-counts" class="md-nav__link">
    Select Cells for Downstream Analysis by Associated ClickTag Counts
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/edit/master/docs/notebooks/filterStarvCells_ClickTags.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                  <h1>Initial cDNA Filtering with Cell Ranger Data (Starvation Data)</h1>
                
                <p><a href="https://colab.research.google.com/github/pachterlab/CWGFLHGCCHAP_2021/blob/master/notebooks/Preprocessing/cDNAFiltering/filterStarvCells_ClickTags.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!date
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">Sat Dec  5 00:33:19 UTC 2020</span>
</code></pre></div>

<h3 id="download-data"><strong>Download Data</strong><a class="headerlink" href="#download-data" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import requests
from tqdm import tnrange, tqdm_notebook
def download_file(doi,ext):
    url = &#39;https://api.datacite.org/dois/&#39;+doi+&#39;/media&#39;
    r = requests.get(url).json()
    netcdf_url = r[&#39;data&#39;][0][&#39;attributes&#39;][&#39;url&#39;]
    r = requests.get(netcdf_url,stream=True)
    #Set file name
    fname = doi.split(&#39;/&#39;)[-1]+ext
    #Download file with progress bar
    if r.status_code == 403:
        print(&quot;File Unavailable&quot;)
    if &#39;content-length&#39; not in r.headers:
        print(&quot;Did not get file&quot;)
    else:
        with open(fname, &#39;wb&#39;) as f:
            total_length = int(r.headers.get(&#39;content-length&#39;))
            pbar = tnrange(int(total_length/1024), unit=&quot;B&quot;)
            for chunk in r.iter_content(chunk_size=1024):
                if chunk:
                    pbar.update()
                    f.write(chunk)
        return fname
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#From CellRanger (cDNA)
#matrix
download_file(&#39;10.22002/D1.1802&#39;,&#39;.gz&#39;)

#features
download_file(&#39;10.22002/D1.1803&#39;,&#39;.gz&#39;)

#barcodes
download_file(&#39;10.22002/D1.1801&#39;,&#39;.gz&#39;)

#ClickTag lane 1 counts
download_file(&#39;10.22002/D1.1799&#39;,&#39;.gz&#39;)

#ClickTag lane 2 counts
download_file(&#39;10.22002/D1.1800&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">111488</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">329</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>






<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">4453</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">2984</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">2737</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1800.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!gunzip *.gz
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!pip install --quiet anndata
!pip install --quiet scanpy
!pip3 install --quiet leidenalg
!pip install --quiet louvain
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 122kB 4.3MB/s </span>
<span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 7.7MB 3.9MB/s </span>
<span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 71kB 9.4MB/s </span>
<span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 51kB 7.1MB/s </span>
<span class="err">[?25h  Building wheel for sinfo (setup.py) ... [?25l[?25hdone</span>
<span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 2.4MB 4.3MB/s </span>
<span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3.2MB 43.5MB/s </span>
<span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 2.2MB 4.4MB/s </span>
<span class="err">[?25h</span>
</code></pre></div>

<h3 id="import-packages"><strong>Import Packages</strong><a class="headerlink" href="#import-packages" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import pandas as pd
import anndata
import scanpy as sc
import numpy as np
import scipy.sparse

import warnings
warnings.filterwarnings(&#39;ignore&#39;)

from sklearn.neighbors import (KNeighborsClassifier,NeighborhoodComponentsAnalysis)
from sklearn.pipeline import Pipeline
from sklearn.manifold import TSNE
from sklearn.decomposition import PCA

from sklearn.preprocessing import scale

import random

import matplotlib.pyplot as plt
%matplotlib inline
sc.set_figure_params(dpi=125)

import seaborn as sns
sns.set(style=&quot;whitegrid&quot;)
%load_ext rpy2.ipython
</code></pre></div>
</td></tr></table>
<h3 id="select-cells-for-downstream-analysis-by-associated-clicktag-counts"><strong>Select Cells for Downstream Analysis by Associated ClickTag Counts</strong><a class="headerlink" href="#select-cells-for-downstream-analysis-by-associated-clicktag-counts" title="Permanent link">&para;</a></h3>
<p>Filtering of CellRanger cDNA data by ClickTag counts</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>path = &quot;&quot; 
!mv D1.1802 D1.1802.mtx
!mv D1.1803 D1.1803.tsv
!mv D1.1801 D1.1801.tsv

jelly3trin = sc.read(path+&#39;D1.1802.mtx&#39;, cache=True).T
jelly3trin.var_names = pd.read_csv(path+&#39;D1.1803.tsv&#39;, header=None, sep=&#39;\t&#39;)[1]
jelly3trin.obs_names = pd.read_csv(path+&#39;D1.1801.tsv&#39;, header=None)[0]
jelly3trin
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">1474560</span> <span class="err">Ã—</span> <span class="mi">46716</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>jelly3trin.var_names_make_unique()
sc.pp.filter_cells(jelly3trin,min_counts=1)
jelly3trin.obs[&#39;n_countslog&#39;]=np.log10(jelly3trin.obs[&#39;n_counts&#39;])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>jelly3trin
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">882404</span> <span class="err">Ã—</span> <span class="mi">46716</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span>
</code></pre></div>

<p>Selecting well-tagged cells by counts of ClickTags on embedding of cell x ClickTag counts matrix (previously calculated)</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!mv D1.1799 tag1_counts.csv
adata = sc.read_csv(&quot;tag1_counts.csv&quot;)
adata=adata.transpose()

sc.pp.filter_cells(adata, min_counts=100)
adata.obs[&#39;n_countslog&#39;] = np.log(adata.obs[&#39;n_counts&#39;])

sc.pp.log1p(adata)

#For visualization of tag counts
sc.tl.pca(adata)
sc.tl.tsne(adata)
sc.pp.neighbors(adata)
sc.tl.louvain(adata, resolution=0.35)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">WARNING</span><span class="o">:</span> <span class="n">Consider</span> <span class="n">installing</span> <span class="n">the</span> <span class="k">package</span> <span class="nn">MulticoreTSNE</span> <span class="o">(</span><span class="n">https</span><span class="o">://</span><span class="n">github</span><span class="o">.</span><span class="na">com</span><span class="sr">/DmitryUlyanov/</span><span class="n">Multicore</span><span class="o">-</span><span class="n">TSNE</span><span class="o">).</span> <span class="n">Even</span> <span class="k">for</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span> <span class="k">this</span> <span class="n">speeds</span> <span class="n">up</span> <span class="n">the</span> <span class="n">computation</span> <span class="n">considerably</span> <span class="n">and</span> <span class="n">might</span> <span class="n">yield</span> <span class="n">better</span> <span class="n">converged</span> <span class="n">results</span><span class="o">.</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.louvain(adata, resolution=0.35)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.tsne(adata, color=[&#39;louvain&#39;])
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../filterStarvCells_ClickTags_files/filterStarvCells_ClickTags_17_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.tsne(adata, color=[&#39;BC_21&#39;, &#39;BC_22&#39;,&#39;BC_23&#39;,&#39;BC_24&#39;,&#39;BC_25&#39;,&#39;BC_26&#39;,&#39;BC_27&#39;,&#39;BC_28&#39;,&#39;BC_29&#39;,&#39;BC_30&#39;,&#39;BC_31&#39;,&#39;BC_32&#39;,&#39;BC_33&#39;,
                         &#39;BC_34&#39;, &#39;BC_35&#39;,&#39;BC_36&#39;,&#39;BC_37&#39;,&#39;BC_38&#39;,&#39;BC_39&#39;,&#39;BC_40&#39;,&#39;n_countslog&#39;])
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../filterStarvCells_ClickTags_files/filterStarvCells_ClickTags_18_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!mv D1.1800 tag2_counts.csv
adata2 = sc.read_csv(&quot;tag2_counts.csv&quot;)
adata2=adata2.transpose()

sc.pp.filter_cells(adata2, min_counts=80)
adata2.obs[&#39;n_countslog&#39;] = np.log(adata2.obs[&#39;n_counts&#39;])

sc.pp.log1p(adata2)

#For visualization of tag counts
sc.tl.pca(adata2)
sc.tl.tsne(adata2)
sc.pp.neighbors(adata2)
sc.tl.louvain(adata2, resolution=0.35)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">WARNING</span><span class="o">:</span> <span class="n">Consider</span> <span class="n">installing</span> <span class="n">the</span> <span class="k">package</span> <span class="nn">MulticoreTSNE</span> <span class="o">(</span><span class="n">https</span><span class="o">://</span><span class="n">github</span><span class="o">.</span><span class="na">com</span><span class="sr">/DmitryUlyanov/</span><span class="n">Multicore</span><span class="o">-</span><span class="n">TSNE</span><span class="o">).</span> <span class="n">Even</span> <span class="k">for</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span> <span class="k">this</span> <span class="n">speeds</span> <span class="n">up</span> <span class="n">the</span> <span class="n">computation</span> <span class="n">considerably</span> <span class="n">and</span> <span class="n">might</span> <span class="n">yield</span> <span class="n">better</span> <span class="n">converged</span> <span class="n">results</span><span class="o">.</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.tsne(adata2, color=[&#39;BC_21&#39;, &#39;BC_22&#39;,&#39;BC_23&#39;,&#39;BC_24&#39;,&#39;BC_25&#39;,&#39;BC_26&#39;,&#39;BC_27&#39;,&#39;BC_28&#39;,&#39;BC_29&#39;,&#39;BC_30&#39;,&#39;BC_31&#39;,&#39;BC_32&#39;,&#39;BC_33&#39;,
                         &#39;BC_34&#39;, &#39;BC_35&#39;,&#39;BC_36&#39;,&#39;BC_37&#39;,&#39;BC_38&#39;,&#39;BC_39&#39;,&#39;BC_40&#39;,&#39;n_countslog&#39;])
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../filterStarvCells_ClickTags_files/filterStarvCells_ClickTags_20_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.tsne(adata2,color=[&#39;louvain&#39;])
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../filterStarvCells_ClickTags_files/filterStarvCells_ClickTags_21_0.png" /></p>
<p>Save cell clusters from lane1 and lane2 ClickTag counts</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Make names unique (For each lane)
adata.obs_names=adata.obs_names+[&#39;-1&#39;]
adata2.obs_names=adata2.obs_names+[&#39;-2&#39;]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Based on tsne Plots assign organism IDs/numbers from experiment (1 - 10 jellies)

adataorgIDdict={&#39;7&#39;:&#39;1&#39;, &#39;4&#39;:&#39;2&#39;, &#39;6&#39;:&#39;3&#39;,&#39;10&#39;:&#39;4&#39;, &#39;1&#39;:&#39;5&#39;, &#39;9&#39;:&#39;6&#39;, &#39;5&#39;:&#39;7&#39;, &#39;3&#39;:&#39;8&#39;, &#39;2&#39;:&#39;9&#39;, &#39;8&#39;:&#39;10&#39; }
adata.obs[&#39;filtered_by_oligos&#39;] = adata.obs[&#39;louvain&#39;].isin([&#39;3&#39;,&#39;4&#39;,&#39;8&#39;,&#39;7&#39;,&#39;5&#39;,&#39;2&#39;,&#39;9&#39;,&#39;10&#39;,&#39;6&#39;,&#39;11&#39;])


adata2orgIDdict={&#39;8&#39;:&#39;2&#39;, &#39;5&#39;:&#39;8&#39;, &#39;7&#39;:&#39;1&#39;,&#39;11&#39;:&#39;6&#39;, &#39;2&#39;:&#39;5&#39;, &#39;10&#39;:&#39;4&#39;, &#39;6&#39;:&#39;3&#39;, &#39;4&#39;:&#39;7&#39;, &#39;3&#39;:&#39;9&#39;, &#39;9&#39;:&#39;10&#39;, &#39;louvain&#39;:&#39;org&#39;}
adata2.obs[&#39;filtered_by_oligos&#39;] = adata2.obs[&#39;louvain&#39;].isin([&#39;2&#39;,&#39;3&#39;,&#39;4&#39;,&#39;7&#39;,&#39;5&#39;,&#39;9&#39;,&#39;10&#39;,&#39;6&#39;,&#39;11&#39;,&#39;8&#39;])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>HDfilteradata = adata[adata.obs[&#39;louvain&#39;].isin([&#39;3&#39;,&#39;4&#39;,&#39;8&#39;,&#39;7&#39;,&#39;5&#39;,&#39;2&#39;,&#39;9&#39;,&#39;10&#39;,&#39;6&#39;,&#39;1&#39;])]
HDfilteradata2 = adata2[adata2.obs[&#39;louvain&#39;].isin([&#39;2&#39;,&#39;3&#39;,&#39;4&#39;,&#39;7&#39;,&#39;5&#39;,&#39;9&#39;,&#39;10&#39;,&#39;6&#39;,&#39;11&#39;,&#39;8&#39;])]

HDfilteradata.obs[&#39;orgID&#39;]=[adataorgIDdict.get(x) for x in list(HDfilteradata.obs[&#39;louvain&#39;])]
HDfilteradata2.obs[&#39;orgID&#39;]=[adata2orgIDdict.get(x) for x in list(HDfilteradata2.obs[&#39;louvain&#39;])]
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">Trying to set attribute `.obs` of view, copying.</span>
<span class="err">Trying to set attribute `.obs` of view, copying.</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>ojelly3trin=jelly3trin[jelly3trin.obs_names.isin(list(HDfilteradata.obs_names)+list(HDfilteradata2.obs_names))]
sc.pp.filter_cells(ojelly3trin, min_counts=0)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">View</span> <span class="n">of</span> <span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">16818</span> <span class="err">Ã—</span> <span class="mi">46716</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>holder=pd.DataFrame()
#I need to make dictionaries to map the organismIDs from both 10x lanes back to my experimental conditions
for cell in list(ojelly3trin.obs_names):
    if cell[-1:]==&#39;1&#39;:
        holder[cell]=[list(HDfilteradata.obs[&#39;orgID&#39;])[list(HDfilteradata.obs_names).index(cell)]]
    elif cell[-1:]==&#39;2&#39;:
        holder[cell]=[list(HDfilteradata2.obs[&#39;orgID&#39;])[list(HDfilteradata2.obs_names).index(cell)]]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>holder.head()
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>AAACCTGAGAGGGCTT-1</th>
      <th>AAACCTGAGAGGTTAT-1</th>
      <th>AAACCTGAGCGAGAAA-1</th>
      <th>AAACCTGAGCGTGTCC-1</th>
      <th>AAACCTGAGGATTCGG-1</th>
      <th>AAACCTGAGGTGATAT-1</th>
      <th>AAACCTGCAAAGTGCG-1</th>
      <th>AAACCTGCAAGCTGGA-1</th>
      <th>AAACCTGCAATCACAC-1</th>
      <th>AAACCTGCATAACCTG-1</th>
      <th>AAACCTGCATGCAACT-1</th>
      <th>AAACCTGCATGCAATC-1</th>
      <th>AAACCTGGTACATGTC-1</th>
      <th>AAACCTGGTAGTACCT-1</th>
      <th>AAACCTGGTCAGGACA-1</th>
      <th>AAACCTGGTCGGCACT-1</th>
      <th>AAACCTGGTCGGCTCA-1</th>
      <th>AAACCTGGTCTTCTCG-1</th>
      <th>AAACCTGGTGCAGTAG-1</th>
      <th>AAACCTGTCAACTCTT-1</th>
      <th>AAACCTGTCAGAAATG-1</th>
      <th>AAACCTGTCAGGTTCA-1</th>
      <th>AAACCTGTCATATCGG-1</th>
      <th>AAACCTGTCCTAGTGA-1</th>
      <th>AAACCTGTCGGAAATA-1</th>
      <th>AAACCTGTCGGATGGA-1</th>
      <th>AAACCTGTCTCTAGGA-1</th>
      <th>AAACGGGAGCTGCCCA-1</th>
      <th>AAACGGGAGTGAACAT-1</th>
      <th>AAACGGGCAAACCCAT-1</th>
      <th>AAACGGGCAAAGCAAT-1</th>
      <th>AAACGGGGTCATATCG-1</th>
      <th>AAACGGGGTCCTGCTT-1</th>
      <th>AAACGGGGTGCATCTA-1</th>
      <th>AAACGGGGTTCCGTCT-1</th>
      <th>AAACGGGGTTCGTGAT-1</th>
      <th>AAACGGGTCATTGCCC-1</th>
      <th>AAACGGGTCCACTGGG-1</th>
      <th>AAACGGGTCTAACGGT-1</th>
      <th>AAACGGGTCTGACCTC-1</th>
      <th>...</th>
      <th>TTTGCGCGTATATGAG-2</th>
      <th>TTTGCGCGTCCAGTAT-2</th>
      <th>TTTGCGCGTCCTCTTG-2</th>
      <th>TTTGCGCGTCTTGATG-2</th>
      <th>TTTGCGCGTTATCCGA-2</th>
      <th>TTTGCGCGTTGGTAAA-2</th>
      <th>TTTGCGCTCAACCAAC-2</th>
      <th>TTTGCGCTCAGCTGGC-2</th>
      <th>TTTGCGCTCAGTCAGT-2</th>
      <th>TTTGCGCTCGCCCTTA-2</th>
      <th>TTTGCGCTCTTAGCCC-2</th>
      <th>TTTGGTTAGACTAGAT-2</th>
      <th>TTTGGTTAGCGCCTTG-2</th>
      <th>TTTGGTTAGCTAACAA-2</th>
      <th>TTTGGTTAGGACTGGT-2</th>
      <th>TTTGGTTCACAACGCC-2</th>
      <th>TTTGGTTCACAAGCCC-2</th>
      <th>TTTGGTTCAGCCAATT-2</th>
      <th>TTTGGTTCATTCACTT-2</th>
      <th>TTTGGTTGTCGAGATG-2</th>
      <th>TTTGGTTGTGTATGGG-2</th>
      <th>TTTGGTTTCAGTGCAT-2</th>
      <th>TTTGGTTTCCAAGCCG-2</th>
      <th>TTTGGTTTCCCGGATG-2</th>
      <th>TTTGTCAAGAGATGAG-2</th>
      <th>TTTGTCAAGCAGGTCA-2</th>
      <th>TTTGTCAAGGGCATGT-2</th>
      <th>TTTGTCACAAACGCGA-2</th>
      <th>TTTGTCACACTCTGTC-2</th>
      <th>TTTGTCACACTTAACG-2</th>
      <th>TTTGTCACAGAAGCAC-2</th>
      <th>TTTGTCAGTACCTACA-2</th>
      <th>TTTGTCAGTATGGTTC-2</th>
      <th>TTTGTCAGTCATTAGC-2</th>
      <th>TTTGTCAGTCTCCACT-2</th>
      <th>TTTGTCAGTGGACGAT-2</th>
      <th>TTTGTCATCATCTGTT-2</th>
      <th>TTTGTCATCGCCTGAG-2</th>
      <th>TTTGTCATCTAACGGT-2</th>
      <th>TTTGTCATCTGAGTGT-2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>3</td>
      <td>9</td>
      <td>3</td>
      <td>10</td>
      <td>8</td>
      <td>7</td>
      <td>6</td>
      <td>7</td>
      <td>9</td>
      <td>3</td>
      <td>4</td>
      <td>5</td>
      <td>9</td>
      <td>5</td>
      <td>9</td>
      <td>9</td>
      <td>5</td>
      <td>6</td>
      <td>4</td>
      <td>8</td>
      <td>2</td>
      <td>9</td>
      <td>2</td>
      <td>7</td>
      <td>5</td>
      <td>2</td>
      <td>9</td>
      <td>10</td>
      <td>1</td>
      <td>9</td>
      <td>3</td>
      <td>2</td>
      <td>1</td>
      <td>9</td>
      <td>5</td>
      <td>1</td>
      <td>1</td>
      <td>10</td>
      <td>3</td>
      <td>9</td>
      <td>...</td>
      <td>5</td>
      <td>7</td>
      <td>2</td>
      <td>2</td>
      <td>7</td>
      <td>7</td>
      <td>3</td>
      <td>8</td>
      <td>10</td>
      <td>2</td>
      <td>4</td>
      <td>2</td>
      <td>3</td>
      <td>9</td>
      <td>10</td>
      <td>9</td>
      <td>5</td>
      <td>5</td>
      <td>7</td>
      <td>9</td>
      <td>4</td>
      <td>6</td>
      <td>10</td>
      <td>10</td>
      <td>2</td>
      <td>10</td>
      <td>7</td>
      <td>5</td>
      <td>9</td>
      <td>7</td>
      <td>4</td>
      <td>5</td>
      <td>4</td>
      <td>7</td>
      <td>9</td>
      <td>4</td>
      <td>7</td>
      <td>7</td>
      <td>3</td>
      <td>9</td>
    </tr>
  </tbody>
</table>
<p>1 rows Ã— 16818 columns</p>
</div>

<p>Save condition labels for cells based on ClickTag assignment</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import copy
ojelly3trin.obs[&#39;temp&#39;]=holder.iloc[0]
ojelly3trin.obs[&#39;orgID&#39;]=copy.deepcopy(ojelly3trin.obs[&#39;temp&#39;])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Color by Condition Labels
ojelly3trin.obs[&#39;fed&#39;]=ojelly3trin.obs[&#39;orgID&#39;].isin([&#39;1&#39;,&#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;])

ojelly3trin.obs[&#39;starved&#39;]=ojelly3trin.obs[&#39;orgID&#39;].isin([&#39;6&#39;,&#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>ojelly3trin.write(&#39;jelly3trin_conditions.h5ad&#39;)
pd.DataFrame(ojelly3trin.obs_names).to_csv(&quot;jelly3trin_barcodes.csv&quot;, index=None)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">... storing &#39;temp&#39; as categorical</span>
<span class="err">... storing &#39;orgID&#39; as categorical</span>
</code></pre></div>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../kallistobusStimClickTags/" class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Quantification of ClickTag Counts with kallisto-bustools (Stimulation Data)
              </div>
            </div>
          </a>
        
        
          <a href="../kallistoBusRuns_StarvAndStim/" class="md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Quantification of cDNA Counts with kallisto-bustools (All Data)
              </div>
            </div>
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/vendor.53cc9318.min.js"></script>
      <script src="../../assets/javascripts/bundle.e9c9f54f.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "../..",
          features: ['search.suggest', 'search.highlight'],
          search: Object.assign({
            worker: "../../assets/javascripts/worker/search.9c0e82ba.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>