
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Walkthrough for Whole Organisms Mutliplexed Data Analysis">
      
      
      
        <meta name="author" content="Tara Chari">
      
      
        <link rel="canonical" href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/notebooks/MARIMBAAnnosAnalysis/">
      
      <link rel="shortcut icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.2.7">
    
    
      
        <title>Generation Cell Atlas with MARIMBA Transcriptome - WHAMSeqAnalysis</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.cb6bc1d0.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.39b8e14a.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#download-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/" title="WHAMSeqAnalysis" class="md-header-nav__button md-logo" aria-label="WHAMSeqAnalysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      <div class="md-header-nav__ellipsis">
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            WHAMSeqAnalysis
          </span>
        </div>
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            
              Generation Cell Atlas with MARIMBA Transcriptome
            
          </span>
        </div>
      </div>
    </div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    




<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/" title="WHAMSeqAnalysis" class="md-nav__button md-logo" aria-label="WHAMSeqAnalysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    WHAMSeqAnalysis
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" >
      
      <label class="md-nav__link" for="nav-2">
        ClickTag Pre-processing
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="ClickTag Pre-processing" data-md-level="1">
        <label class="md-nav__title" for="nav-2">
          <span class="md-nav__icon md-icon"></span>
          ClickTag Pre-processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellRangerClickTagCounts/" class="md-nav__link">
        Initial Quantification of ClickTag Counts with Cell Ranger (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistobusStarvClickTags/" class="md-nav__link">
        Quantification of ClickTag Counts with kallisto-bustools (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistobusStimClickTags/" class="md-nav__link">
        Quantification of ClickTag Counts with kallisto-bustools (Stimulation Data)
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3" >
      
      <label class="md-nav__link" for="nav-3">
        cDNA Pre-processing
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="cDNA Pre-processing" data-md-level="1">
        <label class="md-nav__title" for="nav-3">
          <span class="md-nav__icon md-icon"></span>
          cDNA Pre-processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../filterStarvCells_ClickTags/" class="md-nav__link">
        Initial cDNA Filtering with Cell Ranger Data (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistoBusRuns_StarvAndStim/" class="md-nav__link">
        Quantification of cDNA Counts with kallisto-bustools (All Data)
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4" >
      
      <label class="md-nav__link" for="nav-4">
        Cell Atlas and Perturbation Response
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Cell Atlas and Perturbation Response" data-md-level="1">
        <label class="md-nav__title" for="nav-4">
          <span class="md-nav__icon md-icon"></span>
          Cell Atlas and Perturbation Response
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellRangerClustering_Starvation/" class="md-nav__link">
        Initial Cell Atlas Clustering from Cell Ranger
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../starvation_Analysis/" class="md-nav__link">
        Cell Atlas Clustering and Analysis from kallisto-bustools
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../neuronSubpop_Analysis/" class="md-nav__link">
        Analysis of Neuron Subpopulations
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../pseudotime_Analysis/" class="md-nav__link">
        Pseudotime Analysis
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../deSeq2Analysis_StarvationResponse/" class="md-nav__link">
        Extraction of Differentially Expressed Genes Under Starvation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5" >
      
      <label class="md-nav__link" for="nav-5">
        Stimulation Data Analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Stimulation Data Analysis" data-md-level="1">
        <label class="md-nav__title" for="nav-5">
          <span class="md-nav__icon md-icon"></span>
          Stimulation Data Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../stimulation_Analysis/" class="md-nav__link">
        Cell Atlas Concordance and Perturbation Response
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../deSeq2Analysis_StimulationResponse/" class="md-nav__link">
        Extraction of Differentially Expressed Genes Under Starvation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6" >
      
      <label class="md-nav__link" for="nav-6">
        Inter-Experimental Analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Inter-Experimental Analysis" data-md-level="1">
        <label class="md-nav__title" for="nav-6">
          <span class="md-nav__icon md-icon"></span>
          Inter-Experimental Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../allDistanceCalculations/" class="md-nav__link">
        Comparative Analysis of Cell Atlas Concordance and Perturbation Responses
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-7" type="checkbox" id="nav-7" checked>
      
      <label class="md-nav__link" for="nav-7">
        Explore Mutliplexed Datasets
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Explore Mutliplexed Datasets" data-md-level="1">
        <label class="md-nav__title" for="nav-7">
          <span class="md-nav__icon md-icon"></span>
          Explore Mutliplexed Datasets
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Generation Cell Atlas with MARIMBA Transcriptome
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Generation Cell Atlas with MARIMBA Transcriptome
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import Packages
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-kallisto-bus-on-data-with-marimba-annotation" class="md-nav__link">
    Run kallisto bus on data with MARIMBA annotation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-in-saved-dataset" class="md-nav__link">
    Read in saved dataset
  </a>
  
    <nav class="md-nav" aria-label="Read in saved dataset">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#plot-overlap-in-cell-types" class="md-nav__link">
    Plot Overlap in Cell Types
  </a>
  
    <nav class="md-nav" aria-label="Plot Overlap in Cell Types">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#comparing-gene-markers" class="md-nav__link">
    Comparing Gene Markers
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#neuron-analysis" class="md-nav__link">
    Neuron Analysis
  </a>
  
    <nav class="md-nav" aria-label="Neuron Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#check-if-application-of-trinity-louvain-clusters-is-appropriate" class="md-nav__link">
    Check if application of TRINITY Louvain clusters is appropriate
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../MARIMBAAtlasSearchAndPlot/" class="md-nav__link">
        Explore MARIMBA-Annotated Cell Atlas
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellAtlasSearchAndPlot/" class="md-nav__link">
        Explore Original Cell Atlas
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import Packages
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-kallisto-bus-on-data-with-marimba-annotation" class="md-nav__link">
    Run kallisto bus on data with MARIMBA annotation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-in-saved-dataset" class="md-nav__link">
    Read in saved dataset
  </a>
  
    <nav class="md-nav" aria-label="Read in saved dataset">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#plot-overlap-in-cell-types" class="md-nav__link">
    Plot Overlap in Cell Types
  </a>
  
    <nav class="md-nav" aria-label="Plot Overlap in Cell Types">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#comparing-gene-markers" class="md-nav__link">
    Comparing Gene Markers
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#neuron-analysis" class="md-nav__link">
    Neuron Analysis
  </a>
  
    <nav class="md-nav" aria-label="Neuron Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#check-if-application-of-trinity-louvain-clusters-is-appropriate" class="md-nav__link">
    Check if application of TRINITY Louvain clusters is appropriate
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/edit/master/docs/notebooks/MARIMBAAnnosAnalysis.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                  <h1>Generation Cell Atlas with MARIMBA Transcriptome</h1>
                
                <p><a href="https://colab.research.google.com/github/pachterlab/CWGFLHGCCHAP_2021/blob/master/notebooks/SearchAndPlotInteractive/MARIMBAAnnosAnalysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<h3 id="download-data"><strong>Download Data</strong><a class="headerlink" href="#download-data" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Install kallisto and bustools

!wget --quiet https://github.com/pachterlab/kallisto/releases/download/v0.46.2/kallisto_linux-v0.46.2.tar.gz
!tar -xf kallisto_linux-v0.46.2.tar.gz
!cp kallisto/kallisto /usr/local/bin/

!wget --quiet https://github.com/BUStools/bustools/releases/download/v0.40.0/bustools_linux-v0.40.0.tar.gz
!tar -xf bustools_linux-v0.40.0.tar.gz
!cp bustools/bustools /usr/local/bin/
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import requests
from tqdm import tnrange, tqdm_notebook
def download_file(doi,ext):
    url = &#39;https://api.datacite.org/dois/&#39;+doi+&#39;/media&#39;
    r = requests.get(url).json()
    netcdf_url = r[&#39;data&#39;][0][&#39;attributes&#39;][&#39;url&#39;]
    r = requests.get(netcdf_url,stream=True)
    #Set file name
    fname = doi.split(&#39;/&#39;)[-1]+ext
    #Download file with progress bar
    if r.status_code == 403:
        print(&quot;File Unavailable&quot;)
    if &#39;content-length&#39; not in r.headers:
        print(&quot;Did not get file&quot;)
    else:
        with open(fname, &#39;wb&#39;) as f:
            total_length = int(r.headers.get(&#39;content-length&#39;))
            pbar = tnrange(int(total_length/1024), unit=&quot;B&quot;)
            for chunk in r.iter_content(chunk_size=1024):
                if chunk:
                    pbar.update()
                    f.write(chunk)
        return fname
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Get reference data (transcripts, transcriptome gff, fastq&#39;s)

#Transcripts fasta (Marimba)
download_file(&#39;10.22002/D1.1829&#39;,&#39;.gz&#39;)

#Gff3 (Marimba)
download_file(&#39;10.22002/D1.1830&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">22180</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">4334</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1830.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Get TRINITY analyzed starvation data (with clustering,embedding etc)
download_file(&#39;10.22002/D1.1796&#39;,&#39;.gz&#39;)

#Raw starvation data (all nonzero genes included)
download_file(&#39;10.22002/D1.1797&#39;,&#39;.gz&#39;)

#Neurons data
download_file(&#39;10.22002/D1.1804&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">479630</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">26058</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">1595</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1804.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Raw Marimba-processed data
download_file(&#39;10.22002/D1.1832&#39;,&#39;.gz&#39;)

#Marimba filtered/clustered data
download_file(&#39;10.22002/D1.1833&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">76516</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">520896</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>








<span class="s1">&#39;D1.1833.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># Get doi links for all Starvation cDNA fastq.gz files
starvFiles = []
dois = [&#39;10.22002/D1.1840&#39;,&#39;10.22002/D1.1841&#39;,&#39;10.22002/D1.1842&#39;,&#39;10.22002/D1.1843&#39;,
        &#39;10.22002/D1.1844&#39;,&#39;10.22002/D1.1845&#39;,&#39;10.22002/D1.1846&#39;,&#39;10.22002/D1.1847&#39;,
        &#39;10.22002/D1.1848&#39;,&#39;10.22002/D1.1849&#39;,&#39;10.22002/D1.1850&#39;,&#39;10.22002/D1.1851&#39;,
        &#39;10.22002/D1.1852&#39;,&#39;10.22002/D1.1853&#39;,&#39;10.22002/D1.1854&#39;,&#39;10.22002/D1.1855&#39;] #16 doi numbers
for doi in dois:
  url = &#39;https://api.datacite.org/dois/&#39;+doi+&#39;/media&#39;
  r = requests.get(url).json()
  netcdf_url = r[&#39;data&#39;][0][&#39;attributes&#39;][&#39;url&#39;]

  starvFiles += [netcdf_url]

s1 = starvFiles[0]
s2 = starvFiles[1]
s3 = starvFiles[2]
s4 = starvFiles[3]
s5 = starvFiles[4]
s6 = starvFiles[5]
s7 = starvFiles[6]
s8 = starvFiles[7]

s9 = starvFiles[8]
s10 = starvFiles[9]
s11 = starvFiles[10]
s12 = starvFiles[11]
s13 = starvFiles[12]
s14 = starvFiles[13]
s15 = starvFiles[14]
s16 = starvFiles[15]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!gunzip *.gz
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!pip install --quiet anndata
!pip install --quiet scanpy
!pip install --quiet louvain
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">[K     |████████████████████████████████| 122kB 5.8MB/s </span>
<span class="err">[K     |████████████████████████████████| 7.7MB 4.1MB/s </span>
<span class="err">[K     |████████████████████████████████| 71kB 9.0MB/s </span>
<span class="err">[K     |████████████████████████████████| 51kB 7.8MB/s </span>
<span class="err">[?25h  Building wheel for sinfo (setup.py) ... [?25l[?25hdone</span>
<span class="err">[K     |████████████████████████████████| 2.2MB 5.7MB/s </span>
<span class="err">[K     |████████████████████████████████| 3.2MB 28.4MB/s </span>
<span class="err">[?25h</span>
</code></pre></div>

<h3 id="import-packages"><strong>Import Packages</strong><a class="headerlink" href="#import-packages" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import pandas as pd
import anndata
import scanpy as sc
import numpy as np
import scipy.sparse

import matplotlib.pyplot as plt
%matplotlib inline
sc.set_figure_params(dpi=125)
</code></pre></div>
</td></tr></table>
<h3 id="run-kallisto-bus-on-data-with-marimba-annotation"><strong>Run kallisto bus on data with MARIMBA annotation</strong><a class="headerlink" href="#run-kallisto-bus-on-data-with-marimba-annotation" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Make Kallisto index (reference https://www.kallistobus.tools/getting_started)
!mv D1.1829 transcripts.fa
!kallisto index -i cufflinksTrans_clytia.idx -k 31 transcripts.fa
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="nl">mv</span><span class="p">:</span><span class="w"> </span><span class="n">cannot</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="s1">&#39;D1.1829&#39;</span><span class="err">:</span><span class="w"> </span><span class="k">No</span><span class="w"> </span><span class="n">such</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="ow">or</span><span class="w"> </span><span class="n">directory</span><span class="w"></span>

<span class="o">[</span><span class="n">build</span><span class="o">]</span><span class="w"> </span><span class="n">loading</span><span class="w"> </span><span class="n">fasta</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="n">transcripts</span><span class="p">.</span><span class="n">fa</span><span class="w"></span>
<span class="o">[</span><span class="n">build</span><span class="o">]</span><span class="w"> </span><span class="n">k</span><span class="o">-</span><span class="n">mer</span><span class="w"> </span><span class="nl">length</span><span class="p">:</span><span class="w"> </span><span class="mi">31</span><span class="w"></span>
<span class="o">[</span><span class="n">build</span><span class="o">]</span><span class="w"> </span><span class="nl">warning</span><span class="p">:</span><span class="w"> </span><span class="n">clipped</span><span class="w"> </span><span class="k">off</span><span class="w"> </span><span class="n">poly</span><span class="o">-</span><span class="n">A</span><span class="w"> </span><span class="n">tail</span><span class="w"> </span><span class="p">(</span><span class="n">longer</span><span class="w"> </span><span class="k">than</span><span class="w"> </span><span class="mi">10</span><span class="p">)</span><span class="w"></span>
<span class="w">        </span><span class="k">from</span><span class="w"> </span><span class="mi">16</span><span class="w"> </span><span class="n">target</span><span class="w"> </span><span class="n">sequences</span><span class="w"></span>
<span class="o">[</span><span class="n">build</span><span class="o">]</span><span class="w"> </span><span class="nl">warning</span><span class="p">:</span><span class="w"> </span><span class="n">replaced</span><span class="w"> </span><span class="mi">40896</span><span class="w"> </span><span class="n">non</span><span class="o">-</span><span class="n">ACGUT</span><span class="w"> </span><span class="n">characters</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="k">input</span><span class="w"> </span><span class="k">sequence</span><span class="w"></span>
<span class="w">        </span><span class="k">with</span><span class="w"> </span><span class="n">pseudorandom</span><span class="w"> </span><span class="n">nucleotides</span><span class="w"></span>
<span class="o">[</span><span class="n">build</span><span class="o">]</span><span class="w"> </span><span class="n">counting</span><span class="w"> </span><span class="n">k</span><span class="o">-</span><span class="n">mers</span><span class="w"> </span><span class="p">...</span><span class="w"> </span><span class="nl">tcmalloc</span><span class="p">:</span><span class="w"> </span><span class="k">large</span><span class="w"> </span><span class="n">alloc</span><span class="w"> </span><span class="mi">1610612736</span><span class="w"> </span><span class="n">bytes</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="mh">0x6d5f0000</span><span class="w"> </span><span class="err">@</span><span class="w">  </span><span class="mh">0x7f19fcbbb1e7</span><span class="w"> </span><span class="mh">0x6f46bd</span><span class="w"> </span><span class="mh">0x6f4739</span><span class="w"> </span><span class="mh">0x4af899</span><span class="w"> </span><span class="mh">0x4a7a68</span><span class="w"> </span><span class="mh">0x4aec09</span><span class="w"> </span><span class="mh">0x44e175</span><span class="w"> </span><span class="mh">0x7f19fbbd7bf7</span><span class="w"> </span><span class="mh">0x452f19</span><span class="w"></span>
<span class="n">done</span><span class="p">.</span><span class="w"></span>
<span class="o">[</span><span class="n">build</span><span class="o">]</span><span class="w"> </span><span class="n">building</span><span class="w"> </span><span class="n">target</span><span class="w"> </span><span class="n">de</span><span class="w"> </span><span class="n">Bruijn</span><span class="w"> </span><span class="n">graph</span><span class="w"> </span><span class="p">...</span><span class="w">  </span><span class="n">done</span><span class="w"> </span>
<span class="o">[</span><span class="n">build</span><span class="o">]</span><span class="w"> </span><span class="n">creating</span><span class="w"> </span><span class="n">equivalence</span><span class="w"> </span><span class="n">classes</span><span class="w"> </span><span class="p">...</span><span class="w">  </span><span class="n">done</span><span class="w"></span>
<span class="o">[</span><span class="n">build</span><span class="o">]</span><span class="w"> </span><span class="n">target</span><span class="w"> </span><span class="n">de</span><span class="w"> </span><span class="n">Bruijn</span><span class="w"> </span><span class="n">graph</span><span class="w"> </span><span class="n">has</span><span class="w"> </span><span class="mi">489912</span><span class="w"> </span><span class="n">contigs</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="k">contains</span><span class="w"> </span><span class="mi">58668462</span><span class="w"> </span><span class="n">k</span><span class="o">-</span><span class="n">mers</span><span class="w"></span>
</code></pre></div>

<p>Run kallisto for one set of samples</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Create BUS files from fastq&#39;s, can&#39;t do separate lines
!mkfifo R1.gz R2.gz R1_02.gz R2_02.gz R1_03.gz R2_03.gz R1_04.gz R2_04.gz; curl -Ls $s1 &gt; R1.gz &amp; curl -Ls $s2 &gt; R2.gz &amp; curl -Ls $s3 &gt; R1_02.gz &amp; curl -Ls $s4 &gt; R2_02.gz &amp; curl -Ls $s5 &gt; R1_03.gz &amp; curl -Ls $s6 &gt; R2_03.gz &amp; curl -Ls $s7 &gt; R1_04.gz &amp; curl -Ls $s8 &gt; R2_04.gz &amp; kallisto bus -i cufflinksTrans_clytia.idx -o bus_output/ -x 10xv2 -t 2 R1.gz R2.gz R1_02.gz R2_02.gz R1_03.gz R2_03.gz R1_04.gz R2_04.gz 
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Generate gene-count matrices
!wget --quiet https://github.com/bustools/getting_started/releases/download/getting_started/10xv2_whitelist.txt

#Make t2g file
!mv D1.1830 Marimba_merged_transcript_models.gff3
!awk &#39;$3==&quot;transcript&quot;&#39; Marimba_merged_transcript_models.gff3| awk &#39;{ print $9}&#39; &gt; t2g_unsplit.txt
!sed &#39;s/[ID=Parent]//g&#39; t2g_unsplit.txt &gt; t2g_clean.txt
!tr &#39;;&#39; &#39;\t&#39; &lt; t2g_clean.txt  &gt; t2g.txt
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#!cd bus_output/
!mkdir bus_output/genecount/ bus_output/tmp/

!bustools correct -w 10xv2_whitelist.txt -p bus_output/output.bus | bustools sort -T bus_output/tmptmp/ -t 2 -p - | bustools count -o bus_output/genecount/genes -g t2g.txt -e bus_output/matrix.ec -t bus_output/transcripts.txt --genecounts -
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># !ls test
</code></pre></div>
</td></tr></table>
<p>Run kallisto for other sample set</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Create BUS files from fastq&#39;s
!mkfifo R1_new.gz R2_new.gz R1_02_new.gz R2_02_new.gz R1_03_new.gz R2_03_new.gz R1_04_new.gz R2_04_new.gz; curl -Ls $s9 &gt; R1_new.gz &amp; curl -Ls $s10 &gt; R2_new.gz &amp; curl -Ls $s11 &gt; R1_02_new.gz &amp; curl -Ls $s12 &gt; R2_02_new.gz &amp; curl -Ls $s13 &gt; R1_03_new.gz &amp; curl -Ls $s14 &gt; R2_03_new.gz &amp; curl -Ls $s15 &gt; R1_04_new.gz &amp; curl -Ls $s16 &gt; R2_04_new.gz &amp; kallisto bus -i cufflinksTrans_clytia.idx -o bus_output_02/ -x 10xv2 -t 2 R1_new.gz R2_new.gz R1_02_new.gz R2_02_new.gz R1_03_new.gz R2_03_new.gz R1_04_new.gz R2_04_new.gz 
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Generate gene-count matrices
!cd bus_output_02/
!mkdir bus_output_02/genecount/ bus_output_02/tmp/

!bustools correct -w 10xv2_whitelist.txt -p bus_output_02/output.bus | bustools sort -T bus_output_02/tmp/ -t 2 -p - | bustools count -o bus_output_02/genecount/genes -g t2g.txt -e bus_output_02/matrix.ec -t bus_output_02/transcripts.txt --genecounts -
</code></pre></div>
</td></tr></table>
<p>Merge matrices (Add -1 to first and -2 to second dataset)</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>path = &quot;bus_output/genecount/&quot;
jelly_adata_01 = sc.read(path+&#39;genes.mtx&#39;, cache=True)
jelly_adata_01.var_names = pd.read_csv(path+&#39;genes.genes.txt&#39;, header=None)[0]
jelly_adata_01.obs_names = pd.read_csv(path+&#39;genes.barcodes.txt&#39;, header=None)[0]

jelly_adata_01.obs_names = [i+&quot;-1&quot; for i in jelly_adata_01.obs_names]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>path = &quot;bus_output_02/genecount/&quot;
jelly_adata_02 = sc.read(path+&#39;genes.mtx&#39;, cache=True)
jelly_adata_02.var_names = pd.read_csv(path+&#39;genes.genes.txt&#39;, header=None)[0]
jelly_adata_02.obs_names = pd.read_csv(path+&#39;genes.barcodes.txt&#39;, header=None)[0]

jelly_adata_02.obs_names = [i+&quot;-2&quot; for i in jelly_adata_02.obs_names]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>jelly_adata = jelly_adata_01.concatenate(jelly_adata_02,join=&#39;outer&#39;, index_unique=None)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>jelly_adata.write(&#39;cuff_adata_unfilt.h5ad&#39;)
</code></pre></div>
</td></tr></table>
<h3 id="read-in-saved-dataset"><strong>Read in saved dataset</strong><a class="headerlink" href="#read-in-saved-dataset" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#bus_fs_combo
bus_fs_trin = anndata.read(&quot;D1.1796&quot;)
bus_fs_raw_trin = anndata.read(&quot;D1.1797&quot;)

#Read in jelly_adata saved
jelly_adata = anndata.read(&quot;D1.1832&quot;)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>jelly_adata_filt = jelly_adata[list(bus_fs_trin.obs_names),:]
jelly_adata_filt
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">View</span> <span class="n">of</span> <span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">13673</span> <span class="err">×</span> <span class="mi">45872</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>jelly_adata_filt.obs[&#39;orgID&#39;] = pd.Categorical(bus_fs_trin.obs[&#39;orgID&#39;])
jelly_adata_filt.obs[&#39;fed&#39;] = pd.Categorical(bus_fs_trin.obs[&#39;fed&#39;])
jelly_adata_filt.obs[&#39;starved&#39;] = pd.Categorical(bus_fs_trin.obs[&#39;starved&#39;])
jelly_adata_filt.obs[&#39;cellRanger_louvain&#39;] = pd.Categorical(bus_fs_trin.obs[&#39;cellRanger_louvain&#39;])
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">Trying to set attribute `.obs` of view, copying.</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.filter_cells(jelly_adata_filt, min_counts=0) #1
sc.pp.filter_genes(jelly_adata_filt, min_counts=0)
jelly_adata_filt.obs[&#39;n_countslog&#39;]=np.log10(jelly_adata_filt.obs[&#39;n_counts&#39;])

jelly_adata_filt.raw = sc.pp.log1p(jelly_adata_filt, copy=True)
sc.pp.normalize_per_cell(jelly_adata_filt, counts_per_cell_after=1e4)
filter_result = sc.pp.filter_genes_dispersion(
    jelly_adata_filt.X, min_mean=0.0125, max_mean=4.5, min_disp=0.2)
sc.pl.filter_genes_dispersion(filter_result)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>jelly_adata_filt = jelly_adata_filt[:, filter_result.gene_subset]
sc.pp.scale(jelly_adata_filt, max_value=10)
sc.tl.pca(jelly_adata_filt, n_comps=60)
jelly_adata_filt
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">scanpy</span><span class="o">/</span><span class="n">preprocessing</span><span class="o">/</span><span class="n">_simple</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">806</span><span class="err">:</span><span class="w"> </span><span class="nl">UserWarning</span><span class="p">:</span><span class="w"> </span><span class="n">Revieved</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="k">view</span><span class="w"> </span><span class="k">of</span><span class="w"> </span><span class="n">an</span><span class="w"> </span><span class="n">AnnData</span><span class="p">.</span><span class="w"> </span><span class="n">Making</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">copy</span><span class="p">.</span><span class="w"></span>
<span class="w">  </span><span class="n">view_to_actual</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span><span class="w"></span>





<span class="n">AnnData</span><span class="w"> </span><span class="k">object</span><span class="w"> </span><span class="k">with</span><span class="w"> </span><span class="n">n_obs</span><span class="w"> </span><span class="err">×</span><span class="w"> </span><span class="n">n_vars</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">13673</span><span class="w"> </span><span class="err">×</span><span class="w"> </span><span class="mi">9609</span><span class="w"></span>
<span class="w">    </span><span class="nl">obs</span><span class="p">:</span><span class="w"> </span><span class="s1">&#39;batch&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;orgID&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;fed&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;starved&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;n_counts&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;n_countslog&#39;</span><span class="w"></span>
<span class="w">    </span><span class="nf">var</span><span class="err">:</span><span class="w"> </span><span class="s1">&#39;n_counts&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;mean&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;std&#39;</span><span class="w"></span>
<span class="w">    </span><span class="nl">uns</span><span class="p">:</span><span class="w"> </span><span class="s1">&#39;pca&#39;</span><span class="w"></span>
<span class="w">    </span><span class="nl">obsm</span><span class="p">:</span><span class="w"> </span><span class="s1">&#39;X_pca&#39;</span><span class="w"></span>
<span class="w">    </span><span class="nl">varm</span><span class="p">:</span><span class="w"> </span><span class="s1">&#39;PCs&#39;</span><span class="w"></span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.neighbors(jelly_adata_filt,n_neighbors=150, n_pcs=60,random_state=42) #use_rep=&#39;X_nca&#39;
sc.tl.paga(jelly_adata_filt, groups=&#39;cellRanger_louvain&#39;,)
sc.pl.paga(jelly_adata_filt, color=[&#39;cellRanger_louvain&#39;])

sc.tl.umap(jelly_adata_filt,random_state=42,spread=2.5,min_dist = 0.8,init_pos=&#39;paga&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../MARIMBAAnnosAnalysis_files/MARIMBAAnnosAnalysis_33_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.umap(jelly_adata_filt,random_state=42,spread=3.5,min_dist = 1.5,init_pos=&#39;paga&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">/usr/local/lib/python3.6/dist-packages/umap/umap_.py:1158: RuntimeWarning: divide by zero encountered in power</span>
<span class="err">  return 1.0 / (1.0 + a * x ** (2 * b))</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.umap(jelly_adata_filt,color=[&#39;cellRanger_louvain&#39;])
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../MARIMBAAnnosAnalysis_files/MARIMBAAnnosAnalysis_35_0.png" /></p>
<h4 id="plot-overlap-in-cell-types"><strong>Plot Overlap in Cell Types</strong><a class="headerlink" href="#plot-overlap-in-cell-types" title="Permanent link">&para;</a></h4>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Read in filtered and clustered MARIMBA data

jelly_adata_filt = anndata.read(&quot;D1.1833&quot;)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.umap(jelly_adata_filt,color=[&#39;cellRanger_louvain&#39;])
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../MARIMBAAnnosAnalysis_files/MARIMBAAnnosAnalysis_38_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Check that embedding doesn&#39;t work with random cluster labels
import random
new_labels = list(jelly_adata_filt.obs[&#39;cellRanger_louvain&#39;])
random.shuffle(new_labels)

jelly_adata_filt.obs[&#39;shuffled&#39;] = pd.Categorical(new_labels)

sc.tl.paga(jelly_adata_filt, groups=&#39;shuffled&#39;,)
sc.pl.paga(jelly_adata_filt, color=[&#39;shuffled&#39;])

sc.tl.umap(jelly_adata_filt,random_state=42,spread=3.5,min_dist = 1.5,init_pos=&#39;paga&#39;)

sc.pl.umap(jelly_adata_filt,color=[&#39;cellRanger_louvain&#39;,&#39;shuffled&#39;])
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../MARIMBAAnnosAnalysis_files/MARIMBAAnnosAnalysis_39_0.png" /></p>
<div class="codehilite"><pre><span></span><code><span class="err">/usr/local/lib/python3.6/dist-packages/umap/umap_.py:1158: RuntimeWarning: divide by zero encountered in power</span>
<span class="err">  return 1.0 / (1.0 + a * x ** (2 * b))</span>
</code></pre></div>

<p><img alt="png" src="../MARIMBAAnnosAnalysis_files/MARIMBAAnnosAnalysis_39_2.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># len(set(jelly_adata_filt.var_names).intersection(bus_fs_trin.var_names))/len(jelly_adata_filt.var_names)
</code></pre></div>
</td></tr></table>
<h5 id="comparing-gene-markers"><strong>Comparing Gene Markers</strong><a class="headerlink" href="#comparing-gene-markers" title="Permanent link">&para;</a></h5>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>n=100
bus_fs_trin.obs[&#39;cellRanger_louvain&#39;] = pd.Categorical(bus_fs_trin.obs[&#39;cellRanger_louvain&#39;])
sc.tl.rank_genes_groups(jelly_adata_filt,groupby=&#39;cellRanger_louvain&#39;,n_genes=n,method=&#39;wilcoxon&#39;)
sc.tl.rank_genes_groups(bus_fs_trin,groupby=&#39;cellRanger_louvain&#39;,n_genes=n,method=&#39;wilcoxon&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="o">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">statsmodels</span><span class="o">/</span><span class="n">tools</span><span class="o">/</span><span class="n">_testing</span><span class="o">.</span><span class="n">py</span><span class="p">:</span><span class="mi">19</span><span class="p">:</span> <span class="ne">FutureWarning</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">util</span><span class="o">.</span><span class="n">testing</span> <span class="ow">is</span> <span class="n">deprecated</span><span class="o">.</span> <span class="n">Use</span> <span class="n">the</span> <span class="n">functions</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">public</span> <span class="n">API</span> <span class="n">at</span> <span class="n">pandas</span><span class="o">.</span><span class="n">testing</span> <span class="n">instead</span><span class="o">.</span>
  <span class="kn">import</span> <span class="nn">pandas.util.testing</span> <span class="kn">as</span> <span class="nn">tm</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Show pairwise overlap in top 100 names between all clusters, 36x36 grid
clus = np.unique(bus_fs_trin.obs[&#39;cellRanger_louvain&#39;])

cuff = [[]]*len(clus) #np array of top 100 genes for each of 36 clusters
busFS = [[]]*len(clus)#np array of top 100 genes for each of 36 clusters

for c in clus:
  cuff[c] =  list(jelly_adata_filt.uns[&#39;rank_genes_groups&#39;][&#39;names&#39;][str(c)])
  busFS[c] = list(bus_fs_trin.uns[&#39;rank_genes_groups&#39;][&#39;names&#39;][str(c)])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#https://stackoverflow.com/questions/52408910/python-pairwise-intersection-of-multiple-lists-then-sum-up-all-duplicates
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>from itertools import combinations_with_replacement

#Changed to calculate Jaccard Index
def intersect(i,j):
  return len(set(cuff[i]).intersection(busFS[j]))/len(set(cuff[i]).union(busFS[j]))

def pairwise(clus):        
  # Initialise precomputed matrix (num of clusters, 36x36)
  precomputed = np.zeros((len(clus),len(clus)), dtype=&#39;float&#39;)
  # Initialise iterator over objects in X
  iterator = combinations_with_replacement(range(0,len(clus)), 2)
  # Perform the operation on each pair
  for i,j in iterator:
    precomputed[i,j] = intersect(i,j)          
  # Make symmetric and return
  return precomputed + precomputed.T - np.diag(np.diag(precomputed))
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>pairCorrs = pairwise(clus)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>plt.figure(figsize=(7,7))
plt.imshow(pairCorrs, cmap=&#39;viridis&#39;)
plt.colorbar()


plt.xticks(np.arange(0, 36, 1),fontsize=5)
plt.yticks(np.arange(0, 36, 1),fontsize=5)
plt.grid(color=&#39;black&#39;,linewidth=0.3)
plt.xlabel(&#39;Trinity Clusters&#39;)
plt.ylabel(&#39;MARIMBA Clusters&#39;)

plt.show()
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../MARIMBAAnnosAnalysis_files/MARIMBAAnnosAnalysis_47_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#jelly_adata_filt.write(&#39;cufflinks_clus_UMAP_PAGA.h5ad&#39;)
</code></pre></div>
</td></tr></table>
<h4 id="neuron-analysis"><strong>Neuron Analysis</strong><a class="headerlink" href="#neuron-analysis" title="Permanent link">&para;</a></h4>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Read in previously saved neurons (TRINITY) and filtered/clustered cufflinks data
neurons = anndata.read(&#39;D1.1804&#39;)

#Read in jelly_adata saved 

jelly_adata_filt = anndata.read(&#39;D1.1833&#39;)


jelly_adata_raw = anndata.read(&quot;D1.1832&quot;)
jelly_adata_raw = jelly_adata_raw[list(jelly_adata_filt.obs_names),:]
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>jelly_adata_raw.obs[&#39;orgID&#39;] = pd.Categorical(jelly_adata_filt.obs[&#39;orgID&#39;])
jelly_adata_raw.obs[&#39;fed&#39;] = pd.Categorical(jelly_adata_filt.obs[&#39;fed&#39;])
jelly_adata_raw.obs[&#39;starved&#39;] = pd.Categorical(jelly_adata_filt.obs[&#39;starved&#39;])
jelly_adata_raw.obs[&#39;cellRanger_louvain&#39;] = pd.Categorical(jelly_adata_filt.obs[&#39;cellRanger_louvain&#39;])
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">Trying to set attribute `.obs` of view, copying.</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>cuff_neurons = jelly_adata_raw[jelly_adata_raw.obs[&#39;cellRanger_louvain&#39;].isin([31,26,6,9])]
sc.pp.filter_cells(cuff_neurons, min_counts=0)
sc.pp.filter_genes(cuff_neurons, min_counts=0)

sc.pp.normalize_per_cell(cuff_neurons, counts_per_cell_after=1e4)

sc.pp.log1p(cuff_neurons)

# sc.pp.scale(raw_overlap_combo, max_value=10)
sc.pp.highly_variable_genes(cuff_neurons,n_top_genes=2000,n_bins=50)

#sc.pp.filter_cells(neurons, min_counts=0)
# sc.pp.filter_genes(neurons, min_counts=0)

# sc.pp.highly_variable_genes(neurons,n_top_genes=3000)
cuff_neurons = cuff_neurons[:,cuff_neurons.var[&#39;highly_variable&#39;]]

sc.pp.scale(cuff_neurons, max_value=10)

sc.tl.pca(cuff_neurons, n_comps=60)
sc.pl.pca_variance_ratio(cuff_neurons, log=True)

#applyNCAEmbed(neurons,neurons.obs[&#39;knn_clus&#39;])

sc.pp.neighbors(cuff_neurons,n_neighbors=15, n_pcs=15) #n_neighbors=5, n_pcs=15,use_rep=&#39;X_nca&#39;
sc.tl.louvain(cuff_neurons,resolution=1,key_added=&#39;louvain_neur&#39;,random_state=42)#Clustering algorithm,resolution=0.5
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="n">Trying</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">set</span><span class="w"> </span><span class="n">attribute</span><span class="w"> </span><span class="err">`</span><span class="p">.</span><span class="n">obs</span><span class="err">`</span><span class="w"> </span><span class="k">of</span><span class="w"> </span><span class="k">view</span><span class="p">,</span><span class="w"> </span><span class="n">copying</span><span class="p">.</span><span class="w"></span>
<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">scanpy</span><span class="o">/</span><span class="n">preprocessing</span><span class="o">/</span><span class="n">_simple</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">806</span><span class="err">:</span><span class="w"> </span><span class="nl">UserWarning</span><span class="p">:</span><span class="w"> </span><span class="n">Revieved</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="k">view</span><span class="w"> </span><span class="k">of</span><span class="w"> </span><span class="n">an</span><span class="w"> </span><span class="n">AnnData</span><span class="p">.</span><span class="w"> </span><span class="n">Making</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">copy</span><span class="p">.</span><span class="w"></span>
<span class="w">  </span><span class="n">view_to_actual</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span><span class="w"></span>
</code></pre></div>

<p><img alt="png" src="../MARIMBAAnnosAnalysis_files/MARIMBAAnnosAnalysis_52_1.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.louvain(cuff_neurons,resolution=0.95,key_added=&#39;louvain_neur&#39;,random_state=42)#Clustering algorithm,resolution=0.5
#sc.tl.tsne(neurons, n_pcs=15,perplexity=25,random_state = 42) #learning_rate=250
sc.tl.umap(cuff_neurons,random_state=42,spread=2.5, min_dist=1)

cuff_neurons.obs[&#39;cellAtlasClusters&#39;] = pd.Categorical(cuff_neurons.obs[&#39;cellRanger_louvain&#39;] )
cuff_neurons = cuff_neurons[neurons.obs_names,:]
cuff_neurons.obs[&#39;louvain_neur_TRINITY&#39;] = neurons.obs[&#39;louvain_neur&#39;]

sc.pl.umap(cuff_neurons, color=[&#39;louvain_neur&#39;,&#39;louvain_neur_TRINITY&#39;,&#39;cellAtlasClusters&#39;],color_map=&#39;viridis&#39;,size=50,legend_loc=&#39;on data&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="n">Trying</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">set</span><span class="w"> </span><span class="n">attribute</span><span class="w"> </span><span class="err">`</span><span class="p">.</span><span class="n">obs</span><span class="err">`</span><span class="w"> </span><span class="k">of</span><span class="w"> </span><span class="k">view</span><span class="p">,</span><span class="w"> </span><span class="n">copying</span><span class="p">.</span><span class="w"></span>
</code></pre></div>

<p><img alt="png" src="../MARIMBAAnnosAnalysis_files/MARIMBAAnnosAnalysis_53_1.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.dendrogram(cuff_neurons,&#39;louvain_neur&#39;,linkage_method=&#39;ward&#39;)
cuff_neurons.uns[&#39;dendrogram_louvain_neur&#39;] = cuff_neurons.uns[&quot;dendrogram_[&#39;louvain_neur&#39;]&quot;]
sc.tl.dendrogram(cuff_neurons,&#39;louvain_neur_TRINITY&#39;,linkage_method=&#39;ward&#39;)
cuff_neurons.uns[&#39;dendrogram_louvain_neur_TRINITY&#39;] = cuff_neurons.uns[&quot;dendrogram_[&#39;louvain_neur_TRINITY&#39;]&quot;]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#XLOC_042761 instead of TRINITY_DN20104_c0_g1_i1.mrna1
topGenes = [&#39;XLOC_030920&#39;,&#39;XLOC_035224&#39;,
            &#39;XLOC_040584&#39;,&#39;XLOC_019434&#39;,&#39;XLOC_042761&#39;,&#39;XLOC_017097&#39;,
            &#39;XLOC_021799&#39;,&#39;XLOC_041402&#39;,&#39;XLOC_004021&#39;,&#39;XLOC_008730&#39;,
            &#39;XLOC_014624&#39;,&#39;XLOC_030120&#39;]
forPlotAdata = jelly_adata_raw[jelly_adata_raw.obs[&#39;cellRanger_louvain&#39;].isin([31,26,6,9])]
sc.pp.filter_cells(forPlotAdata, min_counts=0)
sc.pp.filter_genes(forPlotAdata, min_counts=0)

sc.pp.log1p(forPlotAdata)
forPlotAdata = forPlotAdata[cuff_neurons.obs_names,:]
forPlotAdata.obs[&#39;louvain_neur&#39;] = cuff_neurons.obs[&#39;louvain_neur&#39;]
forPlotAdata.obs[&#39;louvain_neur_TRINITY&#39;] = cuff_neurons.obs[&#39;louvain_neur_TRINITY&#39;]

forPlotAdata.uns[&#39;dendrogram_louvain_neur&#39;]  = cuff_neurons.uns[&#39;dendrogram_louvain_neur&#39;] 
forPlotAdata.uns[&#39;dendrogram_louvain_neur_TRINITY&#39;]  = cuff_neurons.uns[&#39;dendrogram_louvain_neur_TRINITY&#39;] 

neurons_sigTop = forPlotAdata[:,topGenes]

#labels = [allGenes[i] for i in forPlot]
labels = [&#39;Ngn&#39;,&#39;PP01&#39;,
          &#39;*GRWGamide&#39;,&#39;*RFamide&#39;,&#39;*PRPamide&#39;,&#39;*GLWamide2&#39;,
          &#39;PP02&#39;,&#39;*YFamide&#39;,&#39;*FLFamide&#39;,&#39;PP03&#39;,
          &#39;PP04&#39;,&#39;PP17&#39;]
neurons_sigTop.var[&#39;names&#39;] = labels

sc.set_figure_params(scanpy=True, fontsize=10)
sc.pl.heatmap(neurons_sigTop, labels, groupby=&#39;louvain_neur&#39;, dendrogram=True,show_gene_labels=True,
              gene_symbols = &#39;names&#39;,swap_axes=False,
              cmap=&#39;PuBuGn&#39;,standard_scale=&#39;var&#39;,save=&#39;neuronMarkers.pdf&#39;) 
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="n">Trying</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">set</span><span class="w"> </span><span class="n">attribute</span><span class="w"> </span><span class="err">`</span><span class="p">.</span><span class="n">obs</span><span class="err">`</span><span class="w"> </span><span class="k">of</span><span class="w"> </span><span class="k">view</span><span class="p">,</span><span class="w"> </span><span class="n">copying</span><span class="p">.</span><span class="w"></span>
<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="n">Trying</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">set</span><span class="w"> </span><span class="n">attribute</span><span class="w"> </span><span class="err">`</span><span class="p">.</span><span class="n">obs</span><span class="err">`</span><span class="w"> </span><span class="k">of</span><span class="w"> </span><span class="k">view</span><span class="p">,</span><span class="w"> </span><span class="n">copying</span><span class="p">.</span><span class="w"></span>
<span class="n">Trying</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">set</span><span class="w"> </span><span class="n">attribute</span><span class="w"> </span><span class="err">`</span><span class="p">.</span><span class="nf">var</span><span class="err">`</span><span class="w"> </span><span class="k">of</span><span class="w"> </span><span class="k">view</span><span class="p">,</span><span class="w"> </span><span class="n">copying</span><span class="p">.</span><span class="w"></span>
<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1192</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="n">is_string_dtype</span><span class="p">(</span><span class="n">df</span><span class="o">[</span><span class="n">key</span><span class="o">]</span><span class="p">)</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df</span><span class="o">[</span><span class="n">key</span><span class="o">]</span><span class="p">)</span><span class="w"></span>


<span class="nl">WARNING</span><span class="p">:</span><span class="w"> </span><span class="n">saving</span><span class="w"> </span><span class="n">figure</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="n">figures</span><span class="o">/</span><span class="n">heatmapneuronMarkers</span><span class="p">.</span><span class="n">pdf</span><span class="w"></span>
</code></pre></div>

<p><img alt="png" src="../MARIMBAAnnosAnalysis_files/MARIMBAAnnosAnalysis_55_2.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.heatmap(neurons_sigTop, labels, groupby=&#39;louvain_neur_TRINITY&#39;, dendrogram=True,show_gene_labels=True,
              gene_symbols = &#39;names&#39;,swap_axes=False,
              cmap=&#39;PuBuGn&#39;,standard_scale=&#39;var&#39;,save=&#39;neuronMarkers.pdf&#39;) 
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="nl">WARNING</span><span class="p">:</span><span class="w"> </span><span class="n">saving</span><span class="w"> </span><span class="n">figure</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">file</span><span class="w"> </span><span class="n">figures</span><span class="o">/</span><span class="n">heatmapneuronMarkers</span><span class="p">.</span><span class="n">pdf</span><span class="w"></span>


<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1192</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="n">is_string_dtype</span><span class="p">(</span><span class="n">df</span><span class="o">[</span><span class="n">key</span><span class="o">]</span><span class="p">)</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df</span><span class="o">[</span><span class="n">key</span><span class="o">]</span><span class="p">)</span><span class="w"></span>
<span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
</code></pre></div>

<p><img alt="png" src="../MARIMBAAnnosAnalysis_files/MARIMBAAnnosAnalysis_56_2.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Show pairwise overlap in top 100 names between all clusters, 36x36 grid
clus = np.unique(cuff_neurons.obs[&#39;louvain_neur&#39;])

cuff = [[]]*len(clus) #np array of top 100 genes for each of 36 clusters
busFS = [[]]*len(clus)#np array of top 100 genes for each of 36 clusters

for c in clus:
  cuff[int(c)] =  list(cuff_neurons[cuff_neurons.obs[&#39;louvain_neur&#39;].isin([c])].obs_names)
  busFS[int(c)] = list(cuff_neurons[cuff_neurons.obs[&#39;louvain_neur_TRINITY&#39;].isin([c])].obs_names)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>pairCorrs = pairwise(clus)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>plt.figure(figsize=(7,7))
plt.imshow(pairCorrs, cmap=&#39;viridis&#39;)
plt.colorbar()


plt.xticks(np.arange(0, len(clus), 1),fontsize=5)
plt.yticks(np.arange(0, len(clus), 1),fontsize=5)
plt.grid(color=&#39;black&#39;,linewidth=0.3)
plt.xlabel(&#39;TRINITY Clusters&#39;)
plt.ylabel(&#39;Cufflinks Clusters&#39;)

plt.show()
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../MARIMBAAnnosAnalysis_files/MARIMBAAnnosAnalysis_59_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">True</span>
</code></pre></div>

<h5 id="check-if-application-of-trinity-louvain-clusters-is-appropriate"><strong>Check if application of TRINITY Louvain clusters is appropriate</strong><a class="headerlink" href="#check-if-application-of-trinity-louvain-clusters-is-appropriate" title="Permanent link">&para;</a></h5>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>cuff_neurons = jelly_adata_filt[jelly_adata_filt.obs[&#39;cellRanger_louvain&#39;].isin([31,26,6,9])]
cuff_neurons = cuff_neurons[neurons.obs_names,:]
cuff_neurons.obs[&#39;louvain_neur&#39;] = neurons.obs[&#39;louvain_neur&#39;]
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">anndata</span><span class="o">/</span><span class="n">_core</span><span class="o">/</span><span class="n">anndata</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">1094</span><span class="err">:</span><span class="w"> </span><span class="nl">FutureWarning</span><span class="p">:</span><span class="w"> </span><span class="n">is_categorical</span><span class="w"> </span><span class="k">is</span><span class="w"> </span><span class="n">deprecated</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">will</span><span class="w"> </span><span class="n">be</span><span class="w"> </span><span class="n">removed</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="n">future</span><span class="w"> </span><span class="n">version</span><span class="p">.</span><span class="w">  </span><span class="k">Use</span><span class="w"> </span><span class="n">is_categorical_dtype</span><span class="w"> </span><span class="n">instead</span><span class="w"></span>
<span class="w">  </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">is_categorical</span><span class="p">(</span><span class="n">df_full</span><span class="o">[</span><span class="n">k</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>
<span class="n">Trying</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">set</span><span class="w"> </span><span class="n">attribute</span><span class="w"> </span><span class="err">`</span><span class="p">.</span><span class="n">obs</span><span class="err">`</span><span class="w"> </span><span class="k">of</span><span class="w"> </span><span class="k">view</span><span class="p">,</span><span class="w"> </span><span class="n">copying</span><span class="p">.</span><span class="w"></span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>n=5
neurons.obs[&#39;louvain_neur&#39;] = pd.Categorical(neurons.obs[&#39;louvain_neur&#39;])
sc.tl.rank_genes_groups(cuff_neurons,groupby=&#39;louvain_neur&#39;,n_genes=n,method=&#39;wilcoxon&#39;)
sc.tl.rank_genes_groups(neurons,groupby=&#39;louvain_neur&#39;,n_genes=n,method=&#39;wilcoxon&#39;,)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="mf">.6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">scanpy</span><span class="o">/</span><span class="n">tools</span><span class="o">/</span><span class="n">_rank_genes_groups</span><span class="p">.</span><span class="nl">py</span><span class="p">:</span><span class="mi">408</span><span class="err">:</span><span class="w"> </span><span class="nl">RuntimeWarning</span><span class="p">:</span><span class="w"> </span><span class="n">invalid</span><span class="w"> </span><span class="k">value</span><span class="w"> </span><span class="n">encountered</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">log2</span><span class="w"></span>
<span class="w">  </span><span class="n">foldchanges</span><span class="o">[</span><span class="n">global_indices</span><span class="o">]</span><span class="w"></span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Show pairwise overlap in top 100 names between all clusters, 36x36 grid
clus = np.unique(neurons.obs[&#39;louvain_neur&#39;])

cuff = [[]]*len(clus) #np array of top 100 genes for each of 36 clusters
busFS = [[]]*len(clus)#np array of top 100 genes for each of 36 clusters

for c in clus:
  cuff[int(c)] =  list(cuff_neurons.uns[&#39;rank_genes_groups&#39;][&#39;names&#39;][str(c)])
  busFS[int(c)] = list(neurons.uns[&#39;rank_genes_groups&#39;][&#39;names&#39;][str(c)])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>pairCorrs = pairwise(clus)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>plt.figure(figsize=(7,7))
plt.imshow(pairCorrs, cmap=&#39;viridis&#39;)
plt.colorbar()


plt.xticks(np.arange(0, len(clus), 1),fontsize=5)
plt.yticks(np.arange(0, len(clus), 1),fontsize=5)
plt.grid(color=&#39;black&#39;,linewidth=0.3)
plt.xlabel(&#39;TRINITY Clusters&#39;)
plt.ylabel(&#39;Cufflinks Clusters&#39;)

plt.show()
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../MARIMBAAnnosAnalysis_files/MARIMBAAnnosAnalysis_66_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>neurons.uns[&#39;rank_genes_groups&#39;][&#39;names&#39;][&#39;0&#39;]
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">array([&#39;XLOC_011074&#39;, &#39;XLOC_015355&#39;, &#39;XLOC_044603&#39;, &#39;XLOC_030920&#39;,</span>
<span class="err">       &#39;XLOC_004445&#39;], dtype=object)</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>cuff_neurons.uns[&#39;rank_genes_groups&#39;][&#39;names&#39;][&#39;0&#39;]
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">array([&#39;XLOC_030930&#39;, &#39;XLOC_029992&#39;, &#39;XLOC_010444&#39;, &#39;XLOC_003896&#39;,</span>
<span class="err">       &#39;XLOC_006929&#39;], dtype=object)</span>
</code></pre></div>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../allDistanceCalculations/" class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Comparative Analysis of Cell Atlas Concordance and Perturbation Responses
              </div>
            </div>
          </a>
        
        
          <a href="../MARIMBAAtlasSearchAndPlot/" class="md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Explore MARIMBA-Annotated Cell Atlas
              </div>
            </div>
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/vendor.53cc9318.min.js"></script>
      <script src="../../assets/javascripts/bundle.e9c9f54f.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "../..",
          features: ['search.suggest', 'search.highlight'],
          search: Object.assign({
            worker: "../../assets/javascripts/worker/search.9c0e82ba.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>