
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Walkthrough for Whole Organisms Mutliplexed Data Analysis">
      
      
      
        <meta name="author" content="Tara Chari">
      
      
        <link rel="canonical" href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/notebooks/cellRangerClustering_Starvation/">
      
      <link rel="shortcut icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.2.7">
    
    
      
        <title>Initial Cell Atlas Clustering from Cell Ranger - WHAMSeqAnalysis</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.cb6bc1d0.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.39b8e14a.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#download-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/" title="WHAMSeqAnalysis" class="md-header-nav__button md-logo" aria-label="WHAMSeqAnalysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      <div class="md-header-nav__ellipsis">
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            WHAMSeqAnalysis
          </span>
        </div>
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            
              Initial Cell Atlas Clustering from Cell Ranger
            
          </span>
        </div>
      </div>
    </div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    




<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/" title="WHAMSeqAnalysis" class="md-nav__button md-logo" aria-label="WHAMSeqAnalysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    WHAMSeqAnalysis
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" >
      
      <label class="md-nav__link" for="nav-2">
        ClickTag Pre-processing
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="ClickTag Pre-processing" data-md-level="1">
        <label class="md-nav__title" for="nav-2">
          <span class="md-nav__icon md-icon"></span>
          ClickTag Pre-processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellRangerClickTagCounts/" class="md-nav__link">
        Initial Quantification of ClickTag Counts with Cell Ranger (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistobusStarvClickTags/" class="md-nav__link">
        Quantification of ClickTag Counts with kallisto-bustools (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistobusStimClickTags/" class="md-nav__link">
        Quantification of ClickTag Counts with kallisto-bustools (Stimulation Data)
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3" >
      
      <label class="md-nav__link" for="nav-3">
        cDNA Pre-processing
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="cDNA Pre-processing" data-md-level="1">
        <label class="md-nav__title" for="nav-3">
          <span class="md-nav__icon md-icon"></span>
          cDNA Pre-processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../filterStarvCells_ClickTags/" class="md-nav__link">
        Initial cDNA Filtering with Cell Ranger Data (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistoBusRuns_StarvAndStim/" class="md-nav__link">
        Quantification of cDNA Counts with kallisto-bustools (All Data)
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4" checked>
      
      <label class="md-nav__link" for="nav-4">
        Cell Atlas and Perturbation Response
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Cell Atlas and Perturbation Response" data-md-level="1">
        <label class="md-nav__title" for="nav-4">
          <span class="md-nav__icon md-icon"></span>
          Cell Atlas and Perturbation Response
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Initial Cell Atlas Clustering from Cell Ranger
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Initial Cell Atlas Clustering from Cell Ranger
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import Packages
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-in-original-cell-ranger-data-for-clustering-analysis" class="md-nav__link">
    Read in original Cell Ranger Data for clustering analysis
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../starvation_Analysis/" class="md-nav__link">
        Cell Atlas Clustering and Analysis from kallisto-bustools
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../neuronSubpop_Analysis/" class="md-nav__link">
        Analysis of Neuron Subpopulations
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../pseudotime_Analysis/" class="md-nav__link">
        Pseudotime Analysis
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../deSeq2Analysis_StarvationResponse/" class="md-nav__link">
        Extraction of Differentially Expressed Genes Under Starvation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5" >
      
      <label class="md-nav__link" for="nav-5">
        Stimulation Data Analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Stimulation Data Analysis" data-md-level="1">
        <label class="md-nav__title" for="nav-5">
          <span class="md-nav__icon md-icon"></span>
          Stimulation Data Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../stimulation_Analysis/" class="md-nav__link">
        Cell Atlas Concordance and Perturbation Response
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../deSeq2Analysis_StimulationResponse/" class="md-nav__link">
        Extraction of Differentially Expressed Genes Under Starvation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6" >
      
      <label class="md-nav__link" for="nav-6">
        Inter-Experimental Analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Inter-Experimental Analysis" data-md-level="1">
        <label class="md-nav__title" for="nav-6">
          <span class="md-nav__icon md-icon"></span>
          Inter-Experimental Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../allDistanceCalculations/" class="md-nav__link">
        Comparative Analysis of Cell Atlas Concordance and Perturbation Responses
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-7" type="checkbox" id="nav-7" >
      
      <label class="md-nav__link" for="nav-7">
        Explore Mutliplexed Datasets
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Explore Mutliplexed Datasets" data-md-level="1">
        <label class="md-nav__title" for="nav-7">
          <span class="md-nav__icon md-icon"></span>
          Explore Mutliplexed Datasets
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../MARIMBAAnnosAnalysis/" class="md-nav__link">
        Generation Cell Atlas with MARIMBA Transcriptome
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../MARIMBAAtlasSearchAndPlot/" class="md-nav__link">
        Explore MARIMBA-Annotated Cell Atlas
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellAtlasSearchAndPlot/" class="md-nav__link">
        Explore Original Cell Atlas
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import Packages
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-in-original-cell-ranger-data-for-clustering-analysis" class="md-nav__link">
    Read in original Cell Ranger Data for clustering analysis
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/edit/master/docs/notebooks/cellRangerClustering_Starvation.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                  <h1>Initial Cell Atlas Clustering from Cell Ranger</h1>
                
                <p><a href="https://colab.research.google.com/github/pachterlab/CWGFLHGCCHAP_2021/blob/master/notebooks/CellAtlasAnalysis/cellRangerClustering_Starvation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<h4 id="download-data"><strong>Download Data</strong><a class="headerlink" href="#download-data" title="Permanent link">&para;</a></h4>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import requests
from tqdm import tnrange, tqdm_notebook
def download_file(doi,ext):
    url = &#39;https://api.datacite.org/dois/&#39;+doi+&#39;/media&#39;
    r = requests.get(url).json()
    netcdf_url = r[&#39;data&#39;][0][&#39;attributes&#39;][&#39;url&#39;]
    r = requests.get(netcdf_url,stream=True)
    #Set file name
    fname = doi.split(&#39;/&#39;)[-1]+ext
    #Download file with progress bar
    if r.status_code == 403:
        print(&quot;File Unavailable&quot;)
    if &#39;content-length&#39; not in r.headers:
        print(&quot;Did not get file&quot;)
    else:
        with open(fname, &#39;wb&#39;) as f:
            total_length = int(r.headers.get(&#39;content-length&#39;))
            pbar = tnrange(int(total_length/1024), unit=&quot;B&quot;)
            for chunk in r.iter_content(chunk_size=1024):
                if chunk:
                    pbar.update()
                    f.write(chunk)
        return fname
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># Cell Barcodes from ClickTag filtering
download_file(&#39;10.22002/D1.1811&#39;,&#39;.gz&#39;)

# Condition/Organism labels per cell
download_file(&#39;10.22002/D1.1812&#39;,&#39;.gz&#39;)

#From CellRanger
#matrix
download_file(&#39;10.22002/D1.1802&#39;,&#39;.gz&#39;)

#features
download_file(&#39;10.22002/D1.1803&#39;,&#39;.gz&#39;)

#barcodes
download_file(&#39;10.22002/D1.1801&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">65</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">49202</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">111488</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">329</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">4453</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1801.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!gunzip *.gz
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!pip install --quiet anndata
!pip install --quiet scanpy

!pip install --quiet louvain
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">[K     |████████████████████████████████| 122kB 6.1MB/s </span>
<span class="err">[K     |████████████████████████████████| 7.7MB 3.1MB/s </span>
<span class="err">[K     |████████████████████████████████| 51kB 6.1MB/s </span>
<span class="err">[K     |████████████████████████████████| 71kB 7.5MB/s </span>
<span class="err">[?25h  Building wheel for sinfo (setup.py) ... [?25l[?25hdone</span>
<span class="err">[K     |████████████████████████████████| 2.4MB 4.4MB/s </span>
<span class="err">[K     |████████████████████████████████| 3.2MB 31.7MB/s </span>
<span class="err">[K     |████████████████████████████████| 2.2MB 4.2MB/s </span>
<span class="err">[?25h</span>
</code></pre></div>

<h4 id="import-packages"><strong>Import Packages</strong><a class="headerlink" href="#import-packages" title="Permanent link">&para;</a></h4>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import pandas as pd
import anndata
import scanpy as sc
import numpy as np
import scipy.sparse

import warnings
warnings.filterwarnings(&#39;ignore&#39;)

from sklearn.neighbors import (KNeighborsClassifier,NeighborhoodComponentsAnalysis)
from sklearn.pipeline import Pipeline
from sklearn.manifold import TSNE
from sklearn.decomposition import PCA

from sklearn.preprocessing import scale

import random

# import scrublet as scr

import matplotlib.pyplot as plt
%matplotlib inline
sc.set_figure_params(dpi=125)

import seaborn as sns
sns.set(style=&quot;whitegrid&quot;)
%load_ext rpy2.ipython
</code></pre></div>
</td></tr></table>
<h4 id="read-in-original-cell-ranger-data-for-clustering-analysis"><strong>Read in original Cell Ranger Data for clustering analysis</strong><a class="headerlink" href="#read-in-original-cell-ranger-data-for-clustering-analysis" title="Permanent link">&para;</a></h4>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>path = &quot;&quot; #/home/jgehring/scRNAseq/clytia/20181229/jelly3_aggr_20181231/outs/raw_feature_bc_matrix/
!mv D1.1802 D1.1802.mtx
!mv D1.1803 D1.1803.tsv
!mv D1.1801 D1.1801.tsv

jelly3trin = sc.read(path+&#39;D1.1802.mtx&#39;, cache=True).T
jelly3trin.var_names = pd.read_csv(path+&#39;D1.1803.tsv&#39;, header=None, sep=&#39;\t&#39;)[1]
jelly3trin.obs_names = pd.read_csv(path+&#39;D1.1801.tsv&#39;, header=None)[0]
jelly3trin
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">1474560</span> <span class="err">×</span> <span class="mi">46716</span>
</code></pre></div>

<p>Filter for highly variable genes and filter cells previously selected by ClickTag pre-processing</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.filter_cells(jelly3trin, min_counts=0)

sc.pp.filter_genes(jelly3trin, min_counts=0)

#The dataset has 1.4M cells and 46,716 genes before any filtering
jelly3trin

sc.pp.filter_cells(jelly3trin, min_counts=1)

sc.pp.filter_genes(jelly3trin, min_counts=1)

#Add n_countslog to the data
jelly3trin.obs[&#39;n_countslog&#39;]=np.log10(jelly3trin.obs[&#39;n_counts&#39;])

#Of the possible cells and genes, 9.9M cells and 22149 genes have at least one read
jelly3trin

#True cells were identified in the filterStarvCells notebook, so we&#39;ll load that scanpy file and use its cells

!mv D1.1811 D1.1811.csv
HDfilteredbarcodes=pd.read_csv(&quot;D1.1811.csv&quot;) #DOWNLOAD


len(list(HDfilteredbarcodes.iloc[:,0]))
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">16813</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Filter out low-quality cells as determined from tag sequencing data
ojelly3=jelly3trin[jelly3trin.obs_names.isin(list(HDfilteredbarcodes.iloc[:,0]))]
ojelly3
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">View</span> <span class="n">of</span> <span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">16813</span> <span class="err">×</span> <span class="mi">20937</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#sc.pl.violin(ojelly3, keys=[&#39;n_countslog&#39;])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.filter_genes(ojelly3,min_cells=3)
print(ojelly3)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Trying</span> <span class="n">to</span> <span class="n">set</span> <span class="n">attribute</span> <span class="err">`</span><span class="o">.</span><span class="k">var</span><span class="err">`</span> <span class="n">of</span> <span class="n">view</span><span class="p">,</span> <span class="n">copying</span><span class="o">.</span>


<span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">16813</span> <span class="err">×</span> <span class="mi">16586</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_cells&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sortedUMIs = ojelly3.obs[&#39;n_counts&#39;].sort_values(ascending=False)
#plt.plot(np.log10(range(len(sortedUMIs))), sortedUMIs.apply(np.log10))
#plt.show()

ojelly3.raw = sc.pp.log1p(ojelly3, copy=True)

sc.pp.normalize_per_cell(ojelly3, counts_per_cell_after=1e4)


filter_result = sc.pp.filter_genes_dispersion(
    ojelly3.X,min_mean=0.0125, max_mean=4.5, min_disp=0.50) 
sc.pl.filter_genes_dispersion(filter_result)
print(&#39;num genes: &#39;+str(sum(filter_result.gene_subset)))

ojelly3 = ojelly3[:, filter_result.gene_subset]

ojelly3
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../cellRangerClustering_Starvation_files/cellRangerClustering_Starvation_15_0.png" /></p>
<div class="codehilite"><pre><span></span><code><span class="n">num</span> <span class="n">genes</span><span class="p">:</span> <span class="mi">2879</span>





<span class="n">View</span> <span class="n">of</span> <span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">16813</span> <span class="err">×</span> <span class="mi">2879</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_cells&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.scale(ojelly3, max_value=10)

sc.tl.pca(ojelly3, n_comps=60)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.tsne(ojelly3, n_pcs=60)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">WARNING</span><span class="o">:</span> <span class="n">Consider</span> <span class="n">installing</span> <span class="n">the</span> <span class="k">package</span> <span class="nn">MulticoreTSNE</span> <span class="o">(</span><span class="n">https</span><span class="o">://</span><span class="n">github</span><span class="o">.</span><span class="na">com</span><span class="sr">/DmitryUlyanov/</span><span class="n">Multicore</span><span class="o">-</span><span class="n">TSNE</span><span class="o">).</span> <span class="n">Even</span> <span class="k">for</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span> <span class="k">this</span> <span class="n">speeds</span> <span class="n">up</span> <span class="n">the</span> <span class="n">computation</span> <span class="n">considerably</span> <span class="n">and</span> <span class="n">might</span> <span class="n">yield</span> <span class="n">better</span> <span class="n">converged</span> <span class="n">results</span><span class="o">.</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.neighbors(ojelly3)

sc.tl.louvain(ojelly3) 

sc.pl.tsne(ojelly3, color=[&#39;louvain&#39;])
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../cellRangerClustering_Starvation_files/cellRangerClustering_Starvation_18_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.umap(ojelly3)

sc.pl.umap(ojelly3, color=[&#39;louvain&#39;])
</code></pre></div>
</td></tr></table>
<p>Filter for low count cells</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>clusterspassingfilter=list(set(ojelly3.obs[&#39;louvain&#39;]).difference({&#39;0&#39;}))
ojf = ojelly3[ojelly3.obs[&#39;louvain&#39;].isin(clusterspassingfilter)]

ojf
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Choose embedding
sc.tl.umap(ojf)
sc.pl.umap(ojf, color=[&#39;louvain&#39;, &#39;XLOC_008048&#39;,&#39;XLOC_019434&#39;])

tempoj=copy.deepcopy(ojf)

sc.tl.umap(tempoj, random_state=None)
sc.pl.umap(tempoj, color=[&#39;louvain&#39;, &#39;XLOC_008048&#39;,&#39;XLOC_019434&#39;])

tempoj2=copy.deepcopy(tempoj)
sc.tl.umap(tempoj2, random_state=None)
sc.pl.umap(tempoj2, color=[&#39;louvain&#39;, &#39;XLOC_008048&#39;,&#39;XLOC_019434&#39;])

#Test different pc #&#39;s
tempoj16=sc.pp.neighbors(tempoj16, n_pcs=16, copy=True)
tempoj40=sc.pp.neighbors(tempoj16, n_pcs=40, copy=True)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.umap(tempoj40, random_state=None)
sc.pl.umap(tempoj40, color=[&#39;louvain&#39;, &#39;XLOC_008048&#39;,&#39;XLOC_019434&#39;])
</code></pre></div>
</td></tr></table>
<p>Save output with all organism labels</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>&quot;&quot;&quot;Differential Expression, cluster identification&quot;&quot;&quot;
!mv D1.1812 D1.1812.h5ad
trinojelly3 = anndata.read(&#39;D1.1812.h5ad&#39;) 

trinojelly3

tempoj2.obs[&#39;orgID&#39;]=trinojelly3.obs[&#39;orgID&#39;]
tempoj2.obs[&#39;fed&#39;]=trinojelly3.obs[&#39;fed&#39;]
tempoj2.obs[&#39;starved&#39;]=trinojelly3.obs[&#39;starved&#39;]

tempoj40.obs[&#39;orgID&#39;]=trinojelly3.obs[&#39;orgID&#39;]
tempoj40.obs[&#39;fed&#39;]=trinojelly3.obs[&#39;fed&#39;]
tempoj40.obs[&#39;starved&#39;]=trinojelly3.obs[&#39;starved&#39;]

tempoj40.obs[&#39;fed_ord&#39;] = pd.Categorical(tempoj40.obs[&#39;fed&#39;])
tempoj40.obs[&#39;starved_ord&#39;] = pd.Categorical(tempoj40.obs[&#39;starved&#39;])

sc.pl.umap(tempoj40, color=[&#39;louvain&#39;])
sc.pl.umap(tempoj40, color=[&#39;orgID&#39;])
sc.pl.umap(tempoj40, color=[&#39;fed_ord&#39;], palette=&#39;viridis_r&#39;)
sc.pl.umap(tempoj40, color=[&#39;starved_ord&#39;], palette=&#39;viridis_r&#39;)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#tempoj40.write(&#39;cellRanger_fs.h5ad&#39;)
</code></pre></div>
</td></tr></table>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../kallistoBusRuns_StarvAndStim/" class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Quantification of cDNA Counts with kallisto-bustools (All Data)
              </div>
            </div>
          </a>
        
        
          <a href="../starvation_Analysis/" class="md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Cell Atlas Clustering and Analysis from kallisto-bustools
              </div>
            </div>
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/vendor.53cc9318.min.js"></script>
      <script src="../../assets/javascripts/bundle.e9c9f54f.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "../..",
          features: ['search.suggest', 'search.highlight'],
          search: Object.assign({
            worker: "../../assets/javascripts/worker/search.9c0e82ba.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>