
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Walkthrough for Whole Organisms Mutliplexed Data Analysis">
      
      
      
        <meta name="author" content="Tara Chari">
      
      
        <link rel="canonical" href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/notebooks/allDistanceCalculations/">
      
      <link rel="shortcut icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.2.7">
    
    
      
        <title>Comparative Analysis of Cell Atlas Concordance and Perturbation Responses - WHAMSeqAnalysis</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.cb6bc1d0.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.39b8e14a.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#download-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/" title="WHAMSeqAnalysis" class="md-header-nav__button md-logo" aria-label="WHAMSeqAnalysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      <div class="md-header-nav__ellipsis">
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            WHAMSeqAnalysis
          </span>
        </div>
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            
              Comparative Analysis of Cell Atlas Concordance and Perturbation Responses
            
          </span>
        </div>
      </div>
    </div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    




<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/" title="WHAMSeqAnalysis" class="md-nav__button md-logo" aria-label="WHAMSeqAnalysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    WHAMSeqAnalysis
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" >
      
      <label class="md-nav__link" for="nav-2">
        ClickTag Pre-processing
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="ClickTag Pre-processing" data-md-level="1">
        <label class="md-nav__title" for="nav-2">
          <span class="md-nav__icon md-icon"></span>
          ClickTag Pre-processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellRangerClickTagCounts/" class="md-nav__link">
        Initial Quantification of ClickTag Counts with Cell Ranger (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistobusStarvClickTags/" class="md-nav__link">
        Quantification of ClickTag Counts with kallisto-bustools (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistobusStimClickTags/" class="md-nav__link">
        Quantification of ClickTag Counts with kallisto-bustools (Stimulation Data)
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3" >
      
      <label class="md-nav__link" for="nav-3">
        cDNA Pre-processing
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="cDNA Pre-processing" data-md-level="1">
        <label class="md-nav__title" for="nav-3">
          <span class="md-nav__icon md-icon"></span>
          cDNA Pre-processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../filterStarvCells_ClickTags/" class="md-nav__link">
        Initial cDNA Filtering with Cell Ranger Data (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistoBusRuns_StarvAndStim/" class="md-nav__link">
        Quantification of cDNA Counts with kallisto-bustools (All Data)
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4" >
      
      <label class="md-nav__link" for="nav-4">
        Cell Atlas and Perturbation Response
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Cell Atlas and Perturbation Response" data-md-level="1">
        <label class="md-nav__title" for="nav-4">
          <span class="md-nav__icon md-icon"></span>
          Cell Atlas and Perturbation Response
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellRangerClustering_Starvation/" class="md-nav__link">
        Initial Cell Atlas Clustering from Cell Ranger
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../starvation_Analysis/" class="md-nav__link">
        Cell Atlas Clustering and Analysis from kallisto-bustools
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../neuronSubpop_Analysis/" class="md-nav__link">
        Analysis of Neuron Subpopulations
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../pseudotime_Analysis/" class="md-nav__link">
        Pseudotime Analysis
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../deSeq2Analysis_StarvationResponse/" class="md-nav__link">
        Extraction of Differentially Expressed Genes Under Starvation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5" >
      
      <label class="md-nav__link" for="nav-5">
        Stimulation Data Analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Stimulation Data Analysis" data-md-level="1">
        <label class="md-nav__title" for="nav-5">
          <span class="md-nav__icon md-icon"></span>
          Stimulation Data Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../stimulation_Analysis/" class="md-nav__link">
        Cell Atlas Concordance and Perturbation Response
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../deSeq2Analysis_StimulationResponse/" class="md-nav__link">
        Extraction of Differentially Expressed Genes Under Starvation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6" checked>
      
      <label class="md-nav__link" for="nav-6">
        Inter-Experimental Analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Inter-Experimental Analysis" data-md-level="1">
        <label class="md-nav__title" for="nav-6">
          <span class="md-nav__icon md-icon"></span>
          Inter-Experimental Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Comparative Analysis of Cell Atlas Concordance and Perturbation Responses
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Comparative Analysis of Cell Atlas Concordance and Perturbation Responses
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import packages
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#do-distance-based-analysis-for-inter-and-intra-cluster-distances" class="md-nav__link">
    Do Distance Based Analysis for Inter- and Intra- Cluster Distances
  </a>
  
    <nav class="md-nav" aria-label="Do Distance Based Analysis for Inter- and Intra- Cluster Distances">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-merged-dataset-joint-pc-space-with-stimulation-and-starvation-experiments-was-made-overlap_combo" class="md-nav__link">
    How merged dataset (joint-PC space) with stimulation and starvation experiments was made (= overlap_combo)
  </a>
  
    <nav class="md-nav" aria-label="How merged dataset (joint-PC space) with stimulation and starvation experiments was made (= overlap_combo)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#plot-percent-of-k-neighbors-for-each-stim-cell-that-are-in-one-unique-cluster" class="md-nav__link">
    Plot percent of K neighbors for each stim cell that are in one unique cluster
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#determine-inter-and-intra-cluster-distances-perturbation-effect-vs-cell-type-distance" class="md-nav__link">
    Determine Inter and Intra Cluster Distances (perturbation effect vs cell type distance)
  </a>
  
    <nav class="md-nav" aria-label="Determine Inter and Intra Cluster Distances (perturbation effect vs cell type distance)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fedstarved-l1-analysis" class="md-nav__link">
    Fed/Starved L1 Analysis
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#repeat-for-stimulation-data" class="md-nav__link">
    Repeat for Stimulation Data
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#organismal-distances-fedstarved-stimulation" class="md-nav__link">
    Organismal Distances, Fed/Starved + Stimulation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-7" type="checkbox" id="nav-7" >
      
      <label class="md-nav__link" for="nav-7">
        Explore Mutliplexed Datasets
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Explore Mutliplexed Datasets" data-md-level="1">
        <label class="md-nav__title" for="nav-7">
          <span class="md-nav__icon md-icon"></span>
          Explore Mutliplexed Datasets
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../MARIMBAAnnosAnalysis/" class="md-nav__link">
        Generation Cell Atlas with MARIMBA Transcriptome
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../MARIMBAAtlasSearchAndPlot/" class="md-nav__link">
        Explore MARIMBA-Annotated Cell Atlas
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellAtlasSearchAndPlot/" class="md-nav__link">
        Explore Original Cell Atlas
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import packages
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#do-distance-based-analysis-for-inter-and-intra-cluster-distances" class="md-nav__link">
    Do Distance Based Analysis for Inter- and Intra- Cluster Distances
  </a>
  
    <nav class="md-nav" aria-label="Do Distance Based Analysis for Inter- and Intra- Cluster Distances">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-merged-dataset-joint-pc-space-with-stimulation-and-starvation-experiments-was-made-overlap_combo" class="md-nav__link">
    How merged dataset (joint-PC space) with stimulation and starvation experiments was made (= overlap_combo)
  </a>
  
    <nav class="md-nav" aria-label="How merged dataset (joint-PC space) with stimulation and starvation experiments was made (= overlap_combo)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#plot-percent-of-k-neighbors-for-each-stim-cell-that-are-in-one-unique-cluster" class="md-nav__link">
    Plot percent of K neighbors for each stim cell that are in one unique cluster
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#determine-inter-and-intra-cluster-distances-perturbation-effect-vs-cell-type-distance" class="md-nav__link">
    Determine Inter and Intra Cluster Distances (perturbation effect vs cell type distance)
  </a>
  
    <nav class="md-nav" aria-label="Determine Inter and Intra Cluster Distances (perturbation effect vs cell type distance)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fedstarved-l1-analysis" class="md-nav__link">
    Fed/Starved L1 Analysis
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#repeat-for-stimulation-data" class="md-nav__link">
    Repeat for Stimulation Data
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#organismal-distances-fedstarved-stimulation" class="md-nav__link">
    Organismal Distances, Fed/Starved + Stimulation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/edit/master/docs/notebooks/allDistanceCalculations.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                  <h1>Comparative Analysis of Cell Atlas Concordance and Perturbation Responses</h1>
                
                <p><a href="https://colab.research.google.com/github/pachterlab/CWGFLHGCCHAP_2021/blob/master/notebooks/ComparativeDistanceAnalysis/allDistanceCalculations.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!date
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">Mon Aug 17 20:20:34 UTC 2020</span>
</code></pre></div>

<h3 id="download-data"><strong>Download Data</strong><a class="headerlink" href="#download-data" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import requests
from tqdm import tnrange, tqdm_notebook
def download_file(doi,ext):
    url = &#39;https://api.datacite.org/dois/&#39;+doi+&#39;/media&#39;
    r = requests.get(url).json()
    netcdf_url = r[&#39;data&#39;][0][&#39;attributes&#39;][&#39;url&#39;]
    r = requests.get(netcdf_url,stream=True)
    #Set file name
    fname = doi.split(&#39;/&#39;)[-1]+ext
    #Download file with progress bar
    if r.status_code == 403:
        print(&quot;File Unavailable&quot;)
    if &#39;content-length&#39; not in r.headers:
        print(&quot;Did not get file&quot;)
    else:
        with open(fname, &#39;wb&#39;) as f:
            total_length = int(r.headers.get(&#39;content-length&#39;))
            pbar = tnrange(int(total_length/1024), unit=&quot;B&quot;)
            for chunk in r.iter_content(chunk_size=1024):
                if chunk:
                    pbar.update()
                    f.write(chunk)
        return fname
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Kallisto bus clustered starvation data, h5ad
download_file(&#39;10.22002/D1.1796&#39;,&#39;.gz&#39;)

#CellRanger Starvation h5ad data
download_file(&#39;10.22002/D1.1798&#39;,&#39;.gz&#39;)

#Import previously saved, clustered, &amp; filtered stimulation data 

download_file(&#39;10.22002/D1.1821&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">479630</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">45376</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">772581</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1821.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Starvation h5ad data, all nonzero genes included, filtered for &#39;real cells&#39; from de-multiplexing
download_file(&#39;10.22002/D1.1797&#39;,&#39;.gz&#39;)

#Import raw, unclustered stimulation data
download_file(&#39;10.22002/D1.1814&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">26058</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">78502</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1814.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Import merged data with knn clusters
download_file(&#39;10.22002/D1.1823&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">257856</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1823.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!gunzip *.gz
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!pip install --quiet anndata
!pip install --quiet scanpy
!pip install --quiet louvain
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">[K     |████████████████████████████████| 122kB 4.3MB/s </span>
<span class="err">[K     |████████████████████████████████| 10.2MB 4.0MB/s </span>
<span class="err">[K     |████████████████████████████████| 71kB 7.6MB/s </span>
<span class="err">[K     |████████████████████████████████| 51kB 6.1MB/s </span>
<span class="err">[K     |████████████████████████████████| 71kB 7.9MB/s </span>
<span class="err">[?25h  Building wheel for sinfo (setup.py) ... [?25l[?25hdone</span>
<span class="err">  Building wheel for umap-learn (setup.py) ... [?25l[?25hdone</span>
<span class="err">[K     |████████████████████████████████| 2.2MB 4.3MB/s </span>
<span class="err">[K     |████████████████████████████████| 3.2MB 44.6MB/s </span>
<span class="err">[?25h</span>
</code></pre></div>

<h3 id="import-packages"><strong>Import packages</strong><a class="headerlink" href="#import-packages" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import pandas as pd
import anndata
import scanpy as sc
import numpy as np
import scipy.sparse

import warnings
warnings.filterwarnings(&#39;ignore&#39;)

from sklearn.neighbors import (KNeighborsClassifier,NeighborhoodComponentsAnalysis)
from sklearn.pipeline import Pipeline
from sklearn.manifold import TSNE
from sklearn.decomposition import PCA

import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline
sc.set_figure_params(dpi=125)

import seaborn as sns
sns.set(style=&quot;whitegrid&quot;)
</code></pre></div>
</td></tr></table>
<h3 id="do-distance-based-analysis-for-inter-and-intra-cluster-distances"><strong>Do Distance Based Analysis for Inter- and Intra- Cluster Distances</strong><a class="headerlink" href="#do-distance-based-analysis-for-inter-and-intra-cluster-distances" title="Permanent link">&para;</a></h3>
<p>Read in previously saved data</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># Description: bus_fs_combo data (filtered + clustered starvation (fed/starved --&gt; fs) data)
# Description: bus_combo_noZero (filtered + clustered stimulation data) from Kallisto bus

#bus_fs_combo
bus_fs_combo = anndata.read(&quot;D1.1796&quot;)

print(bus_fs_combo)

bus_combo_noZero = anndata.read(&#39;D1.1821&#39;) 
print(bus_combo_noZero)

#Previously saved overlap, merged data (from both experiments)
overlap_combo = anndata.read(&quot;D1.1823&quot;)
overlap_combo
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">13673</span> <span class="err">×</span> <span class="mi">8696</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;starved&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_neighbor_score&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;annos&#39;</span><span class="p">,</span> <span class="s1">&#39;new_cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;annosSub&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;annosSub_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;annos_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain_sizes&#39;</span><span class="p">,</span> <span class="s2">&quot;dendrogram_[&#39;new_cellRanger_louvain&#39;]&quot;</span><span class="p">,</span> <span class="s1">&#39;dendrogram_new_cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_neighbor_score_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;new_cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;paga&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_nca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
    <span class="n">obsp</span><span class="p">:</span> <span class="s1">&#39;connectivities&#39;</span><span class="p">,</span> <span class="s1">&#39;distances&#39;</span>
<span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">18921</span> <span class="err">×</span> <span class="mi">10260</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;condition&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain_sizes&#39;</span><span class="p">,</span> <span class="s1">&#39;condition_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;paga&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
    <span class="n">obsp</span><span class="p">:</span> <span class="s1">&#39;connectivities&#39;</span><span class="p">,</span> <span class="s1">&#39;distances&#39;</span>





<span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">32594</span> <span class="err">×</span> <span class="mi">6756</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;starved&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_neighbor_score&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;annos&#39;</span><span class="p">,</span> <span class="s1">&#39;new_cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;annosSub&#39;</span><span class="p">,</span> <span class="s1">&#39;condition&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_origin&#39;</span><span class="p">,</span> <span class="s1">&#39;knn_clus&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts-0&#39;</span><span class="p">,</span> <span class="s1">&#39;mean-0&#39;</span><span class="p">,</span> <span class="s1">&#39;std-0&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts-1&#39;</span><span class="p">,</span> <span class="s1">&#39;mean-1&#39;</span><span class="p">,</span> <span class="s1">&#39;std-1&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;pca&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_nca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#How raw overlap adata was made (no genes filtered out)
bus_fs_raw = anndata.read(&quot;D1.1797&quot;)

bus_stim_raw = anndata.read(&quot;D1.1814&quot;)

#Merge datasets
bus_fs_raw.obs_names = [i+&#39;_fs&#39; for i in bus_fs_raw.obs_names]
bus_stim_raw.obs_names = [i+&#39;_ieg&#39; for i in bus_stim_raw.obs_names]

raw_overlap_combo = bus_fs_raw.concatenate(bus_stim_raw,join=&#39;outer&#39;, index_unique=None)
</code></pre></div>
</td></tr></table>
<p>Predict labels for stimulation data</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Make X_pca with merged, use top n PC&#39;s
def getPredLabels(overlap_fs,overlap_combo, n_PC):

    numFS = len(overlap_fs.obs_names)
    X_train = overlap_combo.obsm[&#39;X_pca&#39;][0:numFS,0:n_PC]

    #X_pca at stim rows = X_test
    X_test = overlap_combo.obsm[&#39;X_pca&#39;][numFS:,0:n_PC]

    #y_train is f/s louvain labels
    y_train = overlap_fs.obs[&#39;cellRanger_louvain&#39;]

    #Fit and predict
    classifier = KNeighborsClassifier(n_neighbors=15)
    classifier.fit(X_train, y_train)

    y_pred = classifier.predict(X_test)

    labels = list(y_train)+list(y_pred)

    print(len(labels),&#39; labels created&#39;)

    return labels
</code></pre></div>
</td></tr></table>
<h4 id="how-merged-dataset-joint-pc-space-with-stimulation-and-starvation-experiments-was-made-overlap_combo"><strong>How merged dataset (joint-PC space) with stimulation and starvation experiments was made (= overlap_combo)</strong><a class="headerlink" href="#how-merged-dataset-joint-pc-space-with-stimulation-and-starvation-experiments-was-made-overlap_combo" title="Permanent link">&para;</a></h4>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Get intersection of var names between the two (making overlap_combo)
overlap = list(set(bus_combo_noZero.var_names).intersection(bus_fs_combo.var_names))

overlap_fs = bus_fs_combo[:,overlap]

overlap_stim = bus_combo_noZero[:,overlap]


#Merge datasets
overlap_fs.obs_names = [i+&#39;_fs&#39; for i in overlap_fs.obs_names]
overlap_stim.obs_names = [i+&#39;_ieg&#39; for i in overlap_stim.obs_names]

origin = list(np.repeat(&#39;FS&#39;,len(overlap_fs.obs_names))) + list(np.repeat(&#39;Stim&#39;,len(overlap_stim.obs_names)))
overlap_combo = overlap_fs.concatenate(overlap_stim,join=&#39;outer&#39;, index_unique=None)

overlap_combo.obs[&#39;cell_origin&#39;] = pd.Categorical(origin)
sc.pp.scale(overlap_combo, max_value=10)

#Do PCA on merged dataset + plot variance explained
sc.tl.pca(overlap_combo, n_comps=60)
#sc.pl.pca_variance_ratio(overlap_combo, log=True)

#overlap_combo

labels = getPredLabels(overlap_fs,overlap_combo,n_PC = 60)

overlap_combo.obs[&#39;knn_clus&#39;] = pd.Categorical(labels)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">32594  labels created</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pp.neighbors(overlap_combo,n_neighbors=150, n_pcs=60,random_state=42)
sc.tl.paga(overlap_combo, groups=&#39;knn_clus&#39;)
sc.pl.paga(overlap_combo, color=[&#39;knn_clus&#39;])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.tsne(overlap_combo,random_state=42,n_pcs=60,early_exaggeration=5)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">WARNING</span><span class="o">:</span> <span class="n">Consider</span> <span class="n">installing</span> <span class="n">the</span> <span class="k">package</span> <span class="nn">MulticoreTSNE</span> <span class="o">(</span><span class="n">https</span><span class="o">://</span><span class="n">github</span><span class="o">.</span><span class="na">com</span><span class="sr">/DmitryUlyanov/</span><span class="n">Multicore</span><span class="o">-</span><span class="n">TSNE</span><span class="o">).</span> <span class="n">Even</span> <span class="k">for</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span> <span class="k">this</span> <span class="n">speeds</span> <span class="n">up</span> <span class="n">the</span> <span class="n">computation</span> <span class="n">considerably</span> <span class="n">and</span> <span class="n">might</span> <span class="n">yield</span> <span class="n">better</span> <span class="n">converged</span> <span class="n">results</span><span class="o">.</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.umap(overlap_combo,random_state=42,spread=2.5,min_dist = 0.8,init_pos=&#39;paga&#39;)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># sc.pl.tsne(overlap_combo,color=[&#39;knn_clus&#39;])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.umap(overlap_combo,color=[&#39;knn_clus&#39;],color_map=&#39;viridis&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../allDistanceCalculations_files/allDistanceCalculations_23_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#overlap_combo.write(&#39;overlap_combo.h5ad&#39;)
</code></pre></div>
</td></tr></table>
<h5 id="plot-percent-of-k-neighbors-for-each-stim-cell-that-are-in-one-unique-cluster"><strong>Plot percent of K neighbors for each stim cell that are in one unique cluster</strong><a class="headerlink" href="#plot-percent-of-k-neighbors-for-each-stim-cell-that-are-in-one-unique-cluster" title="Permanent link">&para;</a></h5>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>from sklearn.neighbors import NearestNeighbors
import random 

#Read in previously saved data
overlap_combo = anndata.read(&quot;D1.1823&quot;)

n = 15

neigh = NearestNeighbors(n_neighbors=n)
n_PC = 60


numFS = len(overlap_fs.obs_names)
X_train = overlap_combo.obsm[&#39;X_pca&#39;][0:numFS,0:n_PC]

#X_pca at stim rows = X_test
X_test = overlap_combo.obsm[&#39;X_pca&#39;][numFS:,0:n_PC]

#y_train is f/s louvain labels
y_train = overlap_fs.obs[&#39;cellRanger_louvain&#39;]

#Fit and predict
#neigh.fit(X_train)


#Subsample 70% of fed/starved data for training, test on remaining 30% and stim data
numSamp = np.int(0.7*X_train.shape[0])
ind = random.sample(range(0,X_train.shape[0]),numSamp)
allInd = range(0,X_train.shape[0])
rest = [i for i in allInd if i not in ind]

#Fit and predict
neigh.fit(X_train[ind,:])

X_rest = X_train[rest,:]



fsNeigh = neigh.kneighbors(X_rest)
stimNeigh = neigh.kneighbors(X_test)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>y_train = y_train[ind]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>stim_neighbors = stimNeigh[1]
fs_neighbors = fsNeigh[1]

#y_train = y_train[ind]
stim_perc = np.zeros(stim_neighbors.shape[0])
for i in range(0,len(stim_perc)): 
  #How many of top neighbors come from same cluster in fed/starved data (out of n=15)
  stim_perc[i] = np.max(y_train[stim_neighbors[i]].value_counts())/n 

fs_perc = np.zeros(fs_neighbors.shape[0])
#y_train = y_train[ind]
for i in range(0,len(fs_perc)): 
  fs_perc[i] = np.max(y_train[fs_neighbors[i]].value_counts())/n
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>plt.hist(stim_perc,density=False,range=(0,1),alpha=0.6 ,bins=15,weights=np.ones(len(stim_perc))/len(stim_perc),label=&#39;stim-neighbors&#39;)
plt.hist(fs_perc,density=False,range=(0,1),alpha=0.6 ,bins=15 ,weights=np.ones(len(fs_perc))/len(fs_perc),label=&#39;fs-neighbors&#39;)
plt.legend(loc=&#39;upper left&#39;)
plt.title(&#39;Histogram of Nearest Neighbor Specificity&#39;,fontsize=11)
plt.grid(b=None)
plt.xlabel(&quot;Fraction of n=15 nearest neighbors in unique cluster&quot;,fontsize=10)
plt.ylabel(&quot;Frequency&quot;,fontsize=10)
plt.show()
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../allDistanceCalculations_files/allDistanceCalculations_29_0.png" /></p>
<h4 id="determine-inter-and-intra-cluster-distances-perturbation-effect-vs-cell-type-distance"><strong>Determine Inter and Intra Cluster Distances (perturbation effect vs cell type distance)</strong><a class="headerlink" href="#determine-inter-and-intra-cluster-distances-perturbation-effect-vs-cell-type-distance" title="Permanent link">&para;</a></h4>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Plot distances between centroid of stim-starved and fed-starved for each cluster
def changeBool(fed):
    if fed == &#39;True&#39;:
        return &#39;fed&#39;
    elif fed == &#39;False&#39;:
        return &#39;starved&#39; 
    else:
        return &#39;stim&#39;

def addConds(overlap_combo):
    fs_conds = [changeBool(i) for i in overlap_combo.obs[&#39;fed&#39;]]
    overlap_combo.obs[&#39;centroid_conds&#39;] = pd.Categorical(fs_conds)

    overlap_combo.obs[&#39;centroid_conds&#39;]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>from sklearn.metrics import pairwise_distances
#Centroids of cell atlas/defined clusters
def getClusCentroids(overlap_combo,pcs=60,clusType=&#39;knn_clus&#39;):
    clusters = np.unique(overlap_combo.obs[clusType])
    centroids = np.zeros((len(clusters),pcs))

    for c in clusters:

        sub_data = overlap_combo[overlap_combo.obs[clusType] == c]
        pca_embed = sub_data.obsm[&#39;X_pca&#39;][:,0:pcs]
        centroid = pca_embed.mean(axis=0)

        centroids[c,:] = list(centroid)

    return centroids
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Distance between conditions within clusters
def getCentroidDists(overlap_combo,pcs=60,clusType=&#39;knn_clus&#39;):
    fedStarv = []
    fsStim = []

    cluster = []

    #Initialize output matrix
    addConds(overlap_combo)
    clusters = np.unique(overlap_combo.obs[clusType])
    centroid_dist = pd.DataFrame(columns =[&#39;fedStarv&#39;,&#39;fsStim&#39;,&#39;cluster&#39;,&#39;clus_color&#39;]) 

    for c in clusters:
        #Get cells in cluster and their conditions
        sub_data = overlap_combo[overlap_combo.obs[clusType] == c]
        centroid_conds = sub_data.obs[&#39;centroid_conds&#39;] 

        #Get 2D tsne embedding
        pca_embed = sub_data.obsm[&#39;X_pca&#39;][:,0:pcs] #PCA

        #Get location of condition cells
        stim_pos = list(centroid_conds == &#39;stim&#39;)
        allFS_pos = list(centroid_conds != &#39;stim&#39;)
        fed_pos = list(centroid_conds == &#39;fed&#39;)
        starved_pos = list(centroid_conds == &#39;starved&#39;)

        #Get column means for x,y coords
        meanFed = pca_embed[fed_pos,].mean(axis=0)

        meanStarv = pca_embed[starved_pos,].mean(axis=0)

        meanStim = pca_embed[stim_pos,].mean(axis=0)

        meanFS = pca_embed[allFS_pos,].mean(axis=0)


        cluster += [c]

        #Dist between fed and starved
        fs_dist = np.linalg.norm(meanFed - meanStarv,1)

        #Dist between (all) starved experiment and stim

        starvStim_dist = np.linalg.norm(meanFS - meanStim,1)

        fedStarv += [fs_dist]
        fsStim += [starvStim_dist]

    centroid_dist[&#39;fedStarv&#39;] = fedStarv
    centroid_dist[&#39;fsStim&#39;] = fsStim

    centroid_dist[&#39;cluster&#39;] = cluster

    return centroid_dist
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Distance between conditions within clusters
def getStimCentroidDists(overlap_combo,pcs=60,clusType=&#39;knn_clus&#39;):
    swKCl = []
    swDI = []

    cluster = []

    #Initialize output matrix
    #addConds(overlap_combo)
    clusters = np.unique(overlap_combo.obs[clusType])
    centroid_dist = pd.DataFrame(columns =[&#39;swKCl&#39;,&#39;swDI&#39;,&#39;cluster&#39;,&#39;clus_color&#39;]) 

    for c in clusters:
        #Get cells in cluster and their conditions
        sub_data = overlap_combo[overlap_combo.obs[clusType] == c]
        centroid_conds = sub_data.obs[&#39;condition&#39;] 

        #Get 2D tsne embedding
        pca_embed = sub_data.obsm[&#39;X_pca&#39;][:,0:pcs] #PCA

        #Get location of condition cells
        sw_pos = list(centroid_conds == &#39;SW&#39;)
        kcl_pos = list(centroid_conds == &#39;KCl&#39;)
        di_pos = list(centroid_conds == &#39;DI&#39;)


        #Get column means for x,y coords
        meanSW = pca_embed[sw_pos,].mean(axis=0)

        meanKCl = pca_embed[kcl_pos,].mean(axis=0)

        meanDI = pca_embed[di_pos,].mean(axis=0)




        cluster += [c]

        #Dist between fed and starved
        swKCl_dist = np.linalg.norm(meanSW - meanKCl,1)

        #Dist between (all) starved experiment and stim

        swDI_dist = np.linalg.norm(meanSW - meanDI,1)

        swKCl += [swKCl_dist]
        swDI += [swDI_dist]

    centroid_dist[&#39;swKCl&#39;] = swKCl 
    centroid_dist[&#39;swDI&#39;] =  swDI

    centroid_dist[&#39;cluster&#39;] = cluster

    return centroid_dist
</code></pre></div>
</td></tr></table>
<h5 id="fedstarved-l1-analysis"><strong>Fed/Starved L1 Analysis</strong><a class="headerlink" href="#fedstarved-l1-analysis" title="Permanent link">&para;</a></h5>
<p>Cell type v state plots</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>withinFS_Dists = getCentroidDists(bus_fs_combo,60,&#39;cellRanger_louvain&#39;)
withinFS_Dists.head()
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>fedStarv</th>
      <th>fsStim</th>
      <th>cluster</th>
      <th>clus_color</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>11.357661</td>
      <td>NaN</td>
      <td>0</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>17.776861</td>
      <td>NaN</td>
      <td>1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>12.411180</td>
      <td>NaN</td>
      <td>2</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>18.126976</td>
      <td>NaN</td>
      <td>3</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>13.799874</td>
      <td>NaN</td>
      <td>4</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Compare to pairwise distances between cell atlas clusters
centroids = getClusCentroids(bus_fs_combo,60,&#39;cellRanger_louvain&#39;)
#centroids_arr = centroids[&#39;centroid&#39;].to_numpy()
pairCentroid_dists = pairwise_distances(centroids, metric = &#39;l1&#39;)
pairCentroid_dists.shape
print(np.mean(pairCentroid_dists))
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">150.34457080272998</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>np.unique(bus_fs_combo.obs[&#39;cellRanger_louvain&#39;])
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,</span>
<span class="err">       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,</span>
<span class="err">       34, 35])</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Early cnidocytes, germ cells, stem cells have distances that overlap with intra-F/S distances for 14&amp;19
np.where(pairCentroid_dists &lt; 62)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">(array([ 0,  0,  1,  2,  2,  3,  3,  4,  5,  6,  7,  7,  8,  9, 10, 11, 12,</span>
<span class="err">        12, 13, 14, 15, 16, 17, 18, 19, 20, 20, 21, 22, 23, 24, 25, 26, 27,</span>
<span class="err">        28, 29, 30, 31, 32, 33, 34, 35]),</span>
<span class="err"> array([ 0, 12,  1,  2, 20,  3,  7,  4,  5,  6,  3,  7,  8,  9, 10, 11,  0,</span>
<span class="err">        12, 13, 14, 15, 16, 17, 18, 19,  2, 20, 21, 22, 23, 24, 25, 26, 27,</span>
<span class="err">        28, 29, 30, 31, 32, 33, 34, 35]))</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>max(pairCentroid_dists.reshape((36*36,)))
reshape_noZero = [i for i in pairCentroid_dists.reshape((36*36,)) if i != 0]
print(np.min(reshape_noZero))

matplotlib.rc(&#39;axes&#39;,edgecolor=&#39;black&#39;)

plt.figure(figsize=(16,10))
plt.hist(withinFS_Dists[&#39;fedStarv&#39;],color=&#39;r&#39;,bins=15,alpha=0.6,label=&#39;fed-starved&#39;,density=True)

plt.hist(reshape_noZero, bins=36,color = &#39;orange&#39;,alpha=0.6,label=&#39;pairwise clusters&#39;,density=True)
plt.legend(loc=&#39;upper right&#39;)
plt.title(&#39;Comparison to pairwise distribution (pairwise on clusters with fed/starved cells)&#39;)
plt.grid(b=None)
plt.xlabel(&quot;L1 Distance with 60 PC&#39;s&quot;)
plt.xticks(fontsize=18)
plt.yticks(fontsize=18)
plt.show()
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">49.54310559108853</span>
</code></pre></div>

<p><img alt="png" src="../allDistanceCalculations_files/allDistanceCalculations_40_1.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Area where line between cell state/type is blurred seems to be only in digestive endodermal clusters 14,19 (strong metabolic responses to starvation)
outlier_clus = withinFS_Dists[withinFS_Dists.fedStarv&gt; 46.69][&#39;cluster&#39;]
outlier_clus
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">14    14</span>
<span class="err">19    19</span>
<span class="c">Name: cluster, dtype: int64</span>
</code></pre></div>

<p>Compare internal distances to differences in numbers of fed &amp; starved cells (Spearman Rank correlation)</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Composition of broad cell types between fed and starved , cell counts
counts = pd.DataFrame(columns =[&#39;count&#39;, &#39;orgID&#39;,&#39;condition&#39;,&#39;cluster&#39;]) 
clusters = np.unique(bus_fs_combo.obs[&#39;cellRanger_louvain&#39;])

c = []
org = []
cond = []
clus = []

orgs = np.unique(bus_fs_combo.obs[&#39;orgID&#39;])
conds = [&#39;True&#39;, &#39;False&#39;]
for cl in clusters:
    data = bus_fs_combo[bus_fs_combo.obs[&#39;cellRanger_louvain&#39;].isin([cl])]

    for cd in conds:
        #c_data = data[data.obs[&#39;condition&#39;].isin([cd])]

        for o in orgs:
            pos = (data.obs[&#39;fed&#39;].isin([cd])) &amp; (data.obs[&#39;orgID&#39;].isin([o]))
            org_data = data[pos]
            c += [org_data.n_obs]
            org += [o]
            if cd == &#39;True&#39;:
              cond += [&#39;fed&#39;]
            else:
              cond += [&#39;starved&#39;]
            clus += [cl]


print(len(c))




counts[&#39;count&#39;] = c
counts[&#39;orgID&#39;] = org
counts[&#39;condition&#39;] = cond
counts[&#39;cluster&#39;] = clus
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">720</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>counts = counts[counts[&#39;count&#39;]&gt; 0 ]
counts.head()
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>count</th>
      <th>orgID</th>
      <th>condition</th>
      <th>cluster</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>118</td>
      <td>1</td>
      <td>fed</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>83</td>
      <td>2</td>
      <td>fed</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>102</td>
      <td>3</td>
      <td>fed</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>102</td>
      <td>4</td>
      <td>fed</td>
      <td>0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>118</td>
      <td>5</td>
      <td>fed</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#One-way anova for each cluster
pvalsFTest = {}

for cl in clusters:
  sub = counts[counts.cluster == cl]
  subFed = sub[sub.condition == &#39;fed&#39;]
  subStarv = sub[sub.condition == &#39;starved&#39;]

  fedVals = list(subFed[&#39;count&#39;])
  starvVals = list(subStarv[&#39;count&#39;])

  pvalsFTest[cl] = scipy.stats.f_oneway(fedVals,starvVals)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import statsmodels.stats as sm
pvals = []
# sizes = [] 
for c in clusters:
  pvals += [pvalsFTest[c][1]] 
  # sizes += [len(bus_fs_combo[bus_fs_combo.obs[&#39;cellRanger_louvain&#39;].isin([c])].obs_names)]

pvals = sm.multitest.fdrcorrection(pvals,alpha=0.05, method=&#39;indep&#39;)
pvals
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">(array([False, False, False, False, False, False, False, False, False,</span>
<span class="err">        False,  True, False, False, False, False, False, False, False,</span>
<span class="err">        False, False, False, False, False, False, False, False, False,</span>
<span class="err">        False, False, False, False, False, False, False, False, False]),</span>
<span class="err"> array([1.64253441e-01, 9.35522778e-01, 5.21386720e-01, 8.07361578e-01,</span>
<span class="err">        9.97826979e-01, 3.31498471e-01, 8.07361578e-01, 5.68508228e-01,</span>
<span class="err">        6.91908915e-01, 5.34369937e-01, 4.49408286e-04, 1.99877010e-01,</span>
<span class="err">        9.63535360e-01, 7.78106033e-01, 2.25187596e-01, 4.82880690e-01,</span>
<span class="err">        6.42595657e-01, 4.19543898e-01, 6.87698796e-01, 9.35522778e-01,</span>
<span class="err">        3.31498471e-01, 9.63535360e-01, 9.63535360e-01, 3.31498471e-01,</span>
<span class="err">        3.31498471e-01, 9.97826979e-01, 4.82880690e-01, 1.64253441e-01,</span>
<span class="err">        4.82880690e-01, 2.09527206e-01, 4.82880690e-01, 3.31498471e-01,</span>
<span class="err">        9.63535360e-01, 1.64253441e-01, 9.35522778e-01, 1.00000000e+00]))</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#counts = counts[counts[&#39;count&#39;]&gt; 0 ]
#https://stackoverflow.com/questions/36578458/how-does-one-insert-statistical-annotations-stars-or-p-values-into-matplotlib/37518947#37518947

counts[&#39;cluster&#39;] = [int(i) for i in counts[&#39;cluster&#39;]] 
plt.figure(figsize=(16, 6))
ax = sns.boxplot(x=&quot;cluster&quot;, y=&quot;count&quot;, hue=&quot;condition&quot;, data=counts,linewidth=0.5,palette=[&quot;darkorange&quot;, &quot;b&quot;])
ax.set(ylabel=&#39;Normalized Cell Count&#39;)
ax.set(xlabel=&#39;Louvain Cluster&#39;)

# statistical annotation
# alpha &lt; 0.05
for i in [11]: 
  w = 0.5
  x1, x2 = i-w,i+w  

  y, h, col = counts[&#39;count&#39;][counts[&#39;cluster&#39;] == i].max() + 2, 4, &#39;k&#39;
  plt.plot([x1, x1, x2, x2], [y, y+h, y+h, y], lw=1.5, c=col)
  plt.text((x1+x2)*.5, y+h, &quot;*&quot;, ha=&#39;center&#39;, va=&#39;bottom&#39;, color=col)


plt.show()
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../allDistanceCalculations_files/allDistanceCalculations_47_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>internalDist = list(withinFS_Dists[&#39;fedStarv&#39;])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>scipy.stats.spearmanr(internalDist,pvals[1])
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">SpearmanrResult(correlation=-0.050871795043468955, pvalue=0.768261688241116)</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>plt.scatter(internalDist,pvals[1])
plt.xlabel(&#39;Internal Distances&#39;)
plt.ylabel(&#39;P-value: Cell Count Difference&#39;)
plt.grid(None)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../allDistanceCalculations_files/allDistanceCalculations_50_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>counts.to_csv(&#39;individ_cellCounts.csv&#39;,index=None)
</code></pre></div>
</td></tr></table>
<h5 id="repeat-for-stimulation-data"><strong>Repeat for Stimulation Data</strong><a class="headerlink" href="#repeat-for-stimulation-data" title="Permanent link">&para;</a></h5>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>overlap_combo = anndata.read(&quot;D1.1823&quot;)
bus_combo_noZero.obs[&#39;cellRanger_louvain&#39;] = pd.Categorical(overlap_combo[overlap_combo.obs[&#39;cell_origin&#39;] == &#39;Stim&#39;].obs[&#39;knn_clus&#39;])
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>withinStim_Dists = getStimCentroidDists(bus_combo_noZero,60,&#39;cellRanger_louvain&#39;)
withinStim_Dists
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>swKCl</th>
      <th>swDI</th>
      <th>cluster</th>
      <th>clus_color</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>7.350214</td>
      <td>8.503921</td>
      <td>0</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>21.065245</td>
      <td>36.732548</td>
      <td>1</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>20.948408</td>
      <td>20.759415</td>
      <td>2</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>13.442171</td>
      <td>34.598927</td>
      <td>3</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>14.205286</td>
      <td>25.923903</td>
      <td>4</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>5</th>
      <td>22.041147</td>
      <td>31.476810</td>
      <td>5</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>6</th>
      <td>14.477929</td>
      <td>25.217993</td>
      <td>6</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>7</th>
      <td>11.867729</td>
      <td>17.921000</td>
      <td>7</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>8</th>
      <td>12.292484</td>
      <td>11.644725</td>
      <td>8</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>9</th>
      <td>22.288679</td>
      <td>22.918928</td>
      <td>9</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>10</th>
      <td>10.736828</td>
      <td>11.939169</td>
      <td>10</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>11</th>
      <td>9.275630</td>
      <td>8.065190</td>
      <td>11</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>12</th>
      <td>5.801665</td>
      <td>22.793827</td>
      <td>12</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>13</th>
      <td>31.335329</td>
      <td>31.087091</td>
      <td>13</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>14</th>
      <td>25.940889</td>
      <td>28.084023</td>
      <td>14</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>15</th>
      <td>10.881435</td>
      <td>18.185766</td>
      <td>15</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>16</th>
      <td>25.211300</td>
      <td>17.230129</td>
      <td>16</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>17</th>
      <td>22.916479</td>
      <td>12.188016</td>
      <td>17</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>18</th>
      <td>21.876759</td>
      <td>14.847662</td>
      <td>18</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>19</th>
      <td>23.442184</td>
      <td>24.017313</td>
      <td>19</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>20</th>
      <td>56.649334</td>
      <td>65.165184</td>
      <td>20</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>21</th>
      <td>24.023872</td>
      <td>16.950150</td>
      <td>21</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>22</th>
      <td>13.108215</td>
      <td>19.507177</td>
      <td>22</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>23</th>
      <td>6.248736</td>
      <td>12.076812</td>
      <td>23</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>24</th>
      <td>32.759808</td>
      <td>45.112389</td>
      <td>24</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>25</th>
      <td>11.500583</td>
      <td>18.344841</td>
      <td>25</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>26</th>
      <td>34.523087</td>
      <td>26.619263</td>
      <td>26</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>27</th>
      <td>25.315067</td>
      <td>31.565296</td>
      <td>27</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>28</th>
      <td>46.271626</td>
      <td>31.329525</td>
      <td>28</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>29</th>
      <td>13.501001</td>
      <td>10.425589</td>
      <td>29</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>30</th>
      <td>18.720173</td>
      <td>38.476334</td>
      <td>30</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>31</th>
      <td>38.197033</td>
      <td>57.892250</td>
      <td>31</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>32</th>
      <td>31.816713</td>
      <td>17.769464</td>
      <td>32</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>33</th>
      <td>40.618649</td>
      <td>44.847408</td>
      <td>33</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>34</th>
      <td>11.309300</td>
      <td>12.301980</td>
      <td>34</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>35</th>
      <td>16.585102</td>
      <td>11.396259</td>
      <td>35</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Compare to pairwise distances between cell atlas clusters
centroids = getClusCentroids(bus_combo_noZero,60,&#39;cellRanger_louvain&#39;)
#centroids_arr = centroids[&#39;centroid&#39;].to_numpy()
pairCentroid_dists = pairwise_distances(centroids, metric = &#39;l1&#39;)
pairCentroid_dists.shape
print(np.mean(pairCentroid_dists))
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">173.6400492134319</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#withinStim_Dists[&#39;swDI&#39;]
print(np.max(withinStim_Dists))
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">swKCl         56.649334</span>
<span class="err">swDI          65.165184</span>
<span class="err">cluster       35.000000</span>
<span class="err">clus_color          NaN</span>
<span class="c">dtype: float64</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>max(pairCentroid_dists.reshape((36*36,)))
reshape_noZero = [i for i in pairCentroid_dists.reshape((36*36,)) if i != 0]
print(np.min(reshape_noZero))

plt.figure(figsize=(16,10))
plt.hist(withinStim_Dists[&#39;swKCl&#39;],color=&#39;r&#39;,bins=15,alpha=0.6,label=&#39;SW-KCl&#39;,density=True)
plt.hist(withinStim_Dists[&#39;swDI&#39;],color=&#39;b&#39;,bins=15,alpha=0.6,label=&#39;SW-DI&#39;,density=True)

plt.hist(reshape_noZero, bins=36,color = &#39;orange&#39;,alpha=0.6,label=&#39;pairwise clusters&#39;,density=True)
plt.legend(loc=&#39;upper right&#39;)
plt.title(&#39;Comparison to pairwise distribution (pairwise on clusters with SW/KCl/DI cells)&#39;)

plt.xlabel(&quot;L1 Distance with 60 PC&#39;s&quot;)
plt.ylabel(&quot;Normalized Frequency&quot;)
plt.grid(b=None)
plt.show()
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">65.46425357810222</span>
</code></pre></div>

<p><img alt="png" src="../allDistanceCalculations_files/allDistanceCalculations_57_1.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>print(np.where(pairCentroid_dists &lt; 65.46))
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">(array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,</span>
<span class="err">       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,</span>
<span class="err">       34, 35]), array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,</span>
<span class="err">       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,</span>
<span class="err">       34, 35]))</span>
</code></pre></div>

<h4 id="organismal-distances-fedstarved-stimulation"><strong>Organismal Distances, Fed/Starved + Stimulation</strong><a class="headerlink" href="#organismal-distances-fedstarved-stimulation" title="Permanent link">&para;</a></h4>
<p>Batch effect plots</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#condition -- Stim, fed -- Fed/Starved data
overlap_combo = anndata.read(&quot;D1.1823&quot;)
overlap_combo
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">32594</span> <span class="err">×</span> <span class="mi">6756</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;starved&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_neighbor_score&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;annos&#39;</span><span class="p">,</span> <span class="s1">&#39;new_cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;annosSub&#39;</span><span class="p">,</span> <span class="s1">&#39;condition&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_origin&#39;</span><span class="p">,</span> <span class="s1">&#39;knn_clus&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts-0&#39;</span><span class="p">,</span> <span class="s1">&#39;mean-0&#39;</span><span class="p">,</span> <span class="s1">&#39;std-0&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts-1&#39;</span><span class="p">,</span> <span class="s1">&#39;mean-1&#39;</span><span class="p">,</span> <span class="s1">&#39;std-1&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;pca&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_nca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Use overlap_combo

#Go through each clustee
clus = np.unique(overlap_combo.obs[&#39;knn_clus&#39;])

fed_data = overlap_combo[overlap_combo.obs[&#39;fed&#39;] == &#39;True&#39;]
sw_data = overlap_combo[overlap_combo.obs[&#39;condition&#39;] == &#39;SW&#39;]

fed_orgs = np.unique(fed_data.obs[&#39;orgID&#39;])
sw_orgs = np.unique(sw_data.obs[&#39;orgID&#39;])

pcs = 60

org_metadata = pd.DataFrame(columns =[&#39;orgID&#39;,&#39;cluster&#39;,&#39;condition&#39;]) #centroid = row mean for all cells in organism
num_points = (len(fed_orgs)+len(sw_orgs))*len(clus)

#Get centroid in PCA space in each cluster for each organism
org_centroids = np.zeros((num_points,pcs))

condition = []
cluster = []
orgID = []

count = 0
toRemove = []
for c in clus:
  #Get all cells in chosen condition
  fed_sub = fed_data[fed_data.obs[&#39;knn_clus&#39;] == c]
  fed_pca_embed = fed_sub.obsm[&#39;X_pca&#39;][:,0:pcs] #PCA matrix

  sw_sub = sw_data[sw_data.obs[&#39;knn_clus&#39;] == c]
  sw_pca_embed = sw_sub.obsm[&#39;X_pca&#39;][:,0:pcs] #PCA matrix

  #For each org, get centroid of cells (X_pca)
  for sw in sw_orgs:
    condition += [&#39;SW&#39;]
    cluster += [c]
    orgID += [sw]

    #Get location of org cells
    org_pos = list(sw_sub.obs[&#39;orgID&#39;] == sw)

    #Sum down rows of cells for each PC
    if len(sw_pca_embed[org_pos,]) == 0:
      toRemove += [count]

    org_centroids[count,:] = list(sw_pca_embed[org_pos,].mean(axis=0))

    count += 1

  for fed in fed_orgs:
    condition += [&#39;Fed&#39;]
    cluster += [c]
    orgID += [fed]

    #Get location of org cells
    org_pos = list(fed_sub.obs[&#39;orgID&#39;] == fed)
    #Sum down rows of cells for each PC
    if len(fed_pca_embed[org_pos,]) == 0:
      toRemove += [count]

    org_centroids[count,:] = list(fed_pca_embed[org_pos,].mean(axis=0))

    count += 1    

  #For each org, get centroid of cells (X_pca)

  #Average pairwise distance between org centroid

org_metadata[&#39;orgID&#39;] = orgID
org_metadata[&#39;cluster&#39;] = cluster
org_metadata[&#39;condition&#39;] = condition
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>toRemove 
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">[]</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>org_metadata = org_metadata.drop(org_metadata.index[toRemove])
org_metadata.head()

org_centroids= np.delete(org_centroids, toRemove, 0)
org_centroids[0,:]
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">array([-2.25471210e+00, -1.51177585e+00, -1.93993020e+00, -9.54584032e-02,</span>
<span class="err">       -2.38718843e+00, -1.51001954e+00, -2.81468654e+00,  1.05100930e+00,</span>
<span class="err">        1.54143572e+00, -7.58547902e-01, -1.50774598e+00, -2.47600722e+00,</span>
<span class="err">        7.53088653e-01, -9.25398111e-01,  5.30199647e-01, -2.36053783e-02,</span>
<span class="err">        9.48955193e-02, -6.50646329e-01,  1.05686748e+00, -2.67228413e+00,</span>
<span class="err">        2.98014688e+00,  5.93579225e-02,  5.69488704e-01, -1.05491745e+00,</span>
<span class="err">        1.12574625e+00, -1.49573892e-01,  5.68022847e-01, -1.29234409e+00,</span>
<span class="err">        1.40452191e-01, -3.84621084e-01,  2.95532733e-01,  1.13393039e-01,</span>
<span class="err">        2.88706899e-01, -5.50938368e-01, -1.17623888e-01, -1.51531205e-01,</span>
<span class="err">       -7.78403163e-01, -7.72583902e-01,  9.31857646e-01,  9.21139657e-01,</span>
<span class="err">       -4.21358228e-01,  9.92068797e-02, -3.57610792e-01,  1.40223965e-01,</span>
<span class="err">        4.30851281e-02,  2.75345147e-01,  5.04794896e-01, -1.59878790e-01,</span>
<span class="err">        1.80910379e-01,  2.61838585e-01, -2.14510873e-01,  1.25861779e-01,</span>
<span class="err">       -3.76004130e-01,  2.04274207e-01, -1.61955715e-03, -9.25957318e-03,</span>
<span class="err">        5.98372109e-02, -2.70651639e-01,  2.62535036e-01,  1.59801412e-02])</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>org_pairDists = pd.DataFrame(columns =[&#39;cluster&#39;,&#39;pair_fed&#39;,&#39;pair_sw&#39;,&#39;pair_mix&#39;])
cluster = []
pair_fed = []
pair_sw = []
pair_mix = []

for c in clus:

  #Fed org centroids
  where_fed = (org_metadata.cluster == c) &amp; (org_metadata.condition == &#39;Fed&#39;)
  fed_mat = org_centroids[where_fed,:]

  pair_fed_dist = pairwise_distances(fed_mat, metric = &#39;l1&#39;)
  pair_fed_mean = np.mean(pair_fed_dist)

  #SW org centroids
  where_sw = (org_metadata.cluster == c) &amp; (org_metadata.condition == &#39;SW&#39;)
  sw_mat = org_centroids[where_sw,:]

  pair_sw_dist = pairwise_distances(sw_mat, metric = &#39;l1&#39;)
  pair_sw_mean = np.mean(pair_sw_dist)

  pair_mix_dist = pairwise_distances(fed_mat,sw_mat, metric = &#39;l1&#39;)
  pair_mix_mean = np.mean(pair_mix_dist)

  cluster += [c]
  pair_fed += [pair_fed_mean]
  pair_sw += [pair_sw_mean]
  pair_mix += [pair_mix_mean]

org_pairDists[&#39;cluster&#39;] = cluster
org_pairDists[&#39;pair_fed&#39;] = pair_fed
org_pairDists[&#39;pair_sw&#39;] = pair_sw
org_pairDists[&#39;pair_mix&#39;] = pair_mix
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>org_pairDists.head()
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cluster</th>
      <th>pair_fed</th>
      <th>pair_sw</th>
      <th>pair_mix</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>8.005131</td>
      <td>7.995117</td>
      <td>23.570756</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>15.629137</td>
      <td>32.140367</td>
      <td>75.364971</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>9.598802</td>
      <td>14.187616</td>
      <td>49.332659</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>19.421792</td>
      <td>22.127487</td>
      <td>42.742088</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>16.030268</td>
      <td>26.287836</td>
      <td>36.652727</td>
    </tr>
  </tbody>
</table>
</div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>plt.figure(figsize=(16,10))
plt.hist(org_pairDists[&#39;pair_fed&#39;],color=&#39;r&#39;,bins=15,alpha=0.6,label=&#39;control org dist&#39;,density=True)
plt.hist(org_pairDists[&#39;pair_sw&#39;],color=&#39;blue&#39;,bins=15,alpha=0.6,label=&#39;sw org dist&#39;,density=True)
plt.hist(org_pairDists[&#39;pair_mix&#39;], bins=15,color = &#39;orange&#39;,alpha=0.6,label=&#39;mix dist&#39;,density=True)
plt.legend(loc=&#39;upper right&#39;)
plt.title(&#39;Inter-Organism Distances (L1)&#39;)
plt.grid(b=None)
plt.xlabel(&quot;Average L1 Distance with 60 PC&#39;s&quot;)
plt.ylabel(&quot;Normalized Frequency&quot;)

plt.show()
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../allDistanceCalculations_files/allDistanceCalculations_66_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>overlap_combo
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">32594</span> <span class="err">×</span> <span class="mi">6756</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;starved&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_neighbor_score&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;annos&#39;</span><span class="p">,</span> <span class="s1">&#39;new_cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;annosSub&#39;</span><span class="p">,</span> <span class="s1">&#39;condition&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_origin&#39;</span><span class="p">,</span> <span class="s1">&#39;knn_clus&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts-0&#39;</span><span class="p">,</span> <span class="s1">&#39;mean-0&#39;</span><span class="p">,</span> <span class="s1">&#39;std-0&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts-1&#39;</span><span class="p">,</span> <span class="s1">&#39;mean-1&#39;</span><span class="p">,</span> <span class="s1">&#39;std-1&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;pca&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_nca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>overlap_combo.write(&#39;overlap_combo.h5ad&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">... storing &#39;louvain&#39; as categorical</span>
<span class="err">... storing &#39;leiden&#39; as categorical</span>
<span class="err">... storing &#39;orgID&#39; as categorical</span>
<span class="err">... storing &#39;fed&#39; as categorical</span>
<span class="err">... storing &#39;starved&#39; as categorical</span>
<span class="err">... storing &#39;cellRanger_louvain&#39; as categorical</span>
<span class="err">... storing &#39;annos&#39; as categorical</span>
<span class="err">... storing &#39;new_cellRanger_louvain&#39; as categorical</span>
<span class="err">... storing &#39;annosSub&#39; as categorical</span>
<span class="err">... storing &#39;condition&#39; as categorical</span>
</code></pre></div>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../deSeq2Analysis_StimulationResponse/" class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Extraction of Differentially Expressed Genes Under Starvation
              </div>
            </div>
          </a>
        
        
          <a href="../MARIMBAAnnosAnalysis/" class="md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Generation Cell Atlas with MARIMBA Transcriptome
              </div>
            </div>
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/vendor.53cc9318.min.js"></script>
      <script src="../../assets/javascripts/bundle.e9c9f54f.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "../..",
          features: ['search.suggest', 'search.highlight'],
          search: Object.assign({
            worker: "../../assets/javascripts/worker/search.9c0e82ba.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>