
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Walkthrough for Whole Organisms Mutliplexed Data Analysis">
      
      
      
        <meta name="author" content="Tara Chari">
      
      
        <link rel="canonical" href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/notebooks/neuronSubpop_Analysis/">
      
      <link rel="shortcut icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.2.7">
    
    
      
        <title>Analysis of Neuron Subpopulations - WHAMSeqAnalysis</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.cb6bc1d0.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.39b8e14a.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#download-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/" title="WHAMSeqAnalysis" class="md-header-nav__button md-logo" aria-label="WHAMSeqAnalysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      <div class="md-header-nav__ellipsis">
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            WHAMSeqAnalysis
          </span>
        </div>
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            
              Analysis of Neuron Subpopulations
            
          </span>
        </div>
      </div>
    </div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    




<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/" title="WHAMSeqAnalysis" class="md-nav__button md-logo" aria-label="WHAMSeqAnalysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    WHAMSeqAnalysis
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" >
      
      <label class="md-nav__link" for="nav-2">
        ClickTag Pre-processing
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="ClickTag Pre-processing" data-md-level="1">
        <label class="md-nav__title" for="nav-2">
          <span class="md-nav__icon md-icon"></span>
          ClickTag Pre-processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellRangerClickTagCounts/" class="md-nav__link">
        Initial Quantification of ClickTag Counts with Cell Ranger (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistobusStarvClickTags/" class="md-nav__link">
        Quantification of ClickTag Counts with kallisto-bustools (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistobusStimClickTags/" class="md-nav__link">
        Quantification of ClickTag Counts with kallisto-bustools (Stimulation Data)
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3" >
      
      <label class="md-nav__link" for="nav-3">
        cDNA Pre-processing
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="cDNA Pre-processing" data-md-level="1">
        <label class="md-nav__title" for="nav-3">
          <span class="md-nav__icon md-icon"></span>
          cDNA Pre-processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../filterStarvCells_ClickTags/" class="md-nav__link">
        Initial cDNA Filtering with Cell Ranger Data (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistoBusRuns_StarvAndStim/" class="md-nav__link">
        Quantification of cDNA Counts with kallisto-bustools (All Data)
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4" checked>
      
      <label class="md-nav__link" for="nav-4">
        Cell Atlas and Perturbation Response
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Cell Atlas and Perturbation Response" data-md-level="1">
        <label class="md-nav__title" for="nav-4">
          <span class="md-nav__icon md-icon"></span>
          Cell Atlas and Perturbation Response
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellRangerClustering_Starvation/" class="md-nav__link">
        Initial Cell Atlas Clustering from Cell Ranger
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../starvation_Analysis/" class="md-nav__link">
        Cell Atlas Clustering and Analysis from kallisto-bustools
      </a>
    </li>
  

          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Analysis of Neuron Subpopulations
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Analysis of Neuron Subpopulations
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import Packages
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-in-previously-saved-data" class="md-nav__link">
    Read in Previously Saved Data
  </a>
  
    <nav class="md-nav" aria-label="Read in Previously Saved Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#generating-neuron-subpopulations" class="md-nav__link">
    Generating Neuron Subpopulations
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#neuron-subpopulation-plots" class="md-nav__link">
    Neuron Subpopulation Plots
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../pseudotime_Analysis/" class="md-nav__link">
        Pseudotime Analysis
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../deSeq2Analysis_StarvationResponse/" class="md-nav__link">
        Extraction of Differentially Expressed Genes Under Starvation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5" >
      
      <label class="md-nav__link" for="nav-5">
        Stimulation Data Analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Stimulation Data Analysis" data-md-level="1">
        <label class="md-nav__title" for="nav-5">
          <span class="md-nav__icon md-icon"></span>
          Stimulation Data Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../stimulation_Analysis/" class="md-nav__link">
        Cell Atlas Concordance and Perturbation Response
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../deSeq2Analysis_StimulationResponse/" class="md-nav__link">
        Extraction of Differentially Expressed Genes Under Starvation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6" >
      
      <label class="md-nav__link" for="nav-6">
        Inter-Experimental Analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Inter-Experimental Analysis" data-md-level="1">
        <label class="md-nav__title" for="nav-6">
          <span class="md-nav__icon md-icon"></span>
          Inter-Experimental Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../allDistanceCalculations/" class="md-nav__link">
        Comparative Analysis of Cell Atlas Concordance and Perturbation Responses
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-7" type="checkbox" id="nav-7" >
      
      <label class="md-nav__link" for="nav-7">
        Explore Mutliplexed Datasets
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Explore Mutliplexed Datasets" data-md-level="1">
        <label class="md-nav__title" for="nav-7">
          <span class="md-nav__icon md-icon"></span>
          Explore Mutliplexed Datasets
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../MARIMBAAnnosAnalysis/" class="md-nav__link">
        Generation Cell Atlas with MARIMBA Transcriptome
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../MARIMBAAtlasSearchAndPlot/" class="md-nav__link">
        Explore MARIMBA-Annotated Cell Atlas
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellAtlasSearchAndPlot/" class="md-nav__link">
        Explore Original Cell Atlas
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import Packages
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#read-in-previously-saved-data" class="md-nav__link">
    Read in Previously Saved Data
  </a>
  
    <nav class="md-nav" aria-label="Read in Previously Saved Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#generating-neuron-subpopulations" class="md-nav__link">
    Generating Neuron Subpopulations
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#neuron-subpopulation-plots" class="md-nav__link">
    Neuron Subpopulation Plots
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/edit/master/docs/notebooks/neuronSubpop_Analysis.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                  <h1>Analysis of Neuron Subpopulations</h1>
                
                <p><a href="https://colab.research.google.com/github/pachterlab/CWGFLHGCCHAP_2021/blob/master/notebooks/CellAtlasAnalysis/neuronSubpop_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!date
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">Fri Aug 21 00:45:00 UTC 2020</span>
</code></pre></div>

<h3 id="download-data"><strong>Download Data</strong><a class="headerlink" href="#download-data" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import requests
from tqdm import tnrange, tqdm_notebook
def download_file(doi,ext):
    url = &#39;https://api.datacite.org/dois/&#39;+doi+&#39;/media&#39;
    r = requests.get(url).json()
    netcdf_url = r[&#39;data&#39;][0][&#39;attributes&#39;][&#39;url&#39;]
    r = requests.get(netcdf_url,stream=True)
    #Set file name
    fname = doi.split(&#39;/&#39;)[-1]+ext
    #Download file with progress bar
    if r.status_code == 403:
        print(&quot;File Unavailable&quot;)
    if &#39;content-length&#39; not in r.headers:
        print(&quot;Did not get file&quot;)
    else:
        with open(fname, &#39;wb&#39;) as f:
            total_length = int(r.headers.get(&#39;content-length&#39;))
            pbar = tnrange(int(total_length/1024), unit=&quot;B&quot;)
            for chunk in r.iter_content(chunk_size=1024):
                if chunk:
                    pbar.update()
                    f.write(chunk)
        return fname
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Kallisto bus clustered starvation data, h5ad
download_file(&#39;10.22002/D1.1796&#39;,&#39;.gz&#39;)

#Starvation h5ad data, all nonzero genes included, filtered for &#39;real cells&#39; from de-multiplexing
download_file(&#39;10.22002/D1.1797&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">479630</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">26058</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1797.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Human ortholog annotations
download_file(&#39;10.22002/D1.1819&#39;,&#39;.gz&#39;)

#Panther annotations
download_file(&#39;10.22002/D1.1820&#39;,&#39;.gz&#39;)

#GO Terms
download_file(&#39;10.22002/D1.1822&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">528</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>






<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">515</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">227</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1822.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Previously saved neuron subpopulations
download_file(&#39;10.22002/D1.1804&#39;,&#39;.gz&#39;)

#Previously saved marker genes for neurons
download_file(&#39;10.22002/D1.1823&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">1595</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">257856</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1823.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!gunzip *.gz
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!pip install --quiet anndata
!pip install --quiet scanpy

!pip install --quiet louvain
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 122kB 6.0MB/s </span>
<span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 10.2MB 3.8MB/s </span>
<span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 71kB 7.6MB/s </span>
<span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 51kB 5.7MB/s </span>
<span class="err">[?25h  Building wheel for sinfo (setup.py) ... [?25l[?25hdone</span>
<span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 2.2MB 4.3MB/s </span>
<span class="err">[K     |â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3.2MB 24.3MB/s </span>
<span class="err">[?25h</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!pip3 install --quiet rpy2
</code></pre></div>
</td></tr></table>
<h3 id="import-packages"><strong>Import Packages</strong><a class="headerlink" href="#import-packages" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Install Packages
import random
import pandas as pd
import anndata
import scanpy as sc
import numpy as np
import scipy.sparse

import warnings
warnings.filterwarnings(&#39;ignore&#39;)

from sklearn.neighbors import (KNeighborsClassifier,NeighborhoodComponentsAnalysis)
from sklearn.pipeline import Pipeline
from sklearn.manifold import TSNE

#import scrublet as scr

import matplotlib.pyplot as plt
%matplotlib inline
sc.set_figure_params(dpi=125)

import seaborn as sns
sns.set(style=&quot;whitegrid&quot;)
%load_ext rpy2.ipython
</code></pre></div>
</td></tr></table>
<h2 id="read-in-previously-saved-data"><strong>Read in Previously Saved Data</strong><a class="headerlink" href="#read-in-previously-saved-data" title="Permanent link">&para;</a></h2>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Read in h5ad file
bus_fs_combo = anndata.read(&quot;D1.1796&quot;)

#Previously saved neurons
neurons = anndata.read(&quot;D1.1804&quot;)

#Read in annotations
from io import StringIO

hg_ortho_df = pd.read_csv(StringIO(&#39;&#39;.join(l.replace(&#39;|&#39;, &#39;\t&#39;) for l in open(&#39;D1.1819&#39;))),
            sep=&quot;\t&quot;,header=None,skiprows=[0,1,2,3])

hg_ortho_df[[&#39;XLOC&#39;,&#39;TCONS&#39;]] = hg_ortho_df[13].str.split(expand=True) 
hg_ortho_df[[&#39;Gene&#39;,&#39;gi&#39;]] = hg_ortho_df[3].str.split(expand=True) 
hg_ortho_df[&#39;Description&#39;]= hg_ortho_df[11]


panther_df = pd.read_csv(&#39;D1.1820&#39;,
            sep=&quot;\t&quot;,header=None) #skiprows=[0,1,2,3]



goTerm_df = pd.read_csv(&#39;D1.1822&#39;,
            sep=&quot; &quot;,header=None) #skiprows=[0,1,2,3]

bus_fs_combo
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">13673</span> <span class="err">Ã—</span> <span class="mi">8696</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;starved&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_neighbor_score&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;annos&#39;</span><span class="p">,</span> <span class="s1">&#39;new_cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;annosSub&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;annosSub_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;annos_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain_sizes&#39;</span><span class="p">,</span> <span class="s2">&quot;dendrogram_[&#39;new_cellRanger_louvain&#39;]&quot;</span><span class="p">,</span> <span class="s1">&#39;dendrogram_new_cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_neighbor_score_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;new_cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;paga&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_nca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
    <span class="n">obsp</span><span class="p">:</span> <span class="s1">&#39;connectivities&#39;</span><span class="p">,</span> <span class="s1">&#39;distances&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>raw_fs_combo = anndata.read(&#39;D1.1797&#39;)

raw_fs_combo = raw_fs_combo[bus_fs_combo.obs_names,]
#Transfer info from embedded version
raw_fs_combo.obs[&#39;cellRanger_louvain&#39;] = pd.Categorical(bus_fs_combo.obs[&#39;cellRanger_louvain&#39;])
raw_fs_combo.obs[&#39;fed&#39;] = pd.Categorical(bus_fs_combo.obs[&#39;fed&#39;])
raw_fs_combo.obsm[&#39;X_tsne&#39;] = bus_fs_combo.obsm[&#39;X_tsne&#39;]

raw_fs_combo
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Trying</span> <span class="n">to</span> <span class="n">set</span> <span class="n">attribute</span> <span class="err">`</span><span class="o">.</span><span class="n">obs</span><span class="err">`</span> <span class="n">of</span> <span class="n">view</span><span class="p">,</span> <span class="n">copying</span><span class="o">.</span>





<span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">13673</span> <span class="err">Ã—</span> <span class="mi">46716</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_tsne&#39;</span>
</code></pre></div>

<h3 id="generating-neuron-subpopulations"><strong>Generating Neuron Subpopulations</strong><a class="headerlink" href="#generating-neuron-subpopulations" title="Permanent link">&para;</a></h3>
<p>We begin by restricting analysis to four clusters that contain neuron subpopulations (31,26,6 and 9). After constraining the single-cell count matrices to the relevant cells, we examine the variance explained in the top principal components, and based on the "elbow plot" continue analysis with 15 principal components. </p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># #Neurons, start from raw counts + unfiltered genes
neurons = raw_fs_combo[raw_fs_combo.obs[&#39;cellRanger_louvain&#39;].isin([31,26,6,9])]
sc.pp.filter_cells(neurons, min_counts=0)
sc.pp.filter_genes(neurons, min_counts=0)

sc.pp.normalize_per_cell(neurons, counts_per_cell_after=1e4)

sc.pp.log1p(neurons)

sc.pp.highly_variable_genes(neurons,n_top_genes=2000,n_bins=50)

neurons = neurons[:,neurons.var[&#39;highly_variable&#39;]]

sc.pp.scale(neurons, max_value=10)

sc.tl.pca(neurons, n_comps=60)
sc.pl.pca_variance_ratio(neurons, log=True)

sc.pp.neighbors(neurons,n_neighbors=15, n_pcs=15) #n_neighbors=5, n_pcs=15,use_rep=&#39;X_nca&#39;
sc.tl.louvain(neurons,resolution=1,key_added=&#39;louvain_neur&#39;,random_state=42)#Clustering algorithm,resolution=0.5
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">Trying to set attribute `.obs` of view, copying.</span>
</code></pre></div>

<p><img alt="png" src="../neuronSubpop_Analysis_files/neuronSubpop_Analysis_16_1.png" /></p>
<p>Next, we perform Louvain clustering to identify distinct neuron subpopulations, and we plot them using a UMAP reduction of the count matrix. The plot on the left shows the 15 Louvain derived clusters; the plot on the right shows the same UMAP representation of the cells but colored according to the clusters obtained from clustering with all the cells. This shows that a reclustering on the restricted cell x gene matrix produces a refinement of the original clusters.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.louvain(neurons,resolution=2.5,key_added=&#39;louvain_neur&#39;,random_state=42)#Clustering algorithm,resolution=0.5
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>neurons
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">1387</span> <span class="err">Ã—</span> <span class="mi">2000</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_neur&#39;</span><span class="p">,</span> <span class="s1">&#39;test_louvain&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;highly_variable&#39;</span><span class="p">,</span> <span class="s1">&#39;means&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions_norm&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;hvg&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_neur_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">,</span> <span class="s1">&#39;test_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
    <span class="n">obsp</span><span class="p">:</span> <span class="s1">&#39;connectivities&#39;</span><span class="p">,</span> <span class="s1">&#39;distances&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.tsne(neurons, n_pcs=15,perplexity=25,random_state = 42) #learning_rate=250
sc.tl.umap(neurons,random_state=42,spread=2.5, min_dist=1)

neurons.obs[&#39;cellAtlasClusters&#39;] = pd.Categorical(neurons.obs[&#39;cellRanger_louvain&#39;] )
sc.pl.umap(neurons, color=[&#39;louvain_neur&#39;,&#39;cellAtlasClusters&#39;],color_map=&#39;viridis&#39;,size=50,legend_loc=&#39;on data&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../neuronSubpop_Analysis_files/neuronSubpop_Analysis_20_0.png" /></p>
<p>The plots below show the cells expressing several different neuropeptides:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.umap(neurons, color=[&#39;XLOC_040584&#39;,&#39;XLOC_019434&#39;,&#39;TRINITY_DN20104_c0_g1_i1.mrna1&#39;,&#39;XLOC_017097&#39;,
                           &#39;XLOC_004021&#39;],color_map=&#39;viridis&#39;,size=50,legend_loc=&#39;on data&#39;,
           title=[&#39;GRWGamide&#39;,&#39;RFamide&#39;,&#39;PRPamide&#39;,&#39;GLWamide2&#39;,
                  &#39;FLFamide&#39;],vmin=0,vmax=5)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../neuronSubpop_Analysis_files/neuronSubpop_Analysis_22_0.png" /></p>
<p>Next, we identify marker genes for each of the 15 Louvain clusters determined for the neuron subpopulation, and we annotate them using the Panther database, gene ontology (GO) terms and via orthology searches.</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Get top n marker genes for each cluster

#Keep top 100 genes, &#39;louvain_neur&#39; is label for neuron clusters determined using Louvain clustering algorithm
sc.tl.rank_genes_groups(neurons, &#39;louvain_neur&#39;,n_genes = 100,method=&#39;wilcoxon&#39;) #Using non-parametric test for significance
neurons
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">1387</span> <span class="err">Ã—</span> <span class="mi">2000</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_neur&#39;</span><span class="p">,</span> <span class="s1">&#39;test_louvain&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;highly_variable&#39;</span><span class="p">,</span> <span class="s1">&#39;means&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions_norm&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;hvg&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_neur_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">,</span> <span class="s1">&#39;test_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
    <span class="n">obsp</span><span class="p">:</span> <span class="s1">&#39;connectivities&#39;</span><span class="p">,</span> <span class="s1">&#39;distances&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Make dataframe, with 100 marker genes for each cluster + annotations
neuron_clusters = np.unique(neurons.obs[&#39;louvain_neur&#39;])
neuron_markers = pd.DataFrame()

neurClus = []
markerGene = []
padj = []
orthoGene = []
orthoDescr = []

pantherNum = []
pantherDescr = []

goTerms = []

for i in neuron_clusters:
  genes = neurons.uns[&#39;rank_genes_groups&#39;][&#39;names&#39;][str(i)]

  neurClus += list(np.repeat(i,len(genes)))
  markerGene += list(genes)
  padj += list(neurons.uns[&#39;rank_genes_groups&#39;][&#39;pvals_adj&#39;][str(i)])

  for g in genes:

    sub_df = hg_ortho_df[hg_ortho_df.XLOC.isin([g])]
    panth_df = panther_df[panther_df[0].isin([g])]
    go_df = goTerm_df[goTerm_df[0].isin([g])]

    if len(sub_df) &gt; 0:
      #Save first result for gene/description
      orthoGene += [list(sub_df.Gene)[0]]
      orthoDescr += [list(sub_df.Description)[0]]
    else:
      orthoGene += [&#39;NA&#39;]
      orthoDescr += [&#39;NA&#39;]


    if len(panth_df) &gt; 0:
      pantherNum += [list(panth_df[1])]
      pantherDescr += [list(panth_df[2])]
    else:
      pantherNum += [&#39;NA&#39;]
      pantherDescr += [&#39;NA&#39;]


    if len(go_df) &gt; 0:
      goTerms += [list(go_df[1])]
    else:
      goTerms += [&#39;NA&#39;]


neuron_markers[&#39;neurClus&#39;] = neurClus
neuron_markers[&#39;markerGene&#39;] = markerGene
neuron_markers[&#39;padj&#39;] = padj

neuron_markers[&#39;orthoGene&#39;] = orthoGene
neuron_markers[&#39;orthoDescr&#39;] = orthoDescr

neuron_markers[&#39;pantherID&#39;] = pantherNum
neuron_markers[&#39;pantherDescr&#39;] = pantherDescr

neuron_markers[&#39;goTerms&#39;] = goTerms

neuron_markers.head()
#list(neurons.uns[&#39;rank_genes_groups&#39;][&#39;names&#39;][&#39;1&#39;])
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>neurClus</th>
      <th>markerGene</th>
      <th>padj</th>
      <th>orthoGene</th>
      <th>orthoDescr</th>
      <th>pantherID</th>
      <th>pantherDescr</th>
      <th>goTerms</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>XLOC_011074</td>
      <td>9.930711e-30</td>
      <td>TUBA1C</td>
      <td>tubulin alpha-1C chain isoform a [Homo sapiens]</td>
      <td>[PTHR11588:SF223]</td>
      <td>[SUBFAMILY NOT NAMED]</td>
      <td>[GO:0045298,GO:0005856,GO:0005200,GO:0005622,G...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0</td>
      <td>XLOC_015355</td>
      <td>2.959498e-09</td>
      <td>NA</td>
      <td>NA</td>
      <td>[PTHR12124:SF53]</td>
      <td>[EXONUCLEASE 3'-5' DOMAIN-CONTAINING PROTEIN 1]</td>
      <td>[GO:0016070,GO:0016787,GO:0016072,GO:0044238,G...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0</td>
      <td>XLOC_044603</td>
      <td>1.123759e-05</td>
      <td>NDUFV2</td>
      <td>NADH dehydrogenase [ubiquinone] flavoprotein ...</td>
      <td>[PTHR10371]</td>
      <td>[NADH DEHYDROGENASE  UBIQUINONE  FLAVOPROTEIN ...</td>
      <td>[GO:0043234,GO:0005886,GO:0016020,GO:0032991,G...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0</td>
      <td>XLOC_030920</td>
      <td>3.931534e-05</td>
      <td>NA</td>
      <td>NA</td>
      <td>[PTHR19290]</td>
      <td>[BASIC HELIX-LOOP-HELIX PROTEIN NEUROGENIN-REL...</td>
      <td>[nan]</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0</td>
      <td>XLOC_004445</td>
      <td>3.710752e-01</td>
      <td>NA</td>
      <td>NA</td>
      <td>[PTHR28618]</td>
      <td>[FAMILY NOT NAMED]</td>
      <td>[nan]</td>
    </tr>
  </tbody>
</table>
</div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Write to csv
neuron_markers.to_csv(&#39;neuron_marker_annotations.csv&#39;)

#Read in csv (previously saved version, uploaded to Box)
neuron_markers = pd.read_csv(&#39;neuron_marker_annotations.csv&#39;,
            sep=&quot;,&quot;)
neuron_markers.head()
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Unnamed: 0</th>
      <th>neurClus</th>
      <th>markerGene</th>
      <th>padj</th>
      <th>orthoGene</th>
      <th>orthoDescr</th>
      <th>pantherID</th>
      <th>pantherDescr</th>
      <th>goTerms</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>0</td>
      <td>XLOC_011074</td>
      <td>9.930711e-30</td>
      <td>TUBA1C</td>
      <td>tubulin alpha-1C chain isoform a [Homo sapiens]</td>
      <td>['PTHR11588:SF223']</td>
      <td>['SUBFAMILY NOT NAMED']</td>
      <td>['GO:0045298,GO:0005856,GO:0005200,GO:0005622,...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>0</td>
      <td>XLOC_015355</td>
      <td>2.959498e-09</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>['PTHR12124:SF53']</td>
      <td>["EXONUCLEASE 3'-5' DOMAIN-CONTAINING PROTEIN 1"]</td>
      <td>['GO:0016070,GO:0016787,GO:0016072,GO:0044238,...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>0</td>
      <td>XLOC_044603</td>
      <td>1.123759e-05</td>
      <td>NDUFV2</td>
      <td>NADH dehydrogenase [ubiquinone] flavoprotein ...</td>
      <td>['PTHR10371']</td>
      <td>['NADH DEHYDROGENASE  UBIQUINONE  FLAVOPROTEIN...</td>
      <td>['GO:0043234,GO:0005886,GO:0016020,GO:0032991,...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>0</td>
      <td>XLOC_030920</td>
      <td>3.931534e-05</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>['PTHR19290']</td>
      <td>['BASIC HELIX-LOOP-HELIX PROTEIN NEUROGENIN-RE...</td>
      <td>[nan]</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>0</td>
      <td>XLOC_004445</td>
      <td>3.710752e-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>['PTHR28618']</td>
      <td>['FAMILY NOT NAMED']</td>
      <td>[nan]</td>
    </tr>
  </tbody>
</table>
</div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>neurons.write(&#39;neuron_subpops_fs.h5ad&#39;)
</code></pre></div>
</td></tr></table>
<h3 id="neuron-subpopulation-plots"><strong>Neuron Subpopulation Plots</strong><a class="headerlink" href="#neuron-subpopulation-plots" title="Permanent link">&para;</a></h3>
<p>Hierarchical Clustering of Neuron Subpopulations</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Previously saved neurons
neurons = anndata.read(&quot;D1.1804&quot;)
print(neurons)

raw_fs_combo = anndata.read(&#39;D1.1797&#39;)

raw_fs_combo = raw_fs_combo[bus_fs_combo.obs_names,]
#Transfer info from embedded version
raw_fs_combo.obs[&#39;cellRanger_louvain&#39;] = pd.Categorical(bus_fs_combo.obs[&#39;cellRanger_louvain&#39;])
raw_fs_combo.obs[&#39;fed&#39;] = pd.Categorical(bus_fs_combo.obs[&#39;fed&#39;])
raw_fs_combo.obsm[&#39;X_tsne&#39;] = bus_fs_combo.obsm[&#39;X_tsne&#39;]

raw_fs_combo
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Trying</span> <span class="n">to</span> <span class="n">set</span> <span class="n">attribute</span> <span class="err">`</span><span class="o">.</span><span class="n">obs</span><span class="err">`</span> <span class="n">of</span> <span class="n">view</span><span class="p">,</span> <span class="n">copying</span><span class="o">.</span>


<span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">1387</span> <span class="err">Ã—</span> <span class="mi">2000</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_neur&#39;</span><span class="p">,</span> <span class="s1">&#39;test_louvain&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;highly_variable&#39;</span><span class="p">,</span> <span class="s1">&#39;means&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions_norm&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;hvg&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_neur_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">,</span> <span class="s1">&#39;test_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
    <span class="n">obsp</span><span class="p">:</span> <span class="s1">&#39;connectivities&#39;</span><span class="p">,</span> <span class="s1">&#39;distances&#39;</span>





<span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">13673</span> <span class="err">Ã—</span> <span class="mi">46716</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_tsne&#39;</span>
</code></pre></div>

<p>To examine similarities between different subpopulations we infer a dendogram based on similarities of cells in the different Louvain derived clusters:</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.dendrogram(neurons,&#39;louvain_neur&#39;,linkage_method=&#39;ward&#39;)
neurons
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">Ã—</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">1387</span> <span class="err">Ã—</span> <span class="mi">2000</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_neur&#39;</span><span class="p">,</span> <span class="s1">&#39;test_louvain&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;highly_variable&#39;</span><span class="p">,</span> <span class="s1">&#39;means&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions_norm&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;hvg&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_neur_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">,</span> <span class="s1">&#39;test_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span><span class="p">,</span> <span class="s2">&quot;dendrogram_[&#39;louvain_neur&#39;]&quot;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
    <span class="n">obsp</span><span class="p">:</span> <span class="s1">&#39;connectivities&#39;</span><span class="p">,</span> <span class="s1">&#39;distances&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>neurons.uns[&#39;dendrogram_louvain_neur&#39;] = neurons.uns[&quot;dendrogram_[&#39;louvain_neur&#39;]&quot;]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.dendrogram(neurons,&#39;louvain_neur&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../neuronSubpop_Analysis_files/neuronSubpop_Analysis_34_0.png" /></p>
<div class="codehilite"><pre><span></span><code><span class="err">&lt;matplotlib.axes._subplots.AxesSubplot at 0x7f984bb074e0&gt;</span>
</code></pre></div>

<p>Plot marker genes</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>neuron_markers = pd.read_csv(&#39;D1.1823&#39;) 
neuron_markers.head()
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Unnamed: 0</th>
      <th>neurClus</th>
      <th>markerGene</th>
      <th>padj</th>
      <th>orthoGene</th>
      <th>orthoDescr</th>
      <th>pantherID</th>
      <th>pantherDescr</th>
      <th>goTerms</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>0</td>
      <td>XLOC_011074</td>
      <td>9.930711e-30</td>
      <td>TUBA1C</td>
      <td>tubulin alpha-1C chain isoform a [Homo sapiens]</td>
      <td>['PTHR11588:SF223']</td>
      <td>['SUBFAMILY NOT NAMED']</td>
      <td>['GO:0045298,GO:0005856,GO:0005200,GO:0005622,...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>0</td>
      <td>XLOC_015355</td>
      <td>2.959498e-09</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>['PTHR12124:SF53']</td>
      <td>["EXONUCLEASE 3'-5' DOMAIN-CONTAINING PROTEIN 1"]</td>
      <td>['GO:0016070,GO:0016787,GO:0016072,GO:0044238,...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>0</td>
      <td>XLOC_044603</td>
      <td>1.123759e-05</td>
      <td>NDUFV2</td>
      <td>NADH dehydrogenase [ubiquinone] flavoprotein ...</td>
      <td>['PTHR10371']</td>
      <td>['NADH DEHYDROGENASE  UBIQUINONE  FLAVOPROTEIN...</td>
      <td>['GO:0043234,GO:0005886,GO:0016020,GO:0032991,...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>0</td>
      <td>XLOC_030920</td>
      <td>3.931534e-05</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>['PTHR19290']</td>
      <td>['BASIC HELIX-LOOP-HELIX PROTEIN NEUROGENIN-RE...</td>
      <td>[nan]</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>0</td>
      <td>XLOC_004445</td>
      <td>3.710752e-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>['PTHR28618']</td>
      <td>['FAMILY NOT NAMED']</td>
      <td>[nan]</td>
    </tr>
  </tbody>
</table>
</div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># topGenes = []
# var_groups = []
# var_labels = []
# ind = 0
# n_genes = 10
# for i in np.unique(neuron_markers.neurClus):
#   sub = neuron_markers[neuron_markers.neurClus == i]
#   sub.sort_values(by=&#39;padj&#39;,ascending=True)

#   #noDups = [i for i in sub.markerGene if i not in topGenes] #Remove duplicate genes
#   topGenes += list(sub.markerGene[0:n_genes])

#   var_groups += [(ind,ind+n_genes-1)]
#   var_labels += [str(i)]
#   ind += n_genes
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># forBLAST = neuron_markers[neuron_markers.markerGene.isin(topGenes)]
# forBLAST.to_csv(&#39;blastNeuroPop_Genes.csv&#39;)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code># Use raw adata with all genes, to plot any gene expression in neuron subpopulations
neurons_raw = raw_fs_combo[raw_fs_combo.obs[&#39;cellRanger_louvain&#39;].isin([31,26,6,9])]


sc.pp.filter_cells(neurons_raw, min_counts=0)
sc.pp.filter_genes(neurons_raw, min_counts=0)

sc.pp.normalize_per_cell(neurons_raw, counts_per_cell_after=1e4)

sc.pp.log1p(neurons_raw)

#sc.pp.scale(neurons_raw,max_value=10)
neurons_raw.obs[&#39;louvain_neur&#39;] = neurons.obs[&#39;louvain_neur&#39;]
neurons_raw.obsm[&#39;X_tsne&#39;] = neurons.obsm[&#39;X_tsne&#39;]
neurons_raw.obsm[&#39;X_umap&#39;] = neurons.obsm[&#39;X_umap&#39;]

neurons_raw.uns[&#39;dendrogram_louvain_neur&#39;] = neurons.uns[&#39;dendrogram_louvain_neur&#39;]
neurons_raw.uns[&#39;louvain_neur_colors&#39;] = neurons.uns[&#39;louvain_neur_colors&#39;]
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">Trying to set attribute `.obs` of view, copying.</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Plot main figure genes

topGenes = [&#39;XLOC_018937&#39;,&#39;XLOC_035224&#39;,&#39;XLOC_040584&#39;,
            &#39;XLOC_008764&#39;,&#39;XLOC_040209&#39;, &#39;XLOC_010892&#39;,&#39;XLOC_003339&#39;,&#39;XLOC_019434&#39;,&#39;XLOC_042761&#39;,&#39;XLOC_017097&#39;,
            &#39;XLOC_021799&#39;,&#39;XLOC_041402&#39;,&#39;XLOC_004021&#39;,&#39;XLOC_008730&#39;,
            &#39;XLOC_014624&#39;,&#39;XLOC_030120&#39;,
            &#39;XLOC_038155&#39;] 
neurons_sigTop = neurons_raw[:,topGenes]

labels = [&#39;bHLH TF 2&#39;,&#39;+PP24&#39;,&#39;PP7&#39;,
          &#39;PP8&#39;,&#39;+PP19&#39;,&#39;+PP21&#39;,&#39;+PP15&#39;,&#39;PP5&#39;,&#39;PP1&#39;,&#39;PP11&#39;,
          &#39;+PP25&#39;,&#39;PP9&#39;,&#39;PP14&#39;,&#39;+PP26&#39;,
          &#39;+PP27&#39;,&#39;+PP17&#39;,
          &#39;+PP20&#39;]
neurons_sigTop.var[&#39;names&#39;] = labels

sc.set_figure_params(scanpy=True, fontsize=10)
sc.pl.heatmap(neurons_sigTop, labels, groupby=&#39;louvain_neur&#39;, dendrogram=True,show_gene_labels=True,
              gene_symbols = &#39;names&#39;,swap_axes=True,
              cmap=&#39;PuBuGn&#39;,standard_scale=&#39;var&#39;,save=&#39;neuronMarkers.pdf&#39;) #var_group_positions =var_groups ,var_group_labels = var_labels
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Trying</span> <span class="n">to</span> <span class="n">set</span> <span class="n">attribute</span> <span class="err">`</span><span class="o">.</span><span class="k">var</span><span class="err">`</span> <span class="n">of</span> <span class="n">view</span><span class="p">,</span> <span class="n">copying</span><span class="o">.</span>


<span class="n">WARNING</span><span class="p">:</span> <span class="n">saving</span> <span class="n">figure</span> <span class="n">to</span> <span class="n">file</span> <span class="n">figures</span><span class="o">/</span><span class="n">heatmapneuronMarkers</span><span class="o">.</span><span class="n">pdf</span>
</code></pre></div>

<p><img alt="png" src="../neuronSubpop_Analysis_files/neuronSubpop_Analysis_40_2.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.umap(neurons_raw,color=[&#39;XLOC_041402&#39;,&#39;XLOC_019434&#39;,&#39;XLOC_040584&#39;],title=[&#39;PP9&#39;,&#39;PP5&#39;,&#39;PP7&#39;],color_map=&#39;plasma&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../neuronSubpop_Analysis_files/neuronSubpop_Analysis_41_0.png" /></p>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../starvation_Analysis/" class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Cell Atlas Clustering and Analysis from kallisto-bustools
              </div>
            </div>
          </a>
        
        
          <a href="../pseudotime_Analysis/" class="md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Pseudotime Analysis
              </div>
            </div>
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/vendor.53cc9318.min.js"></script>
      <script src="../../assets/javascripts/bundle.e9c9f54f.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "../..",
          features: ['search.suggest', 'search.highlight'],
          search: Object.assign({
            worker: "../../assets/javascripts/worker/search.9c0e82ba.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>