
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Walkthrough for Whole Organisms Mutliplexed Data Analysis">
      
      
      
        <meta name="author" content="Tara Chari">
      
      
        <link rel="canonical" href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/notebooks/pseudotime_Analysis/">
      
      <link rel="shortcut icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.2.7">
    
    
      
        <title>Pseudotime Analysis - WHAMSeqAnalysis</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.cb6bc1d0.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.39b8e14a.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#download-data" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/" title="WHAMSeqAnalysis" class="md-header-nav__button md-logo" aria-label="WHAMSeqAnalysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      <div class="md-header-nav__ellipsis">
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            WHAMSeqAnalysis
          </span>
        </div>
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            
              Pseudotime Analysis
            
          </span>
        </div>
      </div>
    </div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    




<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="https://pachterlab.github.io/CWGFLHGCCHAP_2021/" title="WHAMSeqAnalysis" class="md-nav__button md-logo" aria-label="WHAMSeqAnalysis">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 003-3 3 3 0 00-3-3 3 3 0 00-3 3 3 3 0 003 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    WHAMSeqAnalysis
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" >
      
      <label class="md-nav__link" for="nav-2">
        ClickTag Pre-processing
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="ClickTag Pre-processing" data-md-level="1">
        <label class="md-nav__title" for="nav-2">
          <span class="md-nav__icon md-icon"></span>
          ClickTag Pre-processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellRangerClickTagCounts/" class="md-nav__link">
        Initial Quantification of ClickTag Counts with Cell Ranger (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistobusStarvClickTags/" class="md-nav__link">
        Quantification of ClickTag Counts with kallisto-bustools (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistobusStimClickTags/" class="md-nav__link">
        Quantification of ClickTag Counts with kallisto-bustools (Stimulation Data)
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3" >
      
      <label class="md-nav__link" for="nav-3">
        cDNA Pre-processing
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="cDNA Pre-processing" data-md-level="1">
        <label class="md-nav__title" for="nav-3">
          <span class="md-nav__icon md-icon"></span>
          cDNA Pre-processing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../filterStarvCells_ClickTags/" class="md-nav__link">
        Initial cDNA Filtering with Cell Ranger Data (Starvation Data)
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../kallistoBusRuns_StarvAndStim/" class="md-nav__link">
        Quantification of cDNA Counts with kallisto-bustools (All Data)
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4" checked>
      
      <label class="md-nav__link" for="nav-4">
        Cell Atlas and Perturbation Response
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Cell Atlas and Perturbation Response" data-md-level="1">
        <label class="md-nav__title" for="nav-4">
          <span class="md-nav__icon md-icon"></span>
          Cell Atlas and Perturbation Response
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellRangerClustering_Starvation/" class="md-nav__link">
        Initial Cell Atlas Clustering from Cell Ranger
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../starvation_Analysis/" class="md-nav__link">
        Cell Atlas Clustering and Analysis from kallisto-bustools
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../neuronSubpop_Analysis/" class="md-nav__link">
        Analysis of Neuron Subpopulations
      </a>
    </li>
  

          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Pseudotime Analysis
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Pseudotime Analysis
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import Packages
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generate-pseudotime-trajectories-for-i-cells-neurons-nematocytes" class="md-nav__link">
    Generate Pseudotime Trajectories for I-cells, Neurons, Nematocytes
  </a>
  
    <nav class="md-nav" aria-label="Generate Pseudotime Trajectories for I-cells, Neurons, Nematocytes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-regression" class="md-nav__link">
    Run Regression
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plot-results-of-regression-pseudotime-heatmaps" class="md-nav__link">
    Plot Results of Regression (Pseudotime Heatmaps)
  </a>
  
    <nav class="md-nav" aria-label="Plot Results of Regression (Pseudotime Heatmaps)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#final-pseudotime-plots" class="md-nav__link">
    Final Pseudotime Plots
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#check-gland-cell-trajectories-andor-hair-cells" class="md-nav__link">
    Check gland cell trajectories and/or hair cells
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../deSeq2Analysis_StarvationResponse/" class="md-nav__link">
        Extraction of Differentially Expressed Genes Under Starvation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5" >
      
      <label class="md-nav__link" for="nav-5">
        Stimulation Data Analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Stimulation Data Analysis" data-md-level="1">
        <label class="md-nav__title" for="nav-5">
          <span class="md-nav__icon md-icon"></span>
          Stimulation Data Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../stimulation_Analysis/" class="md-nav__link">
        Cell Atlas Concordance and Perturbation Response
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../deSeq2Analysis_StimulationResponse/" class="md-nav__link">
        Extraction of Differentially Expressed Genes Under Starvation
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6" >
      
      <label class="md-nav__link" for="nav-6">
        Inter-Experimental Analysis
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Inter-Experimental Analysis" data-md-level="1">
        <label class="md-nav__title" for="nav-6">
          <span class="md-nav__icon md-icon"></span>
          Inter-Experimental Analysis
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../allDistanceCalculations/" class="md-nav__link">
        Comparative Analysis of Cell Atlas Concordance and Perturbation Responses
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-7" type="checkbox" id="nav-7" >
      
      <label class="md-nav__link" for="nav-7">
        Explore Mutliplexed Datasets
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Explore Mutliplexed Datasets" data-md-level="1">
        <label class="md-nav__title" for="nav-7">
          <span class="md-nav__icon md-icon"></span>
          Explore Mutliplexed Datasets
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../MARIMBAAnnosAnalysis/" class="md-nav__link">
        Generation Cell Atlas with MARIMBA Transcriptome
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../MARIMBAAtlasSearchAndPlot/" class="md-nav__link">
        Explore MARIMBA-Annotated Cell Atlas
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../cellAtlasSearchAndPlot/" class="md-nav__link">
        Explore Original Cell Atlas
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download Data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#import-packages" class="md-nav__link">
    Import Packages
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generate-pseudotime-trajectories-for-i-cells-neurons-nematocytes" class="md-nav__link">
    Generate Pseudotime Trajectories for I-cells, Neurons, Nematocytes
  </a>
  
    <nav class="md-nav" aria-label="Generate Pseudotime Trajectories for I-cells, Neurons, Nematocytes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#run-regression" class="md-nav__link">
    Run Regression
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plot-results-of-regression-pseudotime-heatmaps" class="md-nav__link">
    Plot Results of Regression (Pseudotime Heatmaps)
  </a>
  
    <nav class="md-nav" aria-label="Plot Results of Regression (Pseudotime Heatmaps)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#final-pseudotime-plots" class="md-nav__link">
    Final Pseudotime Plots
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#check-gland-cell-trajectories-andor-hair-cells" class="md-nav__link">
    Check gland cell trajectories and/or hair cells
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/pachterlab/CWGFLHGCCHAP_2021/edit/master/docs/notebooks/pseudotime_Analysis.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                  <h1>Pseudotime Analysis</h1>
                
                <p><a href="https://colab.research.google.com/github/pachterlab/CWGFLHGCCHAP_2021/blob/master/notebooks/CellAtlasAnalysis/pseudotime_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!date
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">Wed Aug 19 16:45:58 UTC 2020</span>
</code></pre></div>

<h3 id="download-data"><strong>Download Data</strong><a class="headerlink" href="#download-data" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>import requests
from tqdm import tnrange, tqdm_notebook
def download_file(doi,ext):
    url = &#39;https://api.datacite.org/dois/&#39;+doi+&#39;/media&#39;
    r = requests.get(url).json()
    netcdf_url = r[&#39;data&#39;][0][&#39;attributes&#39;][&#39;url&#39;]
    r = requests.get(netcdf_url,stream=True)
    #Set file name
    fname = doi.split(&#39;/&#39;)[-1]+ext
    #Download file with progress bar
    if r.status_code == 403:
        print(&quot;File Unavailable&quot;)
    if &#39;content-length&#39; not in r.headers:
        print(&quot;Did not get file&quot;)
    else:
        with open(fname, &#39;wb&#39;) as f:
            total_length = int(r.headers.get(&#39;content-length&#39;))
            pbar = tnrange(int(total_length/1024), unit=&quot;B&quot;)
            for chunk in r.iter_content(chunk_size=1024):
                if chunk:
                    pbar.update()
                    f.write(chunk)
        return fname
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Kallisto bus clustered starvation data, h5ad
download_file(&#39;10.22002/D1.1796&#39;,&#39;.gz&#39;)

#Starvation h5ad data, all nonzero genes included, filtered for &#39;real cells&#39; from de-multiplexing
download_file(&#39;10.22002/D1.1797&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">479106</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">26058</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1797.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Previously saved neuron subpopulations
download_file(&#39;10.22002/D1.1804&#39;,&#39;.gz&#39;)

#Previously saved neuron pseudotime adata
download_file(&#39;10.22002/D1.1806&#39;,&#39;.gz&#39;)

#Previously saved nematocyte pseudotime adata
download_file(&#39;10.22002/D1.1805&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">1595</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">10910</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">12207</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1805.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Human ortholog annotations
download_file(&#39;10.22002/D1.1819&#39;,&#39;.gz&#39;)

#Panther annotations
download_file(&#39;10.22002/D1.1820&#39;,&#39;.gz&#39;)

#GO Terms
download_file(&#39;10.22002/D1.1822&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">528</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>






<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">515</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">227</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1822.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Previously saved ranked pseudotime genes

#Nematocyte trajectory (ranked genes)
#!wget --quiet https://caltech.box.com/shared/static/u19vzppejvt6ky3ysqptt2ia2dqozpkf
download_file(&#39;10.22002/D1.1808&#39;,&#39;.gz&#39;)

#Neuron trajectory (ranked genes)
#!wget --quiet https://caltech.box.com/shared/static/ippfug3xt20etmyx08egzpipgmyfc17m
download_file(&#39;10.22002/D1.1807&#39;,&#39;.gz&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="k">local</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="p">.</span><span class="mi">6</span><span class="o">/</span><span class="n">dist</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel_launcher</span><span class="p">.</span><span class="n">py</span><span class="p">:</span><span class="mi">18</span><span class="p">:</span> <span class="n">TqdmDeprecationWarning</span><span class="p">:</span> <span class="n">Please</span> <span class="n">use</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">notebook</span><span class="p">.</span><span class="n">trange</span><span class="o">`</span> <span class="k">instead</span> <span class="k">of</span> <span class="o">`</span><span class="n">tqdm</span><span class="p">.</span><span class="n">tnrange</span><span class="o">`</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">135</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>



<span class="n">HBox</span><span class="p">(</span><span class="n">children</span><span class="o">=</span><span class="p">(</span><span class="n">FloatProgress</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">.</span><span class="mi">0</span><span class="p">,</span> <span class="k">max</span><span class="o">=</span><span class="mi">134</span><span class="p">.</span><span class="mi">0</span><span class="p">),</span> <span class="n">HTML</span><span class="p">(</span><span class="n">value</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)))</span>





<span class="s1">&#39;D1.1807.gz&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>!gunzip *.gz
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Install packages
!pip install --quiet anndata
!pip install --quiet scanpy

!pip install --quiet louvain
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">[K     |████████████████████████████████| 122kB 5.8MB/s </span>
<span class="err">[K     |████████████████████████████████| 10.2MB 3.7MB/s </span>
<span class="err">[K     |████████████████████████████████| 51kB 6.3MB/s </span>
<span class="err">[K     |████████████████████████████████| 71kB 9.1MB/s </span>
<span class="err">[?25h  Building wheel for sinfo (setup.py) ... [?25l[?25hdone</span>
<span class="err">[K     |████████████████████████████████| 2.2MB 4.4MB/s </span>
<span class="err">[K     |████████████████████████████████| 3.2MB 28.2MB/s </span>
<span class="err">[?25h</span>
</code></pre></div>

<h3 id="import-packages"><strong>Import Packages</strong><a class="headerlink" href="#import-packages" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Import Packages

import random
import pandas as pd
import anndata
import scanpy as sc
import numpy as np
import scipy.sparse

import warnings
warnings.filterwarnings(&#39;ignore&#39;)

from sklearn.neighbors import (KNeighborsClassifier,NeighborhoodComponentsAnalysis)
from sklearn.pipeline import Pipeline
from sklearn.manifold import TSNE

#import scrublet as scr

import matplotlib.pyplot as plt
%matplotlib inline
sc.set_figure_params(dpi=125)

import seaborn as sns
sns.set(style=&quot;whitegrid&quot;)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Read in annotations
from io import StringIO

hg_ortho_df = pd.read_csv(StringIO(&#39;&#39;.join(l.replace(&#39;|&#39;, &#39;\t&#39;) for l in open(&#39;D1.1819&#39;))),
            sep=&quot;\t&quot;,header=None,skiprows=[0,1,2,3])

hg_ortho_df[[&#39;XLOC&#39;,&#39;TCONS&#39;]] = hg_ortho_df[13].str.split(expand=True) 
hg_ortho_df[[&#39;Gene&#39;,&#39;gi&#39;]] = hg_ortho_df[3].str.split(expand=True) 
hg_ortho_df[&#39;Description&#39;]= hg_ortho_df[11]


panther_df = pd.read_csv(&#39;D1.1820&#39;,
            sep=&quot;\t&quot;,header=None) #skiprows=[0,1,2,3]



goTerm_df = pd.read_csv(&#39;D1.1822&#39;,
            sep=&quot; &quot;,header=None) #skiprows=[0,1,2,3]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Read in saved data
bus_fs_combo = anndata.read(&#39;D1.1796&#39;)
print(bus_fs_combo)

bus_fs_raw  = anndata.read(&#39;D1.1797&#39;)
bus_fs_raw.obs[&#39;cellRanger_louvain&#39;] = pd.Categorical(bus_fs_combo.obs[&#39;cellRanger_louvain&#39;])
bus_fs_raw.obs[&#39;fed&#39;] = pd.Categorical(bus_fs_combo.obs[&#39;fed&#39;])
bus_fs_raw.obs[&#39;orgID&#39;] = pd.Categorical(bus_fs_combo.obs[&#39;orgID&#39;])
bus_fs_raw.obs[&#39;annos&#39;] = pd.Categorical(bus_fs_combo.obs[&#39;annos&#39;])
bus_fs_raw.obs[&#39;annosSub&#39;] = pd.Categorical(bus_fs_combo.obs[&#39;annosSub&#39;])

bus_fs_raw.uns[&#39;annos_colors&#39;] = bus_fs_combo.uns[&#39;annos_colors&#39;]
bus_fs_raw.uns[&#39;cellRanger_louvain_colors&#39;] = bus_fs_combo.uns[&#39;annosSub_colors&#39;]
#bus_fs_raw.uns[&#39;annosSub_colors&#39;] = bus_fs_combo.uns[&#39;annosSub_colors&#39;]
print(bus_fs_raw)

neuron_subpops  = anndata.read(&#39;D1.1804&#39;) 
print(neuron_subpops)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">13673</span> <span class="err">×</span> <span class="mi">8696</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;n_countslog&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;starved&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_neighbor_score&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;annos&#39;</span><span class="p">,</span> <span class="s1">&#39;new_cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;annosSub&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;annosSub_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;annos_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain_sizes&#39;</span><span class="p">,</span> <span class="s2">&quot;dendrogram_[&#39;new_cellRanger_louvain&#39;]&quot;</span><span class="p">,</span> <span class="s1">&#39;dendrogram_new_cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_neighbor_score_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;new_cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;paga&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_nca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
    <span class="n">obsp</span><span class="p">:</span> <span class="s1">&#39;connectivities&#39;</span><span class="p">,</span> <span class="s1">&#39;distances&#39;</span>
<span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">13673</span> <span class="err">×</span> <span class="mi">46716</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID&#39;</span><span class="p">,</span> <span class="s1">&#39;annos&#39;</span><span class="p">,</span> <span class="s1">&#39;annosSub&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;annos_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span>
<span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">1387</span> <span class="err">×</span> <span class="mi">2000</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_neur&#39;</span><span class="p">,</span> <span class="s1">&#39;test_louvain&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;highly_variable&#39;</span><span class="p">,</span> <span class="s1">&#39;means&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions_norm&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;hvg&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_neur_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">,</span> <span class="s1">&#39;test_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;umap&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_tsne&#39;</span><span class="p">,</span> <span class="s1">&#39;X_umap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
    <span class="n">obsp</span><span class="p">:</span> <span class="s1">&#39;connectivities&#39;</span><span class="p">,</span> <span class="s1">&#39;distances&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>  1
  2
  3
  4
  5
  6
  7
  8
  9
 10
 11
 12
 13
 14
 15
 16
 17
 18
 19
 20
 21
 22
 23
 24
 25
 26
 27
 28
 29
 30
 31
 32
 33
 34
 35
 36
 37
 38
 39
 40
 41
 42
 43
 44
 45
 46
 47
 48
 49
 50
 51
 52
 53
 54
 55
 56
 57
 58
 59
 60
 61
 62
 63
 64
 65
 66
 67
 68
 69
 70
 71
 72
 73
 74
 75
 76
 77
 78
 79
 80
 81
 82
 83
 84
 85
 86
 87
 88
 89
 90
 91
 92
 93
 94
 95
 96
 97
 98
 99
100
101
102
103
104
105</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Functions used

#Random forest regression to determine best genes to predict pseudotime
from sklearn.ensemble import RandomForestRegressor
from sklearn.inspection import permutation_importance
def runRegr_dptPseudotime(adata,adata_scaled,label=&#39;dpt_pseudotime&#39;):
  X = adata.X
  if scipy.sparse.issparse(X):
    X = X.toarray()

  #80% training data
  perc = 0.8
  n = np.int(perc*X.shape[0])

  np.random.seed(0)
  indices = np.random.choice(X.shape[0], n, replace=False,)
  X_train = X[indices,:]

  y = adata_scaled.obs[label]
  y_train = y[indices]

  X_test = X[[i for i in range(0,X.shape[0]) if i not in indices],:]
  y_test = y[[i for i in range(0,X.shape[0]) if i not in indices]]

  regr = RandomForestRegressor(random_state=42,max_samples=perc)
  regr.fit(X_train, y_train)
  print(regr.score(X_test, y_test))
  result = permutation_importance(regr, X_test, y_test, n_repeats=5, random_state=42)

  return result

#Convert filtered results to dataframe
def resToDF(result,adata):
  neur_gene_res = pd.DataFrame()
  neur_gene_res[&#39;Genes&#39;] = adata.var_names
  neur_gene_res[&#39;Mean&#39;] = result.importances_mean
  neur_gene_res[&#39;Std&#39;] = result.importances_std

  neur_gene_res= neur_gene_res.sort_values(&#39;Mean&#39;, axis=0, ascending=False)

  neur_gene_res= neur_gene_res[neur_gene_res.Mean &gt; 0]

  print(len(neur_gene_res.Genes))

  return neur_gene_res

#Make dataframe, with 100 marker genes for each cluster + annotations
def annotateResDF(resultDF):
  orthoGene = []
  orthoDescr = []

  pantherNum = []
  pantherDescr = []

  goTerms = []

  genes = resultDF.Genes

  for g in genes:

    sub_df = hg_ortho_df[hg_ortho_df.XLOC.isin([g])]
    panth_df = panther_df[panther_df[0].isin([g])]
    go_df = goTerm_df[goTerm_df[0].isin([g])]

    if len(sub_df) &gt; 0:
      #Save first result for gene/description
      orthoGene += [list(sub_df.Gene)[0]]
      orthoDescr += [list(sub_df.Description)[0]]
    else:
      orthoGene += [&#39;NA&#39;]
      orthoDescr += [&#39;NA&#39;]


    if len(panth_df) &gt; 0:
      pantherNum += [list(panth_df[1])]
      pantherDescr += [list(panth_df[2])]
    else:
      pantherNum += [&#39;NA&#39;]
      pantherDescr += [&#39;NA&#39;]


    if len(go_df) &gt; 0:
      goTerms += [list(go_df[1])]
    else:
      goTerms += [&#39;NA&#39;]


  resultDF[&#39;orthoGene&#39;] = orthoGene
  resultDF[&#39;orthoDescr&#39;] = orthoDescr

  resultDF[&#39;pantherID&#39;] = pantherNum
  resultDF[&#39;pantherDescr&#39;] = pantherDescr

  resultDF[&#39;goTerms&#39;] = goTerms

  return resultDF
  #list(neurons.uns[&#39;rank_genes_groups&#39;][&#39;names&#39;][&#39;1&#39;])

#Order cells in cell-by-gene matrix by pseudotime
def orderTime(adata):
  sortTime = np.argsort(adata.obs[&#39;dpt_pseudotime&#39;])
  newNames = adata.obs_names[list(sortTime)]

  adataRet = adata[newNames,:]
  return adataRet
</code></pre></div>
</td></tr></table>
<h3 id="generate-pseudotime-trajectories-for-i-cells-neurons-nematocytes"><strong>Generate Pseudotime Trajectories for I-cells, Neurons, Nematocytes</strong><a class="headerlink" href="#generate-pseudotime-trajectories-for-i-cells-neurons-nematocytes" title="Permanent link">&para;</a></h3>
<p>Select cell types 0, 11, 12, 17, 23 6, 9, 26 &amp; 31 which represent the i-cells, Neural cells, and Nematocytes for downstream analysis</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Subset data

#I-cells, neurons, nematocytes
big_lin = bus_fs_raw[bus_fs_raw.obs[&#39;cellRanger_louvain&#39;].isin([31,26,6,9,0,11,12,17,23])]


sc.pp.filter_cells(big_lin, min_counts=0)
sc.pp.filter_genes(big_lin, min_counts=1)
sc.pp.normalize_per_cell(big_lin, counts_per_cell_after=1e4)

big_lin_copy = big_lin.copy()
sc.pp.log1p(big_lin)

big_lin.raw = sc.pp.log1p(big_lin_copy, copy=True)


sc.pp.highly_variable_genes(big_lin, n_top_genes=4000,n_bins=50)

big_lin = big_lin[:,big_lin.var[&#39;highly_variable&#39;]]

big_lin
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Trying</span> <span class="n">to</span> <span class="n">set</span> <span class="n">attribute</span> <span class="err">`</span><span class="o">.</span><span class="n">obs</span><span class="err">`</span> <span class="n">of</span> <span class="n">view</span><span class="p">,</span> <span class="n">copying</span><span class="o">.</span>





<span class="n">View</span> <span class="n">of</span> <span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">4064</span> <span class="err">×</span> <span class="mi">4000</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID&#39;</span><span class="p">,</span> <span class="s1">&#39;annos&#39;</span><span class="p">,</span> <span class="s1">&#39;annosSub&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;highly_variable&#39;</span><span class="p">,</span> <span class="s1">&#39;means&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions_norm&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;annos_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;log1p&#39;</span><span class="p">,</span> <span class="s1">&#39;hvg&#39;</span>
</code></pre></div>

<p>Cluster and embed cells for pseudotime visualization</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Large lineage PAGA embedding

big_lin_scaled = big_lin
sc.pp.scale(big_lin_scaled, max_value=10)

sc.tl.pca(big_lin_scaled, n_comps=60)
#sc.pl.pca_variance_ratio(neur_lin_scaled, log=True)
sc.pp.neighbors(big_lin_scaled,n_neighbors=50, n_pcs=15,method=&#39;gauss&#39;)

sc.tl.louvain(big_lin_scaled,resolution=1,key_added=&#39;louvain_test&#39;)#Clustering algorithm,resolution=0.5


sc.tl.paga(big_lin_scaled, groups=&#39;louvain_test&#39;,)
sc.pl.paga(big_lin_scaled, color=[&#39;louvain_test&#39;])
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../pseudotime_Analysis_files/pseudotime_Analysis_19_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Add neuron subpopulation labels
labels = []
for c in big_lin_scaled.obs_names:
  if c in neuron_subpops.obs_names:

    labels += [neuron_subpops[c,:].obs[&#39;louvain_neur&#39;][0]]
  else:
    labels += [&#39;rest&#39;]

big_lin_scaled.obs[&#39;louvain_neur&#39;] = pd.Categorical(labels)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.draw_graph(big_lin_scaled, init_pos=&#39;paga&#39;)
#sc.pl.draw_graph(big_lin_scaled, color=[&#39;louvain_test&#39;,&#39;louvain_neur&#39;,&#39;cellRanger_louvain&#39;,&#39;fed&#39;], legend_loc=&#39;on data&#39;,color_map=&#39;viridis&#39;,alpha=0.5)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Calculate pseudotime with diffusion components (dim reduction)

big_lin_scaled.uns[&#39;iroot&#39;] = np.flatnonzero(big_lin_scaled.obs[&#39;louvain_test&#39;]  == &#39;0&#39;)[0]
sc.tl.diffmap(big_lin_scaled,n_comps=10)
sc.tl.dpt(big_lin_scaled,n_dcs=10)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.draw_graph(big_lin_scaled, color=[&#39;louvain_test&#39;,&#39;annos&#39;,&#39;fed&#39;,&#39;dpt_pseudotime&#39;,&#39;louvain_neur&#39;], 
                 color_map=&#39;viridis&#39;) #legend_loc=&#39;on data&#39;
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">*c* argument looks like a single numeric RGB or RGBA sequence, which should be avoided as value-mapping will have precedence in case its length matches with *x* &amp; *y*.  Please use the *color* keyword-argument or provide a 2-D array with a single row if you intend to specify the same RGB or RGBA value for all points.</span>
<span class="err">*c* argument looks like a single numeric RGB or RGBA sequence, which should be avoided as value-mapping will have precedence in case its length matches with *x* &amp; *y*.  Please use the *color* keyword-argument or provide a 2-D array with a single row if you intend to specify the same RGB or RGBA value for all points.</span>
</code></pre></div>

<p><img alt="png" src="../pseudotime_Analysis_files/pseudotime_Analysis_23_1.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.draw_graph(big_lin_scaled, color=[&#39;XLOC_038831&#39;,&#39;XLOC_010752&#39;,&#39;XLOC_018937&#39;,&#39;XLOC_012650&#39;,&#39;XLOC_001436&#39;,&#39;XLOC_030068&#39;], 
                 legend_loc=&#39;on data&#39;,color_map=&#39;plasma&#39;,s=80) 
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../pseudotime_Analysis_files/pseudotime_Analysis_24_0.png" /></p>
<p>Make adata object including more variable genes to rank by their importance/relevance to the pseudotime trajectories</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Make adata for doing regression (more variable genes included)
big_lin_reg = bus_fs_raw[bus_fs_raw.obs[&#39;cellRanger_louvain&#39;].isin([31,26,6,9,0,11,12,17,23])]


sc.pp.filter_cells(big_lin_reg, min_counts=0)
sc.pp.filter_genes(big_lin_reg, min_counts=1)
sc.pp.normalize_per_cell(big_lin_reg, counts_per_cell_after=1e4)

big_lin_copy = big_lin_reg.copy()
sc.pp.log1p(big_lin_reg)

big_lin_reg.raw = sc.pp.log1p(big_lin_copy, copy=True)

#For shorter demonstration use 4000 option
#sc.pp.highly_variable_genes(big_lin_reg, n_top_genes=4000,n_bins=10)
sc.pp.highly_variable_genes(big_lin_reg, n_top_genes=15000,n_bins=10) 

big_lin_reg = big_lin_reg[:,big_lin_reg.var[&#39;highly_variable&#39;]]
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">Trying to set attribute `.obs` of view, copying.</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Convert pseudotime valeus to quantiles of pseudotime 
big_lin_reg.obs[&#39;dpt_pseudotime&#39;] = big_lin_scaled.obs[&#39;dpt_pseudotime&#39;]

nbins = 20 #quantiles
rangeList = range(0,nbins)
strList = [str(i) for i in rangeList]

cnido = big_lin_reg[big_lin_reg.obs[&#39;cellRanger_louvain&#39;].isin([0,11,12,17,23])]
cnido.obs[&#39;quantTimeBlocks&#39;] = pd.qcut(cnido.obs[&#39;dpt_pseudotime&#39;],q=nbins,labels=strList)

neuro = big_lin_reg[big_lin_reg.obs[&#39;cellRanger_louvain&#39;].isin([0,26,31,9,6])]
neuro.obs[&#39;quantTimeBlocks&#39;] = pd.qcut(neuro.obs[&#39;dpt_pseudotime&#39;],q=nbins,labels=strList)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">Trying to set attribute `.obs` of view, copying.</span>
<span class="err">Trying to set attribute `.obs` of view, copying.</span>
<span class="err">Trying to set attribute `.obs` of view, copying.</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>cnido.write(&#39;nemato.h5ad&#39;)
neuro.write(&#39;neuro.h5ad&#39;)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>big_lin_scaled
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">4064</span> <span class="err">×</span> <span class="mi">4000</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID&#39;</span><span class="p">,</span> <span class="s1">&#39;annos&#39;</span><span class="p">,</span> <span class="s1">&#39;annosSub&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_test&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_neur&#39;</span><span class="p">,</span> <span class="s1">&#39;dpt_pseudotime&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;highly_variable&#39;</span><span class="p">,</span> <span class="s1">&#39;means&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions_norm&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;annos_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;log1p&#39;</span><span class="p">,</span> <span class="s1">&#39;hvg&#39;</span><span class="p">,</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="s1">&#39;neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;paga&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_test_sizes&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_test_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;draw_graph&#39;</span><span class="p">,</span> <span class="s1">&#39;iroot&#39;</span><span class="p">,</span> <span class="s1">&#39;diffmap_evals&#39;</span><span class="p">,</span> <span class="s1">&#39;fed_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;louvain_neur_colors&#39;</span>
    <span class="n">obsm</span><span class="p">:</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">,</span> <span class="s1">&#39;X_diffmap&#39;</span>
    <span class="n">varm</span><span class="p">:</span> <span class="s1">&#39;PCs&#39;</span>
    <span class="n">obsp</span><span class="p">:</span> <span class="s1">&#39;distances&#39;</span><span class="p">,</span> <span class="s1">&#39;connectivities&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Create embedding for plotting pseudotime genes
big_lin_reg.uns[&#39;paga&#39;] = big_lin_scaled.uns[&#39;paga&#39;] 

big_lin_reg.uns[&#39;draw_graph&#39;] = big_lin_scaled.uns[&#39;draw_graph&#39;]

big_lin_reg.obsm[&#39;X_draw_graph_fa&#39;] = big_lin_scaled.obsm[&#39;X_draw_graph_fa&#39;]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.draw_graph(big_lin_reg, color=[&#39;XLOC_019434&#39;,&#39;XLOC_018937&#39;], 
                 legend_loc=&#39;on data&#39;,color_map=&#39;plasma&#39;) 
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../pseudotime_Analysis_files/pseudotime_Analysis_31_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>big_lin_reg.write(&#39;pseudoEmbed_fullLineage.h5ad&#39;)
</code></pre></div>
</td></tr></table>
<h4 id="run-regression"><strong>Run Regression</strong><a class="headerlink" href="#run-regression" title="Permanent link">&para;</a></h4>
<p>Will take a while. Can plot with saved results instead of re-running model or run with less genes</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Run regression

result_cnido = runRegr_dptPseudotime(cnido,cnido,&#39;quantTimeBlocks&#39;)
result_neuro = runRegr_dptPseudotime(neuro,neuro,&#39;quantTimeBlocks&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">0.9252565851716389</span>
<span class="err">0.8683961569750528</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Make dataframe from regression results
cnido_gene_res = resToDF(result_cnido,cnido)
neuro_gene_res = resToDF(result_neuro,neuro)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">2202</span>
<span class="err">2186</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>cnido_gene_res = annotateResDF(cnido_gene_res)
neuro_gene_res = annotateResDF(neuro_gene_res)
cnido_gene_res.head()
</code></pre></div>
</td></tr></table>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Genes</th>
      <th>Mean</th>
      <th>Std</th>
      <th>orthoGene</th>
      <th>orthoDescr</th>
      <th>pantherID</th>
      <th>pantherDescr</th>
      <th>goTerms</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>629</th>
      <td>XLOC_001436</td>
      <td>0.264423</td>
      <td>0.018149</td>
      <td>NA</td>
      <td>NA</td>
      <td>[PTHR35378]</td>
      <td>[FAMILY NOT NAMED]</td>
      <td>[nan]</td>
    </tr>
    <tr>
      <th>631</th>
      <td>XLOC_001437</td>
      <td>0.075968</td>
      <td>0.006814</td>
      <td>NA</td>
      <td>NA</td>
      <td>NA</td>
      <td>NA</td>
      <td>NA</td>
    </tr>
    <tr>
      <th>3242</th>
      <td>XLOC_008048</td>
      <td>0.066162</td>
      <td>0.005565</td>
      <td>LOC102723665</td>
      <td>PREDICTED: basic proline-rich protein-like [H...</td>
      <td>[PTHR24023:SF718]</td>
      <td>[COL-129, ISOFORM A]</td>
      <td>[GO:0032502,GO:0016337,GO:0007398,GO:0007498,G...</td>
    </tr>
    <tr>
      <th>7688</th>
      <td>XLOC_021750</td>
      <td>0.017792</td>
      <td>0.001821</td>
      <td>ACTG1</td>
      <td>actin, cytoplasmic 2 [Homo sapiens]</td>
      <td>[PTHR11937:SF347]</td>
      <td>[BETA-ACTIN-LIKE PROTEIN 3-RELATED]</td>
      <td>[GO:0000910,GO:0005856,GO:0005200,GO:0006897,G...</td>
    </tr>
    <tr>
      <th>12680</th>
      <td>XLOC_039385</td>
      <td>0.015740</td>
      <td>0.002258</td>
      <td>CPE</td>
      <td>carboxypeptidase E preproprotein [Homo sapiens]</td>
      <td>[PTHR11532:SF68]</td>
      <td>[CARBOXYPEPTIDASE E]</td>
      <td>[GO:0016787,GO:0005615,GO:0044238,GO:0019538,G...</td>
    </tr>
  </tbody>
</table>
</div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>cnido_gene_res.to_csv(&#39;cnido_stem_lineage_pseudotime_genes.csv&#39;)
neuro_gene_res.to_csv(&#39;neuro_stem_lineage_pseudotime_genes.csv&#39;)
</code></pre></div>
</td></tr></table>
<h3 id="plot-results-of-regression-pseudotime-heatmaps"><strong>Plot Results of Regression (Pseudotime Heatmaps)</strong><a class="headerlink" href="#plot-results-of-regression-pseudotime-heatmaps" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>def returnVal(i):
  if i == i:
    i= i.replace(&quot;[&quot;,&quot;&quot;)
    i = i.replace(&quot;]&quot;,&quot;&quot;)
    i= i.replace(&quot;&#39;&quot;,&quot;&quot;)
    i = i.replace(&quot;&#39;&quot;,&quot;&quot;)
    return i 
  else:
    return &#39;nan&#39;

neuro_gene_res = pd.read_csv(&#39;D1.1807&#39;)
print(neuro_gene_res.head())

cnido = anndata.read(&#39;D1.1805&#39;)
neuro = anndata.read(&#39;D1.1806&#39;)
print(neuro)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="w">   </span><span class="nl">Unnamed</span><span class="p">:</span><span class="w"> </span><span class="mi">0</span><span class="w">  </span><span class="p">...</span><span class="w">                                            </span><span class="n">goTerms</span><span class="w"></span>
<span class="mi">0</span><span class="w">        </span><span class="mi">8941</span><span class="w">  </span><span class="p">...</span><span class="w">                                                </span><span class="n">NaN</span><span class="w"></span>
<span class="mi">1</span><span class="w">        </span><span class="mi">4963</span><span class="w">  </span><span class="p">...</span><span class="w">                                                </span><span class="n">NaN</span><span class="w"></span>
<span class="mi">2</span><span class="w">        </span><span class="mi">4295</span><span class="w">  </span><span class="p">...</span><span class="w">                                              </span><span class="o">[</span><span class="n">nan</span><span class="o">]</span><span class="w"></span>
<span class="mi">3</span><span class="w">        </span><span class="mi">4164</span><span class="w">  </span><span class="p">...</span><span class="w">                                              </span><span class="o">[</span><span class="n">nan</span><span class="o">]</span><span class="w"></span>
<span class="mi">4</span><span class="w">       </span><span class="mi">10102</span><span class="w">  </span><span class="p">...</span><span class="w">  </span><span class="o">[</span><span class="n">&#39;GO:0006996,GO:0044238,GO:0005622,GO:0032991,...</span>

<span class="n">[5 rows x 9 columns</span><span class="o">]</span><span class="w"></span>
<span class="n">AnnData</span><span class="w"> </span><span class="k">object</span><span class="w"> </span><span class="k">with</span><span class="w"> </span><span class="n">n_obs</span><span class="w"> </span><span class="err">×</span><span class="w"> </span><span class="n">n_vars</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">2662</span><span class="w"> </span><span class="err">×</span><span class="w"> </span><span class="mi">14999</span><span class="w"></span>
<span class="w">    </span><span class="nl">obs</span><span class="p">:</span><span class="w"> </span><span class="s1">&#39;batch&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;fed&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;orgID&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;annos&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;annosSub&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;n_counts&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;dpt_pseudotime&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;quantTimeBlocks&#39;</span><span class="w"></span>
<span class="w">    </span><span class="nf">var</span><span class="err">:</span><span class="w"> </span><span class="s1">&#39;n_counts&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;highly_variable&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;means&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;dispersions&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;dispersions_norm&#39;</span><span class="w"></span>
<span class="w">    </span><span class="nl">uns</span><span class="p">:</span><span class="w"> </span><span class="s1">&#39;annos_colors&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;hvg&#39;</span><span class="w"></span>
</code></pre></div>

<p>Plot expression of genes of interest over time, after ordering cells by pseudotime coordinates so that expression is shown over 'time'</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Order cells by pseudotime (0 --&gt; 1)
orderCnido = orderTime(cnido)
orderNeuro = orderTime(neuro)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Plot interesting genes from ranking over time
neuroPeps = [&#39;XLOC_042761&#39;,&#39;XLOC_019434&#39;,&#39;XLOC_017097&#39;,&#39;XLOC_030120&#39;]
n_names = [&#39;*PP1&#39;,&#39;*PP5&#39;,&#39;*PP11&#39;,&#39;PP17&#39;]

early = [&#39;XLOC_038831&#39;,&#39;XLOC_041836&#39;,&#39;XLOC_010226&#39;,&#39;XLOC_041854&#39;,&#39;XLOC_045448&#39;] #&#39;XLOC_041854&#39;,&#39;XLOC_045448&#39;

e_names = [&#39;Semaphorin&#39;,&#39;NACA2&#39;,&#39;aGPCR 3&#39;,&#39;LRR protein 1&#39;,&#39;LRR protein 2&#39;]

late = [&#39;XLOC_034623&#39;,&#39;XLOC_008333&#39;,&#39;XLOC_040506&#39;,&#39;XLOC_012560&#39;,&#39;XLOC_044075&#39;,
        &#39;XLOC_043514&#39;,&#39;XLOC_034910&#39;,&#39;XLOC_001206&#39;,
        &#39;XLOC_003576&#39;,&#39;XLOC_002941&#39;,&#39;XLOC_008328&#39;,&#39;XLOC_044176&#39;,&#39;XLOC_017070&#39;] #&#39;XLOC_034910&#39;,&#39;XLOC_040506&#39;

l_names = [&#39;SIX-like&#39;,&#39;ACTR3&#39;,&#39;CAB39&#39;,&#39;Clytia-Specific 4&#39;,&#39;ASNS&#39;,
           &#39;POLR1A&#39;,&#39;NCALD&#39;,&#39;aGPCR 1&#39;,
           &#39;KCNA6&#39;,&#39;aGPCR 2&#39;,&#39;TRIO&#39;,&#39;Clytia-Specific 5&#39;,&#39;aGPCR 4&#39;]

total = early+neuroPeps+late
total_names = e_names+n_names+l_names
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>set(neuroPeps).intersection(late)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="err">set()</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>toAddNeuroDict_Other = {}
for i in range(0,len(total)):
  toAddNeuroDict_Other[total[i]] = total_names[i]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>labels= [toAddNeuroDict_Other[i] for i in total]
orderNeuro = orderNeuro[:,total]
orderNeuro.var[&#39;names&#39;] = labels


sc.set_figure_params(scanpy=True, fontsize=12)
sc.pl.heatmap(orderNeuro, total_names, groupby=None, 
              show_gene_labels=True,use_raw=False,standard_scale=&#39;var&#39;,
              interpolation=&#39;gaussian&#39;,cmap=&#39;plasma&#39;,swap_axes=True,gene_symbols= &#39;names&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Trying</span> <span class="n">to</span> <span class="n">set</span> <span class="n">attribute</span> <span class="err">`</span><span class="o">.</span><span class="k">var</span><span class="err">`</span> <span class="n">of</span> <span class="n">view</span><span class="p">,</span> <span class="n">copying</span><span class="o">.</span>
</code></pre></div>

<p><img alt="png" src="../pseudotime_Analysis_files/pseudotime_Analysis_45_1.png" /></p>
<h5 id="final-pseudotime-plots"><strong>Final Pseudotime Plots</strong><a class="headerlink" href="#final-pseudotime-plots" title="Permanent link">&para;</a></h5>
<p>Plots include a mixture of neuro-development related genes</p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Add more pseudotime ranked genes for heatmap
toAddNeuro = [&#39;XLOC_037011&#39;,&#39;XLOC_031923&#39;,&#39;XLOC_010412&#39;,&#39;XLOC_030920&#39;,&#39;XLOC_026271&#39;,&#39;XLOC_025094&#39;,&#39;XLOC_018937&#39;,&#39;XLOC_037278&#39;,&#39;XLOC_009173&#39;,&#39;XLOC_033489&#39;,
                                        &#39;XLOC_010752&#39;,&#39;XLOC_045293&#39;,&#39;XLOC_033337&#39;,&#39;XLOC_012650&#39;]
toAddNeuroNames = [&#39;TUBA4A&#39;,&#39;PCNA&#39;,&#39;GLUL&#39;,&#39;bHLH TF 1&#39;,&#39;DCX&#39;,&#39;CALR&#39;,&#39;bHLH TF 2&#39;,&#39;TAGLN-2&#39;,&#39;H2AFV&#39;,&#39;PRDX4&#39;,
                          &#39;bHLH TF 3&#39;,&#39;*SOX-10&#39;,&#39;CL1&#39;,&#39;Clytia-Specific 3&#39;]

toAddNeuroDict = {}
for i in range(0,len(toAddNeuro)):
  toAddNeuroDict[toAddNeuro[i]] = toAddNeuroNames[i]


toAddCnido = [&#39;XLOC_001436&#39;,&#39;XLOC_008048&#39;,&#39;XLOC_030068&#39;,&#39;XLOC_011074&#39;,&#39;XLOC_041365&#39;,&#39;XLOC_041836&#39;,&#39;XLOC_011949&#39;,&#39;XLOC_015554&#39;]
toAddCnidoNames = [&#39;*Mcol3/4&#39;,&#39;Mcol&#39;,&#39;*Dkk3&#39;,&#39;TUBA1C&#39;,&#39;TF IIIA&#39;,&#39;NACA2&#39;,&#39;BOP1&#39;,&#39;Mos3&#39;]
toAddCnidoDict = {}
for i in range(0,len(toAddCnido)):
  toAddCnidoDict[toAddCnido[i]] = toAddCnidoNames[i]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Determine where var_names is in list of genesToPlot
allGenes = {**toAddCnidoDict, **toAddNeuroDict}
allGenes = {**allGenes,**toAddNeuroDict_Other}  #&#39;XLOC_002941&#39;,&#39;XLOC_043514&#39;,&#39;XLOC_037011&#39;,&#39;XLOC_040506&#39;,&#39;XLOC_003576&#39;,

forPlotNeuro = [&#39;XLOC_041836&#39;,&#39;XLOC_026271&#39;,&#39;XLOC_031923&#39;,&#39;XLOC_010226&#39;,
                &#39;XLOC_018937&#39;,&#39;XLOC_045293&#39;,
                &#39;XLOC_034623&#39;,&#39;XLOC_040506&#39;,&#39;XLOC_012560&#39;,&#39;XLOC_030920&#39;,&#39;XLOC_010752&#39;,
                &#39;XLOC_038831&#39;,&#39;XLOC_001206&#39;,&#39;XLOC_042761&#39;,&#39;XLOC_017097&#39;,
                &#39;XLOC_002941&#39;,&#39;XLOC_030120&#39;,&#39;XLOC_019434&#39;,
                &#39;XLOC_034910&#39;,&#39;XLOC_012650&#39;]


forPlotCnido = [&#39;XLOC_011949&#39;,&#39;XLOC_011074&#39;,&#39;XLOC_041836&#39;,&#39;XLOC_015554&#39;,&#39;XLOC_041365&#39;,&#39;XLOC_030068&#39;,&#39;XLOC_008048&#39;,&#39;XLOC_001436&#39;]
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>labelsCnido = [allGenes[i] for i in forPlotCnido]
labelsNeuro = [allGenes[i] for i in forPlotNeuro]

#Order cells by pseudotime (0 --&gt; 1)
orderCnido = orderTime(cnido)
orderNeuro = orderTime(neuro)

orderCnido = orderCnido[:,forPlotCnido]
orderCnido.var[&#39;names&#39;] = labelsCnido


orderNeuro = orderNeuro[:,forPlotNeuro]
orderNeuro.var[&#39;names&#39;] = labelsNeuro
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Trying</span> <span class="n">to</span> <span class="n">set</span> <span class="n">attribute</span> <span class="err">`</span><span class="o">.</span><span class="k">var</span><span class="err">`</span> <span class="n">of</span> <span class="n">view</span><span class="p">,</span> <span class="n">copying</span><span class="o">.</span>
<span class="n">Trying</span> <span class="n">to</span> <span class="n">set</span> <span class="n">attribute</span> <span class="err">`</span><span class="o">.</span><span class="k">var</span><span class="err">`</span> <span class="n">of</span> <span class="n">view</span><span class="p">,</span> <span class="n">copying</span><span class="o">.</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.set_figure_params(scanpy=True, fontsize=12)
sc.pl.heatmap(orderCnido, labelsCnido, groupby=None, 
              show_gene_labels=True,use_raw=False,standard_scale=&#39;var&#39;,
              interpolation=&#39;gaussian&#39;,cmap=&#39;plasma&#39;,swap_axes=True,gene_symbols=&#39;names&#39;,figsize=(8,2),
              save=&#39;cnido_gauss_BigLinTimeGenes.pdf&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">WARNING</span><span class="o">:</span> <span class="n">saving</span> <span class="n">figure</span> <span class="n">to</span> <span class="n">file</span> <span class="n">figures</span><span class="o">/</span><span class="n">heatmapcnido_gauss_BigLinTimeGenes</span><span class="o">.</span><span class="na">pdf</span>
</code></pre></div>

<p><img alt="png" src="../pseudotime_Analysis_files/pseudotime_Analysis_50_1.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.set_figure_params(scanpy=True, fontsize=12)
sc.pl.heatmap(orderNeuro, labelsNeuro, groupby=None, 
              show_gene_labels=True,use_raw=False,standard_scale=&#39;var&#39;,
              interpolation=&#39;gaussian&#39;,cmap=&#39;plasma&#39;,swap_axes=True,gene_symbols=&#39;names&#39;,figsize=(5,4),
              save=&#39;neuro_gauss_BigLinTimeGenes.pdf&#39;)
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">WARNING</span><span class="o">:</span> <span class="n">saving</span> <span class="n">figure</span> <span class="n">to</span> <span class="n">file</span> <span class="n">figures</span><span class="o">/</span><span class="n">heatmapneuro_gauss_BigLinTimeGenes</span><span class="o">.</span><span class="na">pdf</span>
</code></pre></div>

<p><img alt="png" src="../pseudotime_Analysis_files/pseudotime_Analysis_51_1.png" /></p>
<h3 id="check-gland-cell-trajectories-andor-hair-cells"><strong>Check gland cell trajectories and/or hair cells</strong><a class="headerlink" href="#check-gland-cell-trajectories-andor-hair-cells" title="Permanent link">&para;</a></h3>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Subset data

#Gland cells: 27,25,32,34,22 + no 31,26,6,9
#Actually including sensory hair cell populations, 5, 10, 21
gland_lin = bus_fs_raw[bus_fs_raw.obs[&#39;cellRanger_louvain&#39;].isin([5,10,21,0,11,12,17,23])]


sc.pp.filter_cells(gland_lin, min_counts=0)
sc.pp.filter_genes(gland_lin, min_counts=1)
sc.pp.normalize_per_cell(gland_lin, counts_per_cell_after=1e4)

sc.pp.log1p(gland_lin)


sc.pp.highly_variable_genes(gland_lin, n_top_genes=4000,n_bins=50)

gland_lin = gland_lin[:,gland_lin.var[&#39;highly_variable&#39;]]

gland_lin
</code></pre></div>
</td></tr></table>
<div class="codehilite"><pre><span></span><code><span class="n">Trying</span> <span class="n">to</span> <span class="n">set</span> <span class="n">attribute</span> <span class="err">`</span><span class="o">.</span><span class="n">obs</span><span class="err">`</span> <span class="n">of</span> <span class="n">view</span><span class="p">,</span> <span class="n">copying</span><span class="o">.</span>





<span class="n">View</span> <span class="n">of</span> <span class="n">AnnData</span> <span class="n">object</span> <span class="n">with</span> <span class="n">n_obs</span> <span class="err">×</span> <span class="n">n_vars</span> <span class="o">=</span> <span class="mi">4094</span> <span class="err">×</span> <span class="mi">4000</span>
    <span class="n">obs</span><span class="p">:</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain&#39;</span><span class="p">,</span> <span class="s1">&#39;fed&#39;</span><span class="p">,</span> <span class="s1">&#39;orgID&#39;</span><span class="p">,</span> <span class="s1">&#39;annos&#39;</span><span class="p">,</span> <span class="s1">&#39;annosSub&#39;</span><span class="p">,</span> <span class="s1">&#39;n_counts&#39;</span>
    <span class="k">var</span><span class="p">:</span> <span class="s1">&#39;n_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;highly_variable&#39;</span><span class="p">,</span> <span class="s1">&#39;means&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions&#39;</span><span class="p">,</span> <span class="s1">&#39;dispersions_norm&#39;</span>
    <span class="n">uns</span><span class="p">:</span> <span class="s1">&#39;annos_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;cellRanger_louvain_colors&#39;</span><span class="p">,</span> <span class="s1">&#39;log1p&#39;</span><span class="p">,</span> <span class="s1">&#39;hvg&#39;</span>
</code></pre></div>

<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Large lineage PAGA embedding

gland_lin_scaled = gland_lin
sc.pp.scale(gland_lin_scaled, max_value=10)

sc.tl.pca(gland_lin_scaled, n_comps=60)
#sc.pl.pca_variance_ratio(neur_lin_scaled, log=True)
sc.pp.neighbors(gland_lin_scaled,n_neighbors=100, n_pcs=60,method=&#39;gauss&#39;)

sc.tl.louvain(gland_lin_scaled,resolution=2,key_added=&#39;louvain_test&#39;)#Clustering algorithm,resolution=0.5


sc.tl.paga(gland_lin_scaled, groups=&#39;louvain_test&#39;,)
sc.pl.paga(gland_lin_scaled, color=[&#39;louvain_test&#39;])
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../pseudotime_Analysis_files/pseudotime_Analysis_54_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.draw_graph(gland_lin_scaled, init_pos=&#39;paga&#39;) #&#39;louvain_test&#39;
sc.pl.draw_graph(gland_lin_scaled, color=[&#39;louvain_test&#39;,&#39;cellRanger_louvain&#39;,&#39;fed&#39;], 
                 legend_loc=&#39;on data&#39;,color_map=&#39;viridis&#39;,alpha=0.5)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../pseudotime_Analysis_files/pseudotime_Analysis_55_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>#Calculate pseudotime with diffusion components (dim reduction)

gland_lin_scaled.uns[&#39;iroot&#39;] = np.flatnonzero(gland_lin_scaled.obs[&#39;louvain_test&#39;]  == &#39;2&#39;)[0]
sc.tl.diffmap(gland_lin_scaled,n_comps=10)
sc.tl.dpt(gland_lin_scaled,n_dcs=10)
</code></pre></div>
</td></tr></table>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.pl.draw_graph(gland_lin_scaled, color=[&#39;louvain_test&#39;,&#39;cellRanger_louvain&#39;,&#39;fed&#39;,&#39;dpt_pseudotime&#39;], 
                 legend_loc=&#39;on data&#39;,color_map=&#39;viridis&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../pseudotime_Analysis_files/pseudotime_Analysis_57_0.png" /></p>
<table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><code>sc.tl.dendrogram(gland_lin_scaled,&#39;cellRanger_louvain&#39;,linkage_method=&#39;ward&#39;)
gland_lin_scaled.uns[&#39;dendrogram_cellRanger_louvain&#39;] = gland_lin_scaled.uns[&quot;dendrogram_[&#39;cellRanger_louvain&#39;]&quot;]
sc.pl.dendrogram(gland_lin_scaled,&#39;cellRanger_louvain&#39;)
</code></pre></div>
</td></tr></table>
<p><img alt="png" src="../pseudotime_Analysis_files/pseudotime_Analysis_58_0.png" /></p>
<div class="codehilite"><pre><span></span><code><span class="err">&lt;matplotlib.axes._subplots.AxesSubplot at 0x7fbfef430908&gt;</span>
</code></pre></div>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../neuronSubpop_Analysis/" class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Analysis of Neuron Subpopulations
              </div>
            </div>
          </a>
        
        
          <a href="../deSeq2Analysis_StarvationResponse/" class="md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Extraction of Differentially Expressed Genes Under Starvation
              </div>
            </div>
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/vendor.53cc9318.min.js"></script>
      <script src="../../assets/javascripts/bundle.e9c9f54f.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "../..",
          features: ['search.suggest', 'search.highlight'],
          search: Object.assign({
            worker: "../../assets/javascripts/worker/search.9c0e82ba.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>